<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Visualisation of biomolecular data</title>
  <meta name="description" content="Visualisation of biomolecular data">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Visualisation of biomolecular data" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Visualisation of biomolecular data" />
  <meta name="github-repo" content="lgatto/VisualisingBiomolecularData" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Visualisation of biomolecular data" />
  
  <meta name="twitter:description" content="Visualisation of biomolecular data" />
  

<meta name="author" content="Laurent Gatto">



  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="index.html">
<link rel="next" href="exploring-and-visualising-biomolecular-data.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i><b>1.1</b> Installation</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#questions"><i class="fa fa-check"></i><b>1.2</b> Questions</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.3</b> License</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#useful-references"><i class="fa fa-check"></i><b>1.4</b> Useful references</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="example-datasets.html"><a href="example-datasets.html"><i class="fa fa-check"></i><b>2</b> Example datasets</a><ul>
<li class="chapter" data-level="2.1" data-path="example-datasets.html"><a href="example-datasets.html#raw-ms-data"><i class="fa fa-check"></i><b>2.1</b> Raw MS data</a></li>
<li class="chapter" data-level="2.2" data-path="example-datasets.html"><a href="example-datasets.html#the-iprg-data"><i class="fa fa-check"></i><b>2.2</b> The iPRG data</a></li>
<li class="chapter" data-level="2.3" data-path="example-datasets.html"><a href="example-datasets.html#crc-training-data"><i class="fa fa-check"></i><b>2.3</b> CRC training data</a></li>
<li class="chapter" data-level="2.4" data-path="example-datasets.html"><a href="example-datasets.html#time-course-from-mulvey-et-al.-2015"><i class="fa fa-check"></i><b>2.4</b> Time course from Mulvey <em>et al.</em> 2015</a></li>
<li class="chapter" data-level="2.5" data-path="example-datasets.html"><a href="example-datasets.html#all-data"><i class="fa fa-check"></i><b>2.5</b> ALL data</a></li>
<li class="chapter" data-level="2.6" data-path="example-datasets.html"><a href="example-datasets.html#spatial-proteomics-data"><i class="fa fa-check"></i><b>2.6</b> Spatial proteomics data</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html"><i class="fa fa-check"></i><b>3</b> Exploring and visualising biomolecular data</a><ul>
<li class="chapter" data-level="" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html#raw-data"><i class="fa fa-check"></i>Raw data</a></li>
<li class="chapter" data-level="" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html#tracking-an-ion-from-the-chromatogram-to-the-ms2-spectra"><i class="fa fa-check"></i>Tracking an ion from the chromatogram to the MS2 spectra</a></li>
<li class="chapter" data-level="" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html#raw-data-qc"><i class="fa fa-check"></i>Raw data QC</a></li>
<li class="chapter" data-level="" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html#exploring-the-experimental-design"><i class="fa fa-check"></i>Exploring the experimental design</a></li>
<li class="chapter" data-level="" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html#quantitative-data"><i class="fa fa-check"></i>Quantitative data</a></li>
<li class="chapter" data-level="" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html#differential-expression"><i class="fa fa-check"></i>Differential expression</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="using-dplyr-for-data-manipulation.html"><a href="using-dplyr-for-data-manipulation.html"><i class="fa fa-check"></i><b>4</b> Using <code>dplyr</code> for data manipulation</a><ul>
<li class="chapter" data-level="4.1" data-path="using-dplyr-for-data-manipulation.html"><a href="using-dplyr-for-data-manipulation.html#selecting-columns-and-filtering-rows"><i class="fa fa-check"></i><b>4.1</b> Selecting columns and filtering rows</a></li>
<li class="chapter" data-level="4.2" data-path="using-dplyr-for-data-manipulation.html"><a href="using-dplyr-for-data-manipulation.html#pipes"><i class="fa fa-check"></i><b>4.2</b> Pipes</a></li>
<li class="chapter" data-level="4.3" data-path="using-dplyr-for-data-manipulation.html"><a href="using-dplyr-for-data-manipulation.html#mutate"><i class="fa fa-check"></i><b>4.3</b> Mutate</a></li>
<li class="chapter" data-level="4.4" data-path="using-dplyr-for-data-manipulation.html"><a href="using-dplyr-for-data-manipulation.html#split-apply-combine-data-analysis-and-the-summarize-function"><i class="fa fa-check"></i><b>4.4</b> Split-apply-combine data analysis and the summarize() function</a><ul>
<li><a href="using-dplyr-for-data-manipulation.html#the-summarize-function">The <code>summarize()</code> function</a></li>
<li class="chapter" data-level="" data-path="using-dplyr-for-data-manipulation.html"><a href="using-dplyr-for-data-manipulation.html#tallying"><i class="fa fa-check"></i>Tallying</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="plotting-in-r.html"><a href="plotting-in-r.html"><i class="fa fa-check"></i><b>5</b> Plotting in R</a><ul>
<li class="chapter" data-level="5.1" data-path="plotting-in-r.html"><a href="plotting-in-r.html#base-graphics"><i class="fa fa-check"></i><b>5.1</b> Base graphics</a></li>
<li class="chapter" data-level="5.2" data-path="plotting-in-r.html"><a href="plotting-in-r.html#plotting-with-ggplot2"><i class="fa fa-check"></i><b>5.2</b> Plotting with <code>ggplot2</code></a></li>
<li class="chapter" data-level="5.3" data-path="plotting-in-r.html"><a href="plotting-in-r.html#building-complex-plots"><i class="fa fa-check"></i><b>5.3</b> Building complex plots</a></li>
<li class="chapter" data-level="5.4" data-path="plotting-in-r.html"><a href="plotting-in-r.html#colour-scales"><i class="fa fa-check"></i><b>5.4</b> Colour scales</a></li>
<li class="chapter" data-level="5.5" data-path="plotting-in-r.html"><a href="plotting-in-r.html#customising-plots"><i class="fa fa-check"></i><b>5.5</b> Customising plots</a></li>
<li class="chapter" data-level="5.6" data-path="plotting-in-r.html"><a href="plotting-in-r.html#combine-ggplot-figures"><i class="fa fa-check"></i><b>5.6</b> Combine <code>ggplot</code> figures</a></li>
<li class="chapter" data-level="5.7" data-path="plotting-in-r.html"><a href="plotting-in-r.html#saving-your-figures"><i class="fa fa-check"></i><b>5.7</b> Saving your figures</a></li>
<li class="chapter" data-level="5.8" data-path="plotting-in-r.html"><a href="plotting-in-r.html#exercises"><i class="fa fa-check"></i><b>5.8</b> Exercises</a></li>
<li class="chapter" data-level="5.9" data-path="plotting-in-r.html"><a href="plotting-in-r.html#references"><i class="fa fa-check"></i><b>5.9</b> References</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="tools-and-plots.html"><a href="tools-and-plots.html"><i class="fa fa-check"></i><b>6</b> Tools and plots</a><ul>
<li class="chapter" data-level="6.1" data-path="tools-and-plots.html"><a href="tools-and-plots.html#transformations"><i class="fa fa-check"></i><b>6.1</b> Transformations</a><ul>
<li class="chapter" data-level="6.1.1" data-path="tools-and-plots.html"><a href="tools-and-plots.html#expression-data"><i class="fa fa-check"></i><b>6.1.1</b> Expression data</a></li>
<li class="chapter" data-level="6.1.2" data-path="tools-and-plots.html"><a href="tools-and-plots.html#fold-changes"><i class="fa fa-check"></i><b>6.1.2</b> Fold changes</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="tools-and-plots.html"><a href="tools-and-plots.html#comparing-samples-and-linear-models"><i class="fa fa-check"></i><b>6.2</b> Comparing samples and linear models</a></li>
<li class="chapter" data-level="6.3" data-path="tools-and-plots.html"><a href="tools-and-plots.html#plots-for-statistical-analyses"><i class="fa fa-check"></i><b>6.3</b> Plots for statistical analyses</a></li>
<li class="chapter" data-level="6.4" data-path="tools-and-plots.html"><a href="tools-and-plots.html#visualising-intersections"><i class="fa fa-check"></i><b>6.4</b> Visualising intersections</a></li>
<li class="chapter" data-level="6.5" data-path="tools-and-plots.html"><a href="tools-and-plots.html#unsupervised-learning"><i class="fa fa-check"></i><b>6.5</b> Unsupervised learning</a></li>
<li class="chapter" data-level="6.6" data-path="tools-and-plots.html"><a href="tools-and-plots.html#hierarchical-clustering"><i class="fa fa-check"></i><b>6.6</b> Hierarchical clustering</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#how-does-hierarchical-clustering-work"><i class="fa fa-check"></i>How does hierarchical clustering work</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#defining-clusters"><i class="fa fa-check"></i>Defining clusters</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#effect-of-distance-and-clustering-algorithm"><i class="fa fa-check"></i>Effect of distance and clustering algorithm</a></li>
</ul></li>
<li class="chapter" data-level="6.7" data-path="tools-and-plots.html"><a href="tools-and-plots.html#k-means-clustering"><i class="fa fa-check"></i><b>6.7</b> k-means clustering</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#how-does-k-means-work"><i class="fa fa-check"></i>How does k-means work</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#model-selection"><i class="fa fa-check"></i>Model selection</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#how-to-determine-the-number-of-clusters"><i class="fa fa-check"></i>How to determine the number of clusters</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#challenge"><i class="fa fa-check"></i>Challenge</a></li>
</ul></li>
<li class="chapter" data-level="6.8" data-path="tools-and-plots.html"><a href="tools-and-plots.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>6.8</b> Principal component analysis (PCA)</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#how-does-it-work"><i class="fa fa-check"></i>How does it work</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#visualisation"><i class="fa fa-check"></i>Visualisation</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#data-pre-processing"><i class="fa fa-check"></i>Data pre-processing</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#final-comments-on-pca"><i class="fa fa-check"></i>Final comments on PCA</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#challenge-1"><i class="fa fa-check"></i>Challenge</a></li>
</ul></li>
<li class="chapter" data-level="6.9" data-path="tools-and-plots.html"><a href="tools-and-plots.html#t-distributed-stochastic-neighbour-embedding"><i class="fa fa-check"></i><b>6.9</b> t-Distributed Stochastic Neighbour Embedding</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#parameter-tuning"><i class="fa fa-check"></i>Parameter tuning</a></li>
</ul></li>
<li class="chapter" data-level="6.10" data-path="tools-and-plots.html"><a href="tools-and-plots.html#the-curse-of-dimensionality"><i class="fa fa-check"></i><b>6.10</b> The curse of dimensionality</a></li>
<li class="chapter" data-level="6.11" data-path="tools-and-plots.html"><a href="tools-and-plots.html#pca-loadings-vs-differential-expression"><i class="fa fa-check"></i><b>6.11</b> PCA loadings vs differential expression</a></li>
<li class="chapter" data-level="6.12" data-path="tools-and-plots.html"><a href="tools-and-plots.html#heatmaps"><i class="fa fa-check"></i><b>6.12</b> Heatmaps</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html"><i class="fa fa-check"></i><b>7</b> Using R and Bioconductor for MS-based proteomics</a><ul>
<li class="chapter" data-level="7.1" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#introduction-1"><i class="fa fa-check"></i><b>7.1</b> Introduction</a></li>
<li class="chapter" data-level="7.2" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#reading-ms-data-into-r"><i class="fa fa-check"></i><b>7.2</b> Reading MS data into R</a><ul>
<li class="chapter" data-level="" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#raw-data-1"><i class="fa fa-check"></i>Raw data</a></li>
<li class="chapter" data-level="" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#identification-data"><i class="fa fa-check"></i>Identification data</a></li>
<li class="chapter" data-level="" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#quantitative-data-1"><i class="fa fa-check"></i>Quantitative data</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#manipulating-and-visualising-raw-data"><i class="fa fa-check"></i><b>7.3</b> Manipulating and visualising raw data</a></li>
<li class="chapter" data-level="7.4" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#ms-maps"><i class="fa fa-check"></i><b>7.4</b> MS maps</a></li>
<li class="chapter" data-level="7.5" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#peptide-spetrum-matches"><i class="fa fa-check"></i><b>7.5</b> Peptide spetrum matches</a></li>
<li class="chapter" data-level="7.6" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#ms-imaging"><i class="fa fa-check"></i><b>7.6</b> MS imaging</a></li>
<li class="chapter" data-level="7.7" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#missing-values"><i class="fa fa-check"></i><b>7.7</b> Missing values</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html"><i class="fa fa-check"></i><b>8</b> Interactive visualisation</a><ul>
<li class="chapter" data-level="8.1" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#interactivity-graphs"><i class="fa fa-check"></i><b>8.1</b> Interactivity graphs</a><ul>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#interactivity"><i class="fa fa-check"></i>Interactivity</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#layers"><i class="fa fa-check"></i>Layers</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#more-components"><i class="fa fa-check"></i>More components</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#interactive-apps"><i class="fa fa-check"></i><b>8.2</b> Interactive apps</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#motivating-examples"><i class="fa fa-check"></i>Motivating examples</a><ul>
<li class="chapter" data-level="8.2.1" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#longitudinal-proteomics-qc"><i class="fa fa-check"></i><b>8.2.1</b> Longitudinal proteomics QC</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#spatial-proteomics"><i class="fa fa-check"></i>Spatial proteomics</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#introduction-2"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#creation-of-our-fist-shiny-app"><i class="fa fa-check"></i>Creation of our fist shiny app</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#the-shiny-infrastructure"><i class="fa fa-check"></i>The shiny infrastructure</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#another-shiny-app"><i class="fa fa-check"></i>Another shiny app</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#single-file-app"><i class="fa fa-check"></i>Single-file app</a></li>
<li><a href="interactive-visualisation.html#theres-more-to-shiny">There’s more to <code>shiny</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="session-information.html"><a href="session-information.html"><i class="fa fa-check"></i><b>9</b> Session information</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Visualisation of biomolecular data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="example-datasets" class="section level1">
<h1><span class="header-section-number">Chapter 2</span> Example datasets</h1>
<p>We will used various datasets throughout the course. These data are briefly described below, and we will explore them through various visualisations later.</p>
<div id="raw-ms-data" class="section level2">
<h2><span class="header-section-number">2.1</span> Raw MS data</h2>
<p>Section <em>Using R and Bioconductor for MS-based proteomics</em> shows how to visualise raw mass spectrometry data. The raw data that will be using come from the <em><a href="http://bioconductor.org/packages/msdata">msdata</a></em> <em><a href="http://bioconductor.org/packages/MSnbase">MSnbase</a></em> packages. These data will be introduced later.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;msdata&quot;</span>)</code></pre></div>
</div>
<div id="the-iprg-data" class="section level2">
<h2><span class="header-section-number">2.2</span> The iPRG data</h2>
<p>This iPRG data is a spiked-in exeriment, where 6 proteins where spiked at different ratios in a Yeast proteome background. Each run was repeated in triplicates and order was randomised. Participants in the study were asked to identify the differentially abundant spiked-in proteins.</p>
<div class="figure">
<img src="img/iprg.png" />

</div>
<blockquote>
<p>Choi M, Eren-Dogu ZF, Colangelo C, Cottrell J, Hoopmann MR, Kapp EA, Kim S, Lam H, Neubert TA, Palmblad M, Phinney BS, Weintraub ST, MacLean B, Vitek O. <em>ABRF Proteome Informatics Research Group (iPRG) 2015 Study: Detection of Differentially Abundant Proteins in Label-Free Quantitative LC-MS/MS Experiments</em>. J Proteome Res. 2017 Feb 3;16(2):945-957. doi: <a href="https://doi.org/10.1021/acs.jproteome.6b00881">10.1021/acs.jproteome.6b00881</a> Epub 2017 Jan 3. PMID: 27990823.</p>
</blockquote>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">iprg &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;http://bit.ly/VisBiomedDataIprgCsv&quot;</span>)
<span class="kw">head</span>(iprg)</code></pre></div>
<pre><code>##                Protein Log2Intensity                       Run  Condition
## 1 sp|D6VTK4|STE2_YEAST      26.81232 JD_06232014_sample1_B.raw Condition1
## 2 sp|D6VTK4|STE2_YEAST      26.60786 JD_06232014_sample1_C.raw Condition1
## 3 sp|D6VTK4|STE2_YEAST      26.58301 JD_06232014_sample1-A.raw Condition1
## 4 sp|D6VTK4|STE2_YEAST      26.83563 JD_06232014_sample2_A.raw Condition2
## 5 sp|D6VTK4|STE2_YEAST      26.79430 JD_06232014_sample2_B.raw Condition2
## 6 sp|D6VTK4|STE2_YEAST      26.60863 JD_06232014_sample2_C.raw Condition2
##   BioReplicate Intensity TechReplicate
## 1            1 117845016             B
## 2            1 102273602             C
## 3            1 100526837             A
## 4            2 119765106             A
## 5            2 116382798             B
## 6            2 102328260             C</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(iprg)</code></pre></div>
<pre><code>## [1] 36321     7</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(iprg<span class="op">$</span>Condition, iprg<span class="op">$</span>TechReplicate)</code></pre></div>
<pre><code>##             
##                 A    B    C
##   Condition1 3026 3026 3027
##   Condition2 3027 3027 3027
##   Condition3 3027 3027 3027
##   Condition4 3027 3026 3027</code></pre>
<p>This data is in the so-called long format. In some applications, it is more convenient to have the data in wide format, where rows contain the protein expression data for all samples.</p>
<p>Let’s start by simplifying the data to keep only the relevant columns:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(iprg2 &lt;-<span class="st"> </span>iprg[, <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">6</span>)])</code></pre></div>
<pre><code>##                Protein                       Run Intensity
## 1 sp|D6VTK4|STE2_YEAST JD_06232014_sample1_B.raw 117845016
## 2 sp|D6VTK4|STE2_YEAST JD_06232014_sample1_C.raw 102273602
## 3 sp|D6VTK4|STE2_YEAST JD_06232014_sample1-A.raw 100526837
## 4 sp|D6VTK4|STE2_YEAST JD_06232014_sample2_A.raw 119765106
## 5 sp|D6VTK4|STE2_YEAST JD_06232014_sample2_B.raw 116382798
## 6 sp|D6VTK4|STE2_YEAST JD_06232014_sample2_C.raw 102328260</code></pre>
<p>We can convert the <code>iPRG</code> into a wide format with <code>tidyr::spread</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;tidyr&quot;</span>)
iprg3 &lt;-<span class="st"> </span><span class="kw">spread</span>(iprg2, <span class="dt">key =</span> Run, <span class="dt">value =</span> Intensity)
<span class="kw">head</span>(iprg3)</code></pre></div>
<pre><code>##                 Protein JD_06232014_sample1-A.raw
## 1  sp|D6VTK4|STE2_YEAST                 100526837
## 2  sp|O13297|CET1_YEAST                  27598550
## 3  sp|O13329|FOB1_YEAST                  11625198
## 4  sp|O13539|THP2_YEAST                  20606703
## 5 sp|O13547|CCW14_YEAST                 145493943
## 6 sp|O13563|RPN13_YEAST                  75530595
##   JD_06232014_sample1_B.raw JD_06232014_sample1_C.raw
## 1                 117845016                 102273602
## 2                  27618234                  26774670
## 3                  10892143                  16948335
## 4                 192490784                 175282010
## 5                 156581624                 117211277
## 6                  71664672                  82193735
##   JD_06232014_sample2_A.raw JD_06232014_sample2_B.raw
## 1                 119765106                 116382798
## 2                  24114625                  26803677
## 3                   9448487                  11334921
## 4                  61781078                  72052121
## 5                 151407246                 126364703
## 6                  62998676                  73431260
##   JD_06232014_sample2_C.raw JD_06232014_sample3_A.raw
## 1                 102328260                 103830944
## 2                  25055912                  21977898
## 3                  14304554                  19285356
## 4                  62895992                  52653164
## 5                 123774524                  99222081
## 6                  67612042                  75916420
##   JD_06232014_sample3_B.raw JD_06232014_sample3_C.raw
## 1                  94660680                  96919972
## 2                  25901361                  26545544
## 3                  10230925                  13985282
## 4                  84856844                  63400641
## 5                 113287367                 129669642
## 6                  68877388                  72632994
##   JD_06232014_sample4-A.raw JD_06232014_sample4_B.raw
## 1                 105724288                 102150172
## 2                  23860882                  26305424
## 3                   8592981                   9414560
## 4                  35363603                 109094905
## 5                 141404829                 132444723
## 6                  57600421                  70708409
##   JD_06232014_sample4_C.raw
## 1                  87702341
## 2                  26612349
## 3                  12783705
## 4                  63392610
## 5                 117065261
## 6                  69675572</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">nrow</span>(iprg3)</code></pre></div>
<pre><code>## [1] 3027</code></pre>
<div class="figure">
<img src="img/spread.png" alt="Spreading data from long to wide format" />
<p class="caption">Spreading data from long to wide format</p>
</div>
<p>Indeed, we started with</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(<span class="kw">unique</span>(iprg<span class="op">$</span>Protein))</code></pre></div>
<pre><code>## [1] 3027</code></pre>
<p>unique proteins, which corresponds to the number of rows in the new wide dataset.</p>
<p>The long format is ideal when using <code>ggplot2</code>, as we will see in a later chapter. The wide format has also advantages. For example, it becomes straighforward to verify if there are proteins that haven’t been quantified in some samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">(k &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">is.na</span>(iprg3), <span class="dt">arr.ind =</span> <span class="kw">dim</span>(iprg3)))</code></pre></div>
<pre><code>##       row col
## [1,] 2721   2
## [2,] 2721   3
## [3,]  652  12</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">iprg3[<span class="kw">unique</span>(k[, <span class="st">&quot;row&quot;</span>]), ]</code></pre></div>
<pre><code>##                    Protein JD_06232014_sample1-A.raw
## 2721  sp|Q08236|AVO1_YEAST                        NA
## 652  sp|P28320|CWC16_YEAST                   1108670
##      JD_06232014_sample1_B.raw JD_06232014_sample1_C.raw
## 2721                        NA                  10989286
## 652                   961273.8                   1229085
##      JD_06232014_sample2_A.raw JD_06232014_sample2_B.raw
## 2721                 5407419.7                   2041424
## 652                   449839.1                   1334009
##      JD_06232014_sample2_C.raw JD_06232014_sample3_A.raw
## 2721                  13108081                   4677052
## 652                    1633894                   1493440
##      JD_06232014_sample3_B.raw JD_06232014_sample3_C.raw
## 2721                  10308957                  13639794
## 652                    2205688                   2120321
##      JD_06232014_sample4-A.raw JD_06232014_sample4_B.raw
## 2721                10377088.4                   2592511
## 652                   757082.7                        NA
##      JD_06232014_sample4_C.raw
## 2721                  10420338
## 652                    1264712</code></pre>
<p>The opposite operation to <code>spread</code> is <code>gather</code>, also from the <code>tidyr</code> package:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(iprg4 &lt;-<span class="st"> </span><span class="kw">gather</span>(iprg3, <span class="dt">key =</span> Run, <span class="dt">value =</span> Intensity, <span class="op">-</span>Protein))</code></pre></div>
<pre><code>##                 Protein                       Run Intensity
## 1  sp|D6VTK4|STE2_YEAST JD_06232014_sample1-A.raw 100526837
## 2  sp|O13297|CET1_YEAST JD_06232014_sample1-A.raw  27598550
## 3  sp|O13329|FOB1_YEAST JD_06232014_sample1-A.raw  11625198
## 4  sp|O13539|THP2_YEAST JD_06232014_sample1-A.raw  20606703
## 5 sp|O13547|CCW14_YEAST JD_06232014_sample1-A.raw 145493943
## 6 sp|O13563|RPN13_YEAST JD_06232014_sample1-A.raw  75530595</code></pre>
<div class="figure">
<img src="img/gather.png" alt="Gathering data from wide to long format" />
<p class="caption">Gathering data from wide to long format</p>
</div>
<p>The two lond datasets, <code>iprg2</code> and <code>iprg4</code> are different due to the missing values shown above.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">nrow</span>(iprg2)</code></pre></div>
<pre><code>## [1] 36321</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">nrow</span>(iprg4)</code></pre></div>
<pre><code>## [1] 36324</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">nrow</span>(<span class="kw">na.omit</span>(iprg4))</code></pre></div>
<pre><code>## [1] 36321</code></pre>
<p>which can be accounted for by removing rows with missing values by setting <code>na.rm = TRUE</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(iprg5 &lt;-<span class="st"> </span><span class="kw">gather</span>(iprg3, <span class="dt">key =</span> Run, <span class="dt">value =</span> Intensity, <span class="op">-</span>Protein, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>))</code></pre></div>
<pre><code>##                 Protein                       Run Intensity
## 1  sp|D6VTK4|STE2_YEAST JD_06232014_sample1-A.raw 100526837
## 2  sp|O13297|CET1_YEAST JD_06232014_sample1-A.raw  27598550
## 3  sp|O13329|FOB1_YEAST JD_06232014_sample1-A.raw  11625198
## 4  sp|O13539|THP2_YEAST JD_06232014_sample1-A.raw  20606703
## 5 sp|O13547|CCW14_YEAST JD_06232014_sample1-A.raw 145493943
## 6 sp|O13563|RPN13_YEAST JD_06232014_sample1-A.raw  75530595</code></pre>
</div>
<div id="crc-training-data" class="section level2">
<h2><span class="header-section-number">2.3</span> CRC training data</h2>
<p>This dataset comes from the <em><a href="http://bioconductor.org/packages/MSstatsBioData">MSstatsBioData</a></em> package and was generated as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;MSstats&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;MSstatsBioData&quot;</span>)
<span class="kw">data</span>(SRM_crc_training)
Quant &lt;-<span class="st"> </span><span class="kw">dataProcess</span>(SRM_crc_training)
subjectQuant &lt;-<span class="st"> </span><span class="kw">quantification</span>(Quant)</code></pre></div>
<p>It provides quantitative information for 72 proteins, including two standard proteins, <em>AIAG-Bovine</em> and <em>FETUA-Bovine</em>. These proteins were targeted for plasma samples with SRM with isotope labeled reference peptides in order to identify candidate protein biomarker for non-invasive detection of CRC. The training cohort included 100 subjects in control group and 100 subjects with CRC. Each sample for subject was measured in a single injection without technical replicate. The training cohort was analyzed with Skyline. The dataset was already normalized as described in manuscript. User do not need extra normalization. NAs should be considered as censored missing. Two standard proteins can be removed for statistical analysis.</p>
<p>Clinical information where added manually thereafter.</p>
<p>To load this dataset:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">crcdf &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;http://bit.ly/VisBiomedDataCrcCsv&quot;</span>)
crcdf[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</code></pre></div>
<pre><code>##       A1AG2      AFM     AHSG
## 1  14.23816 16.10302 19.95179
## 2  15.02411 16.02071 19.71592
## 3  15.63136 16.14380 19.71085
## 4  15.40137 16.27642 19.70438
## 5  16.00316 16.95821 20.42033
## 6  13.93242 16.52772 19.88985
## 7  14.34155 16.43029 20.10060
## 8  13.57086 17.11052 19.93833
## 9  15.83348 15.88189 18.64270
## 10 15.37996 16.22621 20.09992</code></pre>
<p>This dataset is in the wide format. It contains the intensity of the proteins in columns 1 to 72 for each of the 200 samples along the rows. Generally, omics datasets contain the features (proteins, transcripts, …) along the rows and the samples along the columns.</p>
<p>In columns 73 to 79, we sample metadata.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">crcdf[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">73</span><span class="op">:</span><span class="dv">79</span>]</code></pre></div>
<pre><code>##    Sample Group Age Gender Cancer_stage Tumour_location Sub_group
## 1   P1A10   CRC  60 female            1           colon       CRC
## 2    P1A2   CRC  70   male            1          rectum       CRC
## 3    P1A4   CRC  65   male            1          rectum       CRC
## 4    P1A6   CRC  65 female            4           colon       CRC
## 5   P1B12   CRC  62 female            3           colon       CRC
## 6    P1B2   CRC  55   male            2           colon       CRC
## 7    P1B3   CRC  61   male            2          rectum       CRC
## 8    P1B6   CRC  52   male            4          rectum       CRC
## 9    P1B9   CRC  89 female            4           colon       CRC
## 10  P1C11   CRC  81   male            1          rectum       CRC</code></pre>
<p>A widely used data structure for omics data follows the convention described in the figure below:</p>
<div class="figure">
<img src="img/msnset.png" alt="An eSet-type of expression data container" />
<p class="caption">An eSet-type of expression data container</p>
</div>
<p>This typical omics data structure, as defined by the <code>eSet</code> class in the Bioconductor <code>Biobase</code> package, is represented below. It’s main features are</p>
<ul>
<li><p>An assay data slot containing the quantitative omics data (expression data), stored as a <code>matrix</code> and accessible with <code>exprs</code>. Features defined along the rows and samples along the columns.</p></li>
<li><p>A sample metadata slot containing sample co-variates, stored as an annotated <code>data.frame</code> and accessible with <code>pData</code>. This data frame is stored with rows representing samples and sample covariate along the columns, and its rows match the expression data columns exactly.</p></li>
<li><p>A feature metadata slot containing feature co-variates, stored as an annotated <code>data.frame</code> and accessible with <code>fData</code>. This dataframe’s rows match the expression data rows exactly.</p></li>
</ul>
<p>The coordinated nature of the high throughput data guarantees that the dimensions of the different slots will always match (i.e the columns in the expression data and then rows in the sample metadata, as well as the rows in the expression data and feature metadata) during data manipulation. The metadata slots can grow additional co-variates (columns) without affecting the other structures.</p>
<p>Below, we show how to transform the <code>crc</code> dataset into an <code>MSnSet</code> (implementing the data structure above for quantitative proteomics data) using the <code>readMSnSet2</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;MSnbase&quot;</span>)
i &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="dv">72</span> ## expression columns
e &lt;-<span class="st"> </span><span class="kw">t</span>(crcdf[, i]) ## expression data
<span class="kw">colnames</span>(e) &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="dv">200</span>
crc &lt;-<span class="st"> </span><span class="kw">readMSnSet2</span>(<span class="kw">data.frame</span>(e), <span class="dt">e =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">200</span>)

pd &lt;-<span class="st"> </span>crcdf[, <span class="op">-</span>i]
<span class="kw">rownames</span>(pd) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;X&quot;</span>, <span class="kw">rownames</span>(pd))
<span class="kw">pData</span>(crc) &lt;-<span class="st"> </span>pd
crc</code></pre></div>
<pre><code>## MSnSet (storageMode: lockedEnvironment)
## assayData: 72 features, 200 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: X1 X2 ... X200 (200 total)
##   varLabels: Sample Group ... Sub_group (7 total)
##   varMetadata: labelDescription
## featureData: none
## experimentData: use &#39;experimentData(object)&#39;
## Annotation:  
## - - - Processing information - - -
##  MSnbase version: 2.6.0</code></pre>
<p>Let’s also set the sample names.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sampleNames</span>(crc) &lt;-<span class="st"> </span>crc<span class="op">$</span>Sample</code></pre></div>
<p>To download and load the <code>MSnSet</code> direcly:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">download.file</span>(<span class="st">&quot;http://bit.ly/VisBiomedDataCrcMSnSet&quot;</span>, <span class="st">&quot;./data/crc.rda&quot;</span>)
<span class="kw">load</span>(<span class="st">&quot;./data/crc.rda&quot;</span>)
crc</code></pre></div>
<pre><code>## MSnSet (storageMode: lockedEnvironment)
## assayData: 72 features, 200 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: P1A10 P1A2 ... P3H6 (200 total)
##   varLabels: Sample Group ... Sub_group (7 total)
##   varMetadata: labelDescription
## featureData: none
## experimentData: use &#39;experimentData(object)&#39;
## Annotation:  
## - - - Processing information - - -
##  MSnbase version: 2.6.0</code></pre>
<p>Reference:</p>
<blockquote>
<p>See Surinova, S. et al. (2015) <em>Prediction of colorectal cancer diagnosis based on circulating plasma proteins</em>. EMBO Mol. Med., 7, 1166–1178 for details.</p>
</blockquote>
</div>
<div id="time-course-from-mulvey-et-al.-2015" class="section level2">
<h2><span class="header-section-number">2.4</span> Time course from Mulvey <em>et al.</em> 2015</h2>
<p>This data comes from</p>
<blockquote>
<p>Mulvey CM, Schröter C, Gatto L, Dikicioglu D, Fidaner IB, Christoforou A, Deery MJ, Cho LT, Niakan KK, Martinez-Arias A, Lilley KS. Dynamic Proteomic Profiling of Extra-Embryonic Endoderm Differentiation in Mouse Embryonic Stem Cells. Stem Cells. 2015 Sep;33(9):2712-25. doi: <a href="https://doi.org/10.1002/stem.2067">10.1002/stem.2067</a>. Epub 2015 Jun 23. PMID: 26059426.</p>
</blockquote>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;pRolocdata&quot;</span>)
<span class="kw">data</span>(mulvey2015norm)</code></pre></div>
<p>This <code>MSnSet</code>, available from the <em><a href="http://bioconductor.org/packages/pRolocdata">pRolocdata</a></em> package, measured the expression profiles of 2337 proteins along 6 time points in triplicate.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MSnbase<span class="op">::</span><span class="kw">exprs</span>(mulvey2015norm)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</code></pre></div>
<pre><code>##          rep1_0hr rep1_16hr rep1_24hr
## P48432   2.479592  1.698630 1.0350877
## Q62315-2 1.979592  1.342466 0.9181287
## P55821   1.780612  1.767123 1.1578947
## P17809   1.637755  1.157534 0.9941520
## Q8K3F7   1.852041  1.623288 1.3040936</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">pData</span>(mulvey2015norm)</code></pre></div>
<pre><code>##           rep times cond
## rep1_0hr    1     1    1
## rep1_16hr   1     2    1
## rep1_24hr   1     3    1
## rep1_48hr   1     4    1
## rep1_72hr   1     5    1
## rep1_XEN    1     6    1
## rep2_0hr    2     1    1
## rep2_16hr   2     2    1
## rep2_24hr   2     3    1
## rep2_48hr   2     4    1
## rep2_72hr   2     5    1
## rep2_XEN    2     6    1
## rep3_0hr    3     1    1
## rep3_16hr   3     2    1
## rep3_24hr   3     3    1
## rep3_48hr   3     4    1
## rep3_72hr   3     5    1
## rep3_XEN    3     6    1</code></pre>
</div>
<div id="all-data" class="section level2">
<h2><span class="header-section-number">2.5</span> ALL data</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;ALL&quot;</span>)
<span class="kw">data</span>(ALL)
ALL</code></pre></div>
<pre><code>## ExpressionSet (storageMode: lockedEnvironment)
## assayData: 12625 features, 128 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: 01005 01010 ... LAL4 (128 total)
##   varLabels: cod diagnosis ... date last seen (21 total)
##   varMetadata: labelDescription
## featureData: none
## experimentData: use &#39;experimentData(object)&#39;
##   pubMedIds: 14684422 16243790 
## Annotation: hgu95av2</code></pre>
<p>From the documentation page:</p>
<blockquote>
<p>The <em>Acute Lymphoblastic Leukemia Data from the Ritz Laboratory</em> consist of microarrays from 128 different individuals with acute lymphoblastic leukemia (ALL). A number of additional covariates are available. The data have been normalized (using rma) and it is the jointly normalized data that are available here.</p>
</blockquote>
<p>The <code>ALL</code> data is of class <code>ExpressionSet</code>, which implements the data structure above for microarray expression data, and contains normalised and summarised transcript intensities.</p>
<p>Below, we select will patients with B-cell lymphomas and BCR/ABL abnormality and negative controls.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(ALL<span class="op">$</span>BT)</code></pre></div>
<pre><code>## 
##  B B1 B2 B3 B4  T T1 T2 T3 T4 
##  5 19 36 23 12  5  1 15 10  2</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(ALL<span class="op">$</span>mol.biol)</code></pre></div>
<pre><code>## 
## ALL1/AF4  BCR/ABL E2A/PBX1      NEG   NUP-98  p15/p16 
##       10       37        5       74        1        1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ALL_bcrneg &lt;-<span class="st"> </span>ALL[, ALL<span class="op">$</span>mol.biol <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;NEG&quot;</span>, <span class="st">&quot;BCR/ABL&quot;</span>) <span class="op">&amp;</span><span class="st"> </span><span class="kw">grepl</span>(<span class="st">&quot;B&quot;</span>, ALL<span class="op">$</span>BT)]
ALL_bcrneg<span class="op">$</span>mol.biol &lt;-<span class="st"> </span><span class="kw">factor</span>(ALL_bcrneg<span class="op">$</span>mol.biol)</code></pre></div>
<p>We then use the <em><a href="http://bioconductor.org/packages/limma">limma</a></em> package to</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;limma&quot;</span>)
design &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="dv">0</span><span class="op">+</span>ALL_bcrneg<span class="op">$</span>mol.biol)
<span class="kw">colnames</span>(design) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;BCR.ABL&quot;</span>, <span class="st">&quot;NEG&quot;</span>)

## Step1: linear model. lmFit is a wrapper around lm in R
fit1 &lt;-<span class="st"> </span><span class="kw">lmFit</span>(ALL_bcrneg, design)

## Step 2: fit contrasts: find genes that respond to estrogen
contrast.matrix &lt;-<span class="st"> </span><span class="kw">makeContrasts</span>(BCR.ABL<span class="op">-</span>NEG, <span class="dt">levels =</span> design)
fit2 &lt;-<span class="st"> </span><span class="kw">contrasts.fit</span>(fit1, contrast.matrix)

## Step3: add empirical Bayes moderation
fit3 &lt;-<span class="st"> </span><span class="kw">eBayes</span>(fit2)

## Extract results and set them to the feature data
res &lt;-<span class="st"> </span><span class="kw">topTable</span>(fit3, <span class="dt">n =</span> <span class="ot">Inf</span>)
<span class="kw">fData</span>(ALL_bcrneg) &lt;-<span class="st"> </span>res[<span class="kw">featureNames</span>(ALL_bcrneg), ]</code></pre></div>
<p>This annotated <code>ExpressionSet</code> can be reproduced as shown above or downloaded and loaded using</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">download.file</span>(<span class="st">&quot;http://bit.ly/VisBiomedDataALL_bcrneg&quot;</span>,  <span class="st">&quot;./data/ALL_bcrneg.rda&quot;</span>)
<span class="kw">load</span>(<span class="st">&quot;./data/ALL_bcrneg.rda&quot;</span>)</code></pre></div>
<p>Reference:</p>
<blockquote>
<p>Sabina Chiaretti, Xiaochun Li, Robert Gentleman, Antonella Vitale, Marco Vignetti, Franco Mandelli, Jerome Ritz, and Robin Foa <em>Gene expression profile of adult T-cell acute lymphocytic leukemia identifies distinct subsets of patients with different response to therapy and survival</em>. Blood, 1 April 2004, Vol. 103, No. 7.</p>
</blockquote>
</div>
<div id="spatial-proteomics-data" class="section level2">
<h2><span class="header-section-number">2.6</span> Spatial proteomics data</h2>
<p>The goal of spatial proteomics is to study the sub-cellular localisation of proteins. The data below comes from</p>
<blockquote>
<p>Christoforou A, Mulvey CM, Breckels LM, Geladaki A, Hurrell T, Hayward PC, Naake T, Gatto L, Viner R, Martinez Arias A, Lilley KS. A draft map of the mouse pluripotent stem cell spatial proteome. Nat Commun. 2016 Jan 12;7:8992. doi: <a href="https://doi.org/10.1038/ncomms9992">10.1038/ncomms9992</a>. PMID: 26754106; PMCID: PMC4729960.</p>
</blockquote>
<p>and investigated the sub-cellular location of over 5000 proteins in mouse pluripotent stem cells using a mass spectrometry-based technique called hyperLOPIT.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;pRolocdata&quot;</span>)
<span class="kw">data</span>(hyperLOPIT2015)
<span class="kw">dim</span>(hyperLOPIT2015)</code></pre></div>
<pre><code>## [1] 5032   20</code></pre>
<p>In addition to the quantitative data, another important piece of data here are the spatial markers, proteins for which we can confidently assign them a sub-cellular location apriori. These markers are defined in the <code>markers</code> feature variable:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(<span class="kw">fData</span>(hyperLOPIT2015)<span class="op">$</span>markers)</code></pre></div>
<pre><code>## 
##                          40S Ribosome 
##                                    27 
##                          60S Ribosome 
##                                    43 
##                    Actin cytoskeleton 
##                                    13 
##                               Cytosol 
##                                    43 
## Endoplasmic reticulum/Golgi apparatus 
##                                   107 
##                              Endosome 
##                                    13 
##                  Extracellular matrix 
##                                    13 
##                              Lysosome 
##                                    33 
##                         Mitochondrion 
##                                   383 
##                   Nucleus - Chromatin 
##                                    64 
##               Nucleus - Non-chromatin 
##                                    85 
##                            Peroxisome 
##                                    17 
##                       Plasma membrane 
##                                    51 
##                            Proteasome 
##                                    34 
##                               unknown 
##                                  4106</code></pre>
<p>Here, the columns of the data represent fractions along a density gradient used to separate the sub-cellular proteome presented in this data.</p>
<p>We will be using a simplified version of this data in the next chapters. Below, we retain the first replicate (<code>hyperLOPIT2015$Replicate == 1]</code>) and the marker proteins (using the <code>markerMSnSet</code> function):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hlm &lt;-<span class="st"> </span><span class="kw">markerMSnSet</span>(hyperLOPIT2015[, hyperLOPIT2015<span class="op">$</span>Replicate <span class="op">==</span><span class="st"> </span><span class="dv">1</span>])
<span class="kw">dim</span>(hlm)</code></pre></div>
<pre><code>## [1] 926  10</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">getMarkerClasses</span>(hlm) ## no unknowns anymore</code></pre></div>
<pre><code>##  [1] &quot;40S Ribosome&quot;                         
##  [2] &quot;60S Ribosome&quot;                         
##  [3] &quot;Actin cytoskeleton&quot;                   
##  [4] &quot;Cytosol&quot;                              
##  [5] &quot;Endoplasmic reticulum/Golgi apparatus&quot;
##  [6] &quot;Endosome&quot;                             
##  [7] &quot;Extracellular matrix&quot;                 
##  [8] &quot;Lysosome&quot;                             
##  [9] &quot;Mitochondrion&quot;                        
## [10] &quot;Nucleus - Chromatin&quot;                  
## [11] &quot;Nucleus - Non-chromatin&quot;              
## [12] &quot;Peroxisome&quot;                           
## [13] &quot;Plasma membrane&quot;                      
## [14] &quot;Proteasome&quot;</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="exploring-and-visualising-biomolecular-data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
