<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Visualisation of biomolecular data</title>
  <meta name="description" content="Visualisation of biomolecular data">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Visualisation of biomolecular data" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Visualisation of biomolecular data" />
  <meta name="github-repo" content="lgatto/VisualisingBiomolecularData" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Visualisation of biomolecular data" />
  
  <meta name="twitter:description" content="Visualisation of biomolecular data" />
  

<meta name="author" content="Laurent Gatto">


<meta name="date" content="2018-04-27">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="introduction.html">
<link rel="next" href="plotting-data.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Preface</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i><b>1.1</b> Installation</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.2</b> License</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>2</b> Introduction</a></li>
<li class="chapter" data-level="3" data-path="example-datasets.html"><a href="example-datasets.html"><i class="fa fa-check"></i><b>3</b> Example datasets</a><ul>
<li class="chapter" data-level="3.1" data-path="example-datasets.html"><a href="example-datasets.html#crc-training-data"><i class="fa fa-check"></i><b>3.1</b> CRC training data</a></li>
<li class="chapter" data-level="3.2" data-path="example-datasets.html"><a href="example-datasets.html#all-data"><i class="fa fa-check"></i><b>3.2</b> ALL data</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="plotting-data.html"><a href="plotting-data.html"><i class="fa fa-check"></i><b>4</b> Plotting data</a></li>
<li class="chapter" data-level="5" data-path="plotting-with-ggplot2.html"><a href="plotting-with-ggplot2.html"><i class="fa fa-check"></i><b>5</b> Plotting with <code>ggplot2</code></a></li>
<li class="chapter" data-level="6" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html"><i class="fa fa-check"></i><b>6</b> Using R and Bioconductor for MS-based proteomics</a></li>
<li class="chapter" data-level="7" data-path="tools-and-plots.html"><a href="tools-and-plots.html"><i class="fa fa-check"></i><b>7</b> Tools and plots</a></li>
<li class="chapter" data-level="8" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html"><i class="fa fa-check"></i><b>8</b> Interactive visualisation</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Visualisation of biomolecular data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="example-datasets" class="section level1">
<h1><span class="header-section-number">Chapter 3</span> Example datasets</h1>
<div id="crc-training-data" class="section level2">
<h2><span class="header-section-number">3.1</span> CRC training data</h2>
<p>This dataset comes from the <em><a href="http://bioconductor.org/packages/MSstatsBioData">MSstatsBioData</a></em> package and was generated as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;MSstats&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;MSstatsBioData&quot;</span>)
<span class="kw">data</span>(SRM_crc_training)
Quant &lt;-<span class="st"> </span><span class="kw">dataProcess</span>(SRM_crc_training)
subjectQuant &lt;-<span class="st"> </span><span class="kw">quantification</span>(Quant)</code></pre></div>
<p>which provides quantitative information for 72 proteins, including two standard proteins, <em>AIAG-Bovine</em> and <em>FETUA-Bovine</em>. These proteins were targeted for plasma samples with SRM with isotope labeled reference peptides in order to identify candidate protein biomarker for non-invasive detection of CRC. The training cohort included 100 subjects in control group and 100 subjects with CRC. Each sample for subject was measured in a single injection without technical replicate. The training cohort was analyzed with Skyline. The dataset was already normalized as described in manuscript. User do not need extra normalization. NAs should be considered as censored missing. Two standard proteins can be removed for statistical analysis.</p>
<p>Clinical information where added manually thereafter.</p>
<p>To load this dataset:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">crcdf &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;http://bit.ly/VisBiomedDataCrcCsv&quot;</span>)
crcdf[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</code></pre></div>
<pre><code>##       A1AG2      AFM     AHSG
## 1  14.23816 16.10302 19.95179
## 2  15.02411 16.02071 19.71592
## 3  15.63136 16.14380 19.71085
## 4  15.40137 16.27642 19.70438
## 5  16.00316 16.95821 20.42033
## 6  13.93242 16.52772 19.88985
## 7  14.34155 16.43029 20.10060
## 8  13.57086 17.11052 19.93833
## 9  15.83348 15.88189 18.64270
## 10 15.37996 16.22621 20.09992</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">crcdf[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">73</span><span class="op">:</span><span class="dv">79</span>]</code></pre></div>
<pre><code>##    Sample Group Age Gender Cancer_stage Tumour_location Sub_group
## 1   P1A10   CRC  60 female            1           colon       CRC
## 2    P1A2   CRC  70   male            1          rectum       CRC
## 3    P1A4   CRC  65   male            1          rectum       CRC
## 4    P1A6   CRC  65 female            4           colon       CRC
## 5   P1B12   CRC  62 female            3           colon       CRC
## 6    P1B2   CRC  55   male            2           colon       CRC
## 7    P1B3   CRC  61   male            2          rectum       CRC
## 8    P1B6   CRC  52   male            4          rectum       CRC
## 9    P1B9   CRC  89 female            4           colon       CRC
## 10  P1C11   CRC  81   male            1          rectum       CRC</code></pre>
<p>Conversion to an <code>MSnSet</code></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;MSnbase&quot;</span>)
i &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="dv">72</span>
e &lt;-<span class="st"> </span><span class="kw">t</span>(crcdf[, i])
<span class="kw">colnames</span>(e) &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="dv">200</span>
crc &lt;-<span class="st"> </span><span class="kw">readMSnSet2</span>(<span class="kw">data.frame</span>(e), <span class="dt">e =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">200</span>)

pd &lt;-<span class="st"> </span>crcdf[, <span class="op">-</span>i]
<span class="kw">rownames</span>(pd) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;X&quot;</span>, <span class="kw">rownames</span>(pd))
<span class="kw">pData</span>(crc) &lt;-<span class="st"> </span>pd
crc</code></pre></div>
<pre><code>## MSnSet (storageMode: lockedEnvironment)
## assayData: 72 features, 200 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: X1 X2 ... X200 (200 total)
##   varLabels: Sample Group ... Sub_group (7 total)
##   varMetadata: labelDescription
## featureData: none
## experimentData: use &#39;experimentData(object)&#39;
## Annotation:  
## - - - Processing information - - -
##  MSnbase version: 2.5.14</code></pre>
<p>Or load it direcly:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">download.file</span>(<span class="st">&quot;http://bit.ly/VisBiomedDataCrcMSnSet&quot;</span>, <span class="st">&quot;./data/crc.rda&quot;</span>)
<span class="kw">load</span>(<span class="st">&quot;./data/crc.rda&quot;</span>)
crc</code></pre></div>
<pre><code>## MSnSet (storageMode: lockedEnvironment)
## assayData: 72 features, 200 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: X1 X2 ... X200 (200 total)
##   varLabels: Sample Group ... Sub_group (7 total)
##   varMetadata: labelDescription
## featureData: none
## experimentData: use &#39;experimentData(object)&#39;
## Annotation:  
## - - - Processing information - - -
##  MSnbase version: 2.5.14</code></pre>
<p>Reference:</p>
<p>See Surinova, S. et al. (2015) <em>Prediction of colorectal cancer diagnosis based on circulating plasma proteins</em>. EMBO Mol. Med., 7, 1166â€“1178 for details.</p>
</div>
<div id="all-data" class="section level2">
<h2><span class="header-section-number">3.2</span> ALL data</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;ALL&quot;</span>)
<span class="kw">data</span>(ALL)
ALL</code></pre></div>
<pre><code>## ExpressionSet (storageMode: lockedEnvironment)
## assayData: 12625 features, 128 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: 01005 01010 ... LAL4 (128 total)
##   varLabels: cod diagnosis ... date last seen (21 total)
##   varMetadata: labelDescription
## featureData: none
## experimentData: use &#39;experimentData(object)&#39;
##   pubMedIds: 14684422 16243790 
## Annotation: hgu95av2</code></pre>
<p>From the documentation page:</p>
<blockquote>
<p>The <em>Acute Lymphoblastic Leukemia Data from the Ritz Laboratory</em> consist of microarrays from 128 different individuals with acute lymphoblastic leukemia (ALL). A number of additional covariates are available. The data have been normalized (using rma) and it is the jointly normalized data that are available here.</p>
</blockquote>
<p>Reference:</p>
<p>Sabina Chiaretti, Xiaochun Li, Robert Gentleman, Antonella Vitale, Marco Vignetti, Franco Mandelli, Jerome Ritz, and Robin Foa <em>Gene expression profile of adult T-cell acute lymphocytic leukemia identifies distinct subsets of patients with different response to therapy and survival</em>. Blood, 1 April 2004, Vol. 103, No. 7.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="plotting-data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
