<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 2 Example datasets | Visualisation of biomolecular data</title>
  <meta name="description" content="Chapter 2 Example datasets | Visualisation of biomolecular data">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 2 Example datasets | Visualisation of biomolecular data" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Chapter 2 Example datasets | Visualisation of biomolecular data" />
  <meta name="github-repo" content="lgatto/VisualisingBiomolecularData" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Example datasets | Visualisation of biomolecular data" />
  
  <meta name="twitter:description" content="Chapter 2 Example datasets | Visualisation of biomolecular data" />
  

<meta name="author" content="Laurent Gatto">



  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="index.html">
<link rel="next" href="exploring-and-visualising-biomolecular-data.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script src="libs/htmlwidgets-1.3/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.9.0/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.0.0/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.0.0/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-1.46.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-1.46.1/plotly-latest.min.js"></script>


<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i><b>1.1</b> Installation</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#questions"><i class="fa fa-check"></i><b>1.2</b> Questions</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.3</b> License</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#useful-references"><i class="fa fa-check"></i><b>1.4</b> Useful references</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="example-datasets.html"><a href="example-datasets.html"><i class="fa fa-check"></i><b>2</b> Example datasets</a><ul>
<li class="chapter" data-level="2.1" data-path="example-datasets.html"><a href="example-datasets.html#raw-ms-data"><i class="fa fa-check"></i><b>2.1</b> Raw MS data</a></li>
<li class="chapter" data-level="2.2" data-path="example-datasets.html"><a href="example-datasets.html#the-iprg-data"><i class="fa fa-check"></i><b>2.2</b> The iPRG data</a></li>
<li class="chapter" data-level="2.3" data-path="example-datasets.html"><a href="example-datasets.html#crc-training-data"><i class="fa fa-check"></i><b>2.3</b> CRC training data</a></li>
<li class="chapter" data-level="2.4" data-path="example-datasets.html"><a href="example-datasets.html#time-course-from-mulvey-et-al.-2015"><i class="fa fa-check"></i><b>2.4</b> Time course from Mulvey <em>et al.</em> 2015</a></li>
<li class="chapter" data-level="2.5" data-path="example-datasets.html"><a href="example-datasets.html#all-data"><i class="fa fa-check"></i><b>2.5</b> ALL data</a></li>
<li class="chapter" data-level="2.6" data-path="example-datasets.html"><a href="example-datasets.html#spatial-proteomics-data"><i class="fa fa-check"></i><b>2.6</b> Spatial proteomics data</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html"><i class="fa fa-check"></i><b>3</b> Exploring and visualising biomolecular data</a><ul>
<li class="chapter" data-level="" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html#raw-data"><i class="fa fa-check"></i>Raw data</a></li>
<li class="chapter" data-level="" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html#tracking-an-ion-from-the-chromatogram-to-the-ms2-spectra"><i class="fa fa-check"></i>Tracking an ion from the chromatogram to the MS2 spectra</a></li>
<li class="chapter" data-level="" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html#raw-data-qc"><i class="fa fa-check"></i>Raw data QC</a></li>
<li class="chapter" data-level="" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html#exploring-the-experimental-design"><i class="fa fa-check"></i>Exploring the experimental design</a></li>
<li class="chapter" data-level="" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html#quantitative-data"><i class="fa fa-check"></i>Quantitative data</a></li>
<li class="chapter" data-level="" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html#differential-expression"><i class="fa fa-check"></i>Differential expression</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="using-dplyr-for-data-manipulation.html"><a href="using-dplyr-for-data-manipulation.html"><i class="fa fa-check"></i><b>4</b> Using <code>dplyr</code> for data manipulation</a><ul>
<li class="chapter" data-level="4.1" data-path="using-dplyr-for-data-manipulation.html"><a href="using-dplyr-for-data-manipulation.html#selecting-columns-and-filtering-rows"><i class="fa fa-check"></i><b>4.1</b> Selecting columns and filtering rows</a></li>
<li class="chapter" data-level="4.2" data-path="using-dplyr-for-data-manipulation.html"><a href="using-dplyr-for-data-manipulation.html#pipes"><i class="fa fa-check"></i><b>4.2</b> Pipes</a></li>
<li class="chapter" data-level="4.3" data-path="using-dplyr-for-data-manipulation.html"><a href="using-dplyr-for-data-manipulation.html#mutate"><i class="fa fa-check"></i><b>4.3</b> Mutate</a></li>
<li class="chapter" data-level="4.4" data-path="using-dplyr-for-data-manipulation.html"><a href="using-dplyr-for-data-manipulation.html#split-apply-combine-data-analysis-and-the-summarize-function"><i class="fa fa-check"></i><b>4.4</b> Split-apply-combine data analysis and the summarize() function</a><ul>
<li><a href="using-dplyr-for-data-manipulation.html#the-summarize-function">The <code>summarize()</code> function</a></li>
<li class="chapter" data-level="" data-path="using-dplyr-for-data-manipulation.html"><a href="using-dplyr-for-data-manipulation.html#tallying"><i class="fa fa-check"></i>Tallying</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="plotting-in-r.html"><a href="plotting-in-r.html"><i class="fa fa-check"></i><b>5</b> Plotting in R</a><ul>
<li class="chapter" data-level="5.1" data-path="plotting-in-r.html"><a href="plotting-in-r.html#base-graphics"><i class="fa fa-check"></i><b>5.1</b> Base graphics</a></li>
<li class="chapter" data-level="5.2" data-path="plotting-in-r.html"><a href="plotting-in-r.html#plotting-with-ggplot2"><i class="fa fa-check"></i><b>5.2</b> Plotting with <code>ggplot2</code></a></li>
<li class="chapter" data-level="5.3" data-path="plotting-in-r.html"><a href="plotting-in-r.html#composing-plots"><i class="fa fa-check"></i><b>5.3</b> Composing plots</a></li>
<li class="chapter" data-level="5.4" data-path="plotting-in-r.html"><a href="plotting-in-r.html#colour-scales"><i class="fa fa-check"></i><b>5.4</b> Colour scales</a></li>
<li class="chapter" data-level="5.5" data-path="plotting-in-r.html"><a href="plotting-in-r.html#customising-plots"><i class="fa fa-check"></i><b>5.5</b> Customising plots</a></li>
<li class="chapter" data-level="5.6" data-path="plotting-in-r.html"><a href="plotting-in-r.html#combine-ggplot-figures"><i class="fa fa-check"></i><b>5.6</b> Combine <code>ggplot</code> figures</a></li>
<li class="chapter" data-level="5.7" data-path="plotting-in-r.html"><a href="plotting-in-r.html#saving-your-figures"><i class="fa fa-check"></i><b>5.7</b> Saving your figures</a></li>
<li class="chapter" data-level="5.8" data-path="plotting-in-r.html"><a href="plotting-in-r.html#exercises"><i class="fa fa-check"></i><b>5.8</b> Exercises</a></li>
<li class="chapter" data-level="5.9" data-path="plotting-in-r.html"><a href="plotting-in-r.html#references"><i class="fa fa-check"></i><b>5.9</b> References</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="tools-and-plots.html"><a href="tools-and-plots.html"><i class="fa fa-check"></i><b>6</b> Tools and plots</a><ul>
<li class="chapter" data-level="6.1" data-path="tools-and-plots.html"><a href="tools-and-plots.html#transformations"><i class="fa fa-check"></i><b>6.1</b> Transformations</a><ul>
<li class="chapter" data-level="6.1.1" data-path="tools-and-plots.html"><a href="tools-and-plots.html#expression-data"><i class="fa fa-check"></i><b>6.1.1</b> Expression data</a></li>
<li class="chapter" data-level="6.1.2" data-path="tools-and-plots.html"><a href="tools-and-plots.html#fold-changes"><i class="fa fa-check"></i><b>6.1.2</b> Fold changes</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="tools-and-plots.html"><a href="tools-and-plots.html#comparing-samples-and-linear-models"><i class="fa fa-check"></i><b>6.2</b> Comparing samples and linear models</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#challenge"><i class="fa fa-check"></i>Challenge</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="tools-and-plots.html"><a href="tools-and-plots.html#plots-for-statistical-analyses"><i class="fa fa-check"></i><b>6.3</b> Plots for statistical analyses</a><ul>
<li class="chapter" data-level="6.3.1" data-path="tools-and-plots.html"><a href="tools-and-plots.html#p-value-distributions"><i class="fa fa-check"></i><b>6.3.1</b> P-value distributions</a></li>
<li class="chapter" data-level="6.3.2" data-path="tools-and-plots.html"><a href="tools-and-plots.html#volcano-plot"><i class="fa fa-check"></i><b>6.3.2</b> Volcano plot</a></li>
<li class="chapter" data-level="6.3.3" data-path="tools-and-plots.html"><a href="tools-and-plots.html#treating-the-plot-as-a-statistic"><i class="fa fa-check"></i><b>6.3.3</b> Treating the plot as a statistic</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="tools-and-plots.html"><a href="tools-and-plots.html#visualising-intersections"><i class="fa fa-check"></i><b>6.4</b> Visualising intersections</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#challenge-1"><i class="fa fa-check"></i>Challenge</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="tools-and-plots.html"><a href="tools-and-plots.html#unsupervised-learning"><i class="fa fa-check"></i><b>6.5</b> Unsupervised learning</a></li>
<li class="chapter" data-level="6.6" data-path="tools-and-plots.html"><a href="tools-and-plots.html#hierarchical-clustering"><i class="fa fa-check"></i><b>6.6</b> Hierarchical clustering</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#how-does-hierarchical-clustering-work"><i class="fa fa-check"></i>How does hierarchical clustering work</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#defining-clusters"><i class="fa fa-check"></i>Defining clusters</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#effect-of-distance-and-clustering-algorithm"><i class="fa fa-check"></i>Effect of distance and clustering algorithm</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#challenge-2"><i class="fa fa-check"></i>Challenge</a></li>
</ul></li>
<li class="chapter" data-level="6.7" data-path="tools-and-plots.html"><a href="tools-and-plots.html#k-means-clustering"><i class="fa fa-check"></i><b>6.7</b> k-means clustering</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#how-does-k-means-work"><i class="fa fa-check"></i>How does k-means work</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#model-selection"><i class="fa fa-check"></i>Model selection</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#how-to-determine-the-number-of-clusters"><i class="fa fa-check"></i>How to determine the number of clusters</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#challenge-3"><i class="fa fa-check"></i>Challenge</a></li>
</ul></li>
<li class="chapter" data-level="6.8" data-path="tools-and-plots.html"><a href="tools-and-plots.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>6.8</b> Principal component analysis (PCA)</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#how-does-it-work"><i class="fa fa-check"></i>How does it work</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#visualisation"><i class="fa fa-check"></i>Visualisation</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#data-pre-processing"><i class="fa fa-check"></i>Data pre-processing</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#final-comments-on-pca"><i class="fa fa-check"></i>Final comments on PCA</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#challenge-4"><i class="fa fa-check"></i>Challenge</a></li>
</ul></li>
<li class="chapter" data-level="6.9" data-path="tools-and-plots.html"><a href="tools-and-plots.html#t-distributed-stochastic-neighbour-embedding"><i class="fa fa-check"></i><b>6.9</b> t-Distributed Stochastic Neighbour Embedding</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#parameter-tuning"><i class="fa fa-check"></i>Parameter tuning</a></li>
</ul></li>
<li class="chapter" data-level="6.10" data-path="tools-and-plots.html"><a href="tools-and-plots.html#the-curse-of-dimensionality"><i class="fa fa-check"></i><b>6.10</b> The curse of dimensionality</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#as-the-number-of-noisy-proteins-increases-signal-may-be-lost"><i class="fa fa-check"></i>As the number of (noisy) proteins increases, signal may be lost</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#spurious-correlations-may-arise"><i class="fa fa-check"></i>Spurious correlations may arise</a></li>
</ul></li>
<li class="chapter" data-level="6.11" data-path="tools-and-plots.html"><a href="tools-and-plots.html#pca-loadings-vs-differential-expression"><i class="fa fa-check"></i><b>6.11</b> PCA loadings vs differential expression</a></li>
<li class="chapter" data-level="6.12" data-path="tools-and-plots.html"><a href="tools-and-plots.html#heatmaps"><i class="fa fa-check"></i><b>6.12</b> Heatmaps</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html"><i class="fa fa-check"></i><b>7</b> Using R and Bioconductor for MS-based proteomics</a><ul>
<li class="chapter" data-level="7.1" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#introduction-1"><i class="fa fa-check"></i><b>7.1</b> Introduction</a></li>
<li class="chapter" data-level="7.2" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#reading-ms-data-into-r"><i class="fa fa-check"></i><b>7.2</b> Reading MS data into R</a><ul>
<li class="chapter" data-level="" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#raw-data-1"><i class="fa fa-check"></i>Raw data</a></li>
<li class="chapter" data-level="" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#identification-data"><i class="fa fa-check"></i>Identification data</a></li>
<li class="chapter" data-level="" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#quantitative-data-1"><i class="fa fa-check"></i>Quantitative data</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#manipulating-and-visualising-raw-data"><i class="fa fa-check"></i><b>7.3</b> Manipulating and visualising raw data</a></li>
<li class="chapter" data-level="7.4" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#ms-maps"><i class="fa fa-check"></i><b>7.4</b> MS maps</a></li>
<li class="chapter" data-level="7.5" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#extracted-ion-chromatograms"><i class="fa fa-check"></i><b>7.5</b> Extracted ion chromatograms</a></li>
<li class="chapter" data-level="7.6" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#peptide-spetrum-matches"><i class="fa fa-check"></i><b>7.6</b> Peptide spetrum matches</a></li>
<li class="chapter" data-level="7.7" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#ms-imaging"><i class="fa fa-check"></i><b>7.7</b> MS imaging</a></li>
<li class="chapter" data-level="7.8" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#missing-values"><i class="fa fa-check"></i><b>7.8</b> Missing values</a></li>
<li class="chapter" data-level="7.9" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#pca-plots"><i class="fa fa-check"></i><b>7.9</b> PCA plots</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html"><i class="fa fa-check"></i><b>8</b> Interactive visualisation</a><ul>
<li class="chapter" data-level="8.1" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#interactive-apps-with-plotly"><i class="fa fa-check"></i><b>8.1</b> Interactive apps with <code>plotly</code></a></li>
<li class="chapter" data-level="8.2" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#interactive-apps"><i class="fa fa-check"></i><b>8.2</b> Interactive apps</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#motivating-examples"><i class="fa fa-check"></i>Motivating examples</a><ul>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#longitudinal-proteomics-qc"><i class="fa fa-check"></i>Longitudinal proteomics QC</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#spatial-proteomics"><i class="fa fa-check"></i>Spatial proteomics</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#introduction-2"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#creation-of-our-fist-shiny-app"><i class="fa fa-check"></i>Creation of our fist shiny app</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#single-file-app"><i class="fa fa-check"></i>Single-file app</a></li>
<li><a href="interactive-visualisation.html#theres-more-to-shiny">There’s more to <code>shiny</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="session-information.html"><a href="session-information.html"><i class="fa fa-check"></i><b>9</b> Session information</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Visualisation of biomolecular data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="example-datasets" class="section level1">
<h1><span class="header-section-number">Chapter 2</span> Example datasets</h1>
<p>We will used various datasets throughout the course. These data are briefly described below, and we will explore them through various visualisations later.</p>
<div id="raw-ms-data" class="section level2">
<h2><span class="header-section-number">2.1</span> Raw MS data</h2>
<p>Section <em>Using R and Bioconductor for MS-based proteomics</em> shows how to visualise raw mass spectrometry data. The raw data that will be used come from the <em><a href="https://bioconductor.org/packages/3.9/msdata">msdata</a></em> and <em><a href="https://bioconductor.org/packages/3.9/MSnbase">MSnbase</a></em> packages.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;msdata&quot;</span>)</code></pre></div>
</div>
<div id="the-iprg-data" class="section level2">
<h2><span class="header-section-number">2.2</span> The iPRG data</h2>
<p>This iPRG data is a spiked-in exeriment, where 6 proteins where spiked at different ratios in a Yeast proteome background. Each run was repeated in triplicates and order was randomised. Participants in the study were asked to identify the differentially abundant spiked-in proteins.</p>
<div class="figure">
<img src="img/iprg.png" />

</div>
<blockquote>
<p>Choi M, Eren-Dogu ZF, Colangelo C, Cottrell J, Hoopmann MR, Kapp EA, Kim S, Lam H, Neubert TA, Palmblad M, Phinney BS, Weintraub ST, MacLean B, Vitek O. <em>ABRF Proteome Informatics Research Group (iPRG) 2015 Study: Detection of Differentially Abundant Proteins in Label-Free Quantitative LC-MS/MS Experiments</em>. J Proteome Res. 2017 Feb 3;16(2):945-957. doi: <a href="https://doi.org/10.1021/acs.jproteome.6b00881">10.1021/acs.jproteome.6b00881</a> Epub 2017 Jan 3. PMID: 27990823.</p>
</blockquote>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">iprg &lt;-<span class="st"> </span>readr<span class="op">::</span><span class="kw">read_csv</span>(<span class="st">&quot;http://bit.ly/VisBiomedDataIprgCsv&quot;</span>)</code></pre></div>
<pre><code>## Parsed with column specification:
## cols(
##   Protein = col_character(),
##   Log2Intensity = col_double(),
##   Run = col_character(),
##   Condition = col_character(),
##   BioReplicate = col_double(),
##   Intensity = col_double(),
##   TechReplicate = col_character()
## )</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">iprg</code></pre></div>
<pre><code>## # A tibble: 36,321 x 7
##    Protein Log2Intensity Run   Condition BioReplicate Intensity
##    &lt;chr&gt;           &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;            &lt;dbl&gt;     &lt;dbl&gt;
##  1 sp|D6V…          26.8 JD_0… Conditio…            1    1.18e8
##  2 sp|D6V…          26.6 JD_0… Conditio…            1    1.02e8
##  3 sp|D6V…          26.6 JD_0… Conditio…            1    1.01e8
##  4 sp|D6V…          26.8 JD_0… Conditio…            2    1.20e8
##  5 sp|D6V…          26.8 JD_0… Conditio…            2    1.16e8
##  6 sp|D6V…          26.6 JD_0… Conditio…            2    1.02e8
##  7 sp|D6V…          26.6 JD_0… Conditio…            3    1.04e8
##  8 sp|D6V…          26.5 JD_0… Conditio…            3    9.47e7
##  9 sp|D6V…          26.5 JD_0… Conditio…            3    9.69e7
## 10 sp|D6V…          26.6 JD_0… Conditio…            4    1.02e8
## # … with 36,311 more rows, and 1 more variable: TechReplicate &lt;chr&gt;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(iprg<span class="op">$</span>Condition, iprg<span class="op">$</span>TechReplicate)</code></pre></div>
<pre><code>##             
##                 A    B    C
##   Condition1 3026 3026 3027
##   Condition2 3027 3027 3027
##   Condition3 3027 3027 3027
##   Condition4 3027 3026 3027</code></pre>
<p>This data is in the so-called long format. In some applications, it is more convenient to have the data in wide format, where rows contain the protein expression data for all samples.</p>
<p>Let’s start by simplifying the data to keep only the relevant columns:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">(iprg2 &lt;-<span class="st"> </span>iprg[, <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">6</span>)])</code></pre></div>
<pre><code>## # A tibble: 36,321 x 3
##    Protein              Run                        Intensity
##    &lt;chr&gt;                &lt;chr&gt;                          &lt;dbl&gt;
##  1 sp|D6VTK4|STE2_YEAST JD_06232014_sample1_B.raw 117845016.
##  2 sp|D6VTK4|STE2_YEAST JD_06232014_sample1_C.raw 102273602.
##  3 sp|D6VTK4|STE2_YEAST JD_06232014_sample1-A.raw 100526837.
##  4 sp|D6VTK4|STE2_YEAST JD_06232014_sample2_A.raw 119765106.
##  5 sp|D6VTK4|STE2_YEAST JD_06232014_sample2_B.raw 116382798.
##  6 sp|D6VTK4|STE2_YEAST JD_06232014_sample2_C.raw 102328260.
##  7 sp|D6VTK4|STE2_YEAST JD_06232014_sample3_A.raw 103830944.
##  8 sp|D6VTK4|STE2_YEAST JD_06232014_sample3_B.raw  94660680 
##  9 sp|D6VTK4|STE2_YEAST JD_06232014_sample3_C.raw  96919972.
## 10 sp|D6VTK4|STE2_YEAST JD_06232014_sample4_B.raw 102150172.
## # … with 36,311 more rows</code></pre>
<p>We can convert the <code>iPRG</code> into a wide format with <code>tidyr::spread</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;tidyr&quot;</span>)
(iprg3 &lt;-<span class="st"> </span><span class="kw">spread</span>(iprg2, <span class="dt">key =</span> Run, <span class="dt">value =</span> Intensity))</code></pre></div>
<pre><code>## # A tibble: 3,027 x 13
##    Protein JD_06232014_sam… JD_06232014_sam… `JD_06232014_sa…
##    &lt;chr&gt;              &lt;dbl&gt;            &lt;dbl&gt;            &lt;dbl&gt;
##  1 sp|D6V…       117845016.       102273602.       100526837.
##  2 sp|O13…        27618234.        26774670.        27598550.
##  3 sp|O13…        10892143.        16948335.        11625198.
##  4 sp|O13…       192490784.       175282010.        20606703.
##  5 sp|O13…       156581624.       117211277.       145493943.
##  6 sp|O13…        71664672.        82193735.        75530595.
##  7 sp|O13…         9087607.         9324409.        10479297.
##  8 sp|O14…       788997917.      1012548158        670287886.
##  9 sp|O14…       331891060.       405043875.       383229875.
## 10 sp|O14…         6989247.         4688128.         4769092.
## # … with 3,017 more rows, and 9 more variables:
## #   JD_06232014_sample2_A.raw &lt;dbl&gt;, JD_06232014_sample2_B.raw &lt;dbl&gt;,
## #   JD_06232014_sample2_C.raw &lt;dbl&gt;, JD_06232014_sample3_A.raw &lt;dbl&gt;,
## #   JD_06232014_sample3_B.raw &lt;dbl&gt;, JD_06232014_sample3_C.raw &lt;dbl&gt;,
## #   JD_06232014_sample4_B.raw &lt;dbl&gt;, JD_06232014_sample4_C.raw &lt;dbl&gt;,
## #   `JD_06232014_sample4-A.raw` &lt;dbl&gt;</code></pre>
<div class="figure">
<img src="img/spread.png" alt="Spreading data from long to wide format" />
<p class="caption">Spreading data from long to wide format</p>
</div>
<p>Indeed, we started with</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(<span class="kw">unique</span>(iprg<span class="op">$</span>Protein))</code></pre></div>
<pre><code>## [1] 3027</code></pre>
<p>unique proteins, which corresponds to the number of rows in the new wide dataset.</p>
<p>The long format is ideal when using <code>ggplot2</code>, as we will see in a later chapter. The wide format has also advantages. For example, it becomes straighforward to verify if there are proteins that haven’t been quantified in some samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">(k &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">is.na</span>(iprg3), <span class="dt">arr.ind =</span> <span class="kw">dim</span>(iprg3)))</code></pre></div>
<pre><code>##       row col
## [1,] 2721   2
## [2,] 2721   4
## [3,]  652  11</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">iprg3[<span class="kw">unique</span>(k[, <span class="st">&quot;row&quot;</span>]), ]</code></pre></div>
<pre><code>## # A tibble: 2 x 13
##   Protein JD_06232014_sam… JD_06232014_sam… `JD_06232014_sa…
##   &lt;chr&gt;              &lt;dbl&gt;            &lt;dbl&gt;            &lt;dbl&gt;
## 1 sp|Q08…              NA         10989286.              NA 
## 2 sp|P28…          961274.         1229085.         1108670.
## # … with 9 more variables: JD_06232014_sample2_A.raw &lt;dbl&gt;,
## #   JD_06232014_sample2_B.raw &lt;dbl&gt;, JD_06232014_sample2_C.raw &lt;dbl&gt;,
## #   JD_06232014_sample3_A.raw &lt;dbl&gt;, JD_06232014_sample3_B.raw &lt;dbl&gt;,
## #   JD_06232014_sample3_C.raw &lt;dbl&gt;, JD_06232014_sample4_B.raw &lt;dbl&gt;,
## #   JD_06232014_sample4_C.raw &lt;dbl&gt;, `JD_06232014_sample4-A.raw` &lt;dbl&gt;</code></pre>
<p>The opposite operation to <code>spread</code> is <code>gather</code>, also from the <code>tidyr</code> package:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">(iprg4 &lt;-<span class="st"> </span><span class="kw">gather</span>(iprg3, <span class="dt">key =</span> Run, <span class="dt">value =</span> Intensity, <span class="op">-</span>Protein))</code></pre></div>
<pre><code>## # A tibble: 36,324 x 3
##    Protein               Run                        Intensity
##    &lt;chr&gt;                 &lt;chr&gt;                          &lt;dbl&gt;
##  1 sp|D6VTK4|STE2_YEAST  JD_06232014_sample1_B.raw 117845016.
##  2 sp|O13297|CET1_YEAST  JD_06232014_sample1_B.raw  27618234.
##  3 sp|O13329|FOB1_YEAST  JD_06232014_sample1_B.raw  10892143.
##  4 sp|O13539|THP2_YEAST  JD_06232014_sample1_B.raw 192490784.
##  5 sp|O13547|CCW14_YEAST JD_06232014_sample1_B.raw 156581624.
##  6 sp|O13563|RPN13_YEAST JD_06232014_sample1_B.raw  71664672.
##  7 sp|O13585|YP089_YEAST JD_06232014_sample1_B.raw   9087607.
##  8 sp|O14455|RL36B_YEAST JD_06232014_sample1_B.raw 788997917.
##  9 sp|O14467|MBF1_YEAST  JD_06232014_sample1_B.raw 331891060.
## 10 sp|O14468|YO304_YEAST JD_06232014_sample1_B.raw   6989247.
## # … with 36,314 more rows</code></pre>
<div class="figure">
<img src="img/gather.png" alt="Gathering data from wide to long format" />
<p class="caption">Gathering data from wide to long format</p>
</div>
<p>The two lond datasets, <code>iprg2</code> and <code>iprg4</code> are different due to the missing values shown above.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">nrow</span>(iprg2)</code></pre></div>
<pre><code>## [1] 36321</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">nrow</span>(iprg4)</code></pre></div>
<pre><code>## [1] 36324</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">nrow</span>(<span class="kw">na.omit</span>(iprg4))</code></pre></div>
<pre><code>## [1] 36321</code></pre>
<p>which can be accounted for by removing rows with missing values by setting <code>na.rm = TRUE</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">(iprg5 &lt;-<span class="st"> </span><span class="kw">gather</span>(iprg3, <span class="dt">key =</span> Run, <span class="dt">value =</span> Intensity, <span class="op">-</span>Protein, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>))</code></pre></div>
<pre><code>## # A tibble: 36,321 x 3
##    Protein               Run                        Intensity
##    &lt;chr&gt;                 &lt;chr&gt;                          &lt;dbl&gt;
##  1 sp|D6VTK4|STE2_YEAST  JD_06232014_sample1_B.raw 117845016.
##  2 sp|O13297|CET1_YEAST  JD_06232014_sample1_B.raw  27618234.
##  3 sp|O13329|FOB1_YEAST  JD_06232014_sample1_B.raw  10892143.
##  4 sp|O13539|THP2_YEAST  JD_06232014_sample1_B.raw 192490784.
##  5 sp|O13547|CCW14_YEAST JD_06232014_sample1_B.raw 156581624.
##  6 sp|O13563|RPN13_YEAST JD_06232014_sample1_B.raw  71664672.
##  7 sp|O13585|YP089_YEAST JD_06232014_sample1_B.raw   9087607.
##  8 sp|O14455|RL36B_YEAST JD_06232014_sample1_B.raw 788997917.
##  9 sp|O14467|MBF1_YEAST  JD_06232014_sample1_B.raw 331891060.
## 10 sp|O14468|YO304_YEAST JD_06232014_sample1_B.raw   6989247.
## # … with 36,311 more rows</code></pre>
</div>
<div id="crc-training-data" class="section level2">
<h2><span class="header-section-number">2.3</span> CRC training data</h2>
<p>This dataset comes from the <em><a href="https://bioconductor.org/packages/3.9/MSstatsBioData">MSstatsBioData</a></em> package and was generated as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;MSstats&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;MSstatsBioData&quot;</span>)
<span class="kw">data</span>(SRM_crc_training)
Quant &lt;-<span class="st"> </span><span class="kw">dataProcess</span>(SRM_crc_training)
subjectQuant &lt;-<span class="st"> </span><span class="kw">quantification</span>(Quant)</code></pre></div>
<p>It provides quantitative information for 72 proteins, including two standard proteins, <em>AIAG-Bovine</em> and <em>FETUA-Bovine</em>. These proteins were targeted for plasma samples with SRM with isotope labeled reference peptides in order to identify candidate protein biomarker for non-invasive detection of CRC. The training cohort included 100 subjects in control group and 100 subjects with CRC. Each sample for subject was measured in a single injection without technical replicate. The training cohort was analyzed with Skyline. The dataset was already normalized as described in manuscript. User do not need extra normalization. NAs should be considered as censored missing. Two standard proteins can be removed for statistical analysis.</p>
<p>Clinical information where added manually thereafter.</p>
<p>To load this dataset:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">(crcdf &lt;-<span class="st"> </span>readr<span class="op">::</span><span class="kw">read_csv</span>(<span class="st">&quot;http://bit.ly/VisBiomedDataCrcCsv&quot;</span>))</code></pre></div>
<pre><code>## Parsed with column specification:
## cols(
##   .default = col_double(),
##   Sample = col_character(),
##   Group = col_character(),
##   Gender = col_character(),
##   Tumour_location = col_character(),
##   Sub_group = col_character()
## )</code></pre>
<pre><code>## See spec(...) for full column specifications.</code></pre>
<pre><code>## # A tibble: 200 x 79
##    A1AG2   AFM  AHSG `AIAG-Bovine`  ANT3  AOC3  APOB  ATRN   BTD C20orf3
##    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;         &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;
##  1  14.2  16.1  20.0          15.3  17.2 10.0   15.5  14.4  16.3   10.7 
##  2  15.0  16.0  19.7          15.2  17.3  9.04  15.1  14.0  16.2   10.7 
##  3  15.6  16.1  19.7          15.6  17.6 10.4   16.0  14.6  16.5   11.2 
##  4  15.4  16.3  19.7          15.1  17.4  9.50  15.7  14.1  16.3    9.97
##  5  16.0  17.0  20.4          15.6  18.0  9.65  16.2  14.2  16.5   12.6 
##  6  13.9  16.5  19.9          13.4  16.3  9.52  14.5  13.7  15.8   10.6 
##  7  14.3  16.4  20.1          15.1  17.4  9.72  14.7  13.3  16.4   11.4 
##  8  13.6  17.1  19.9          14.9  17.3  8.79  14.9  12.7  16.1   10.8 
##  9  15.8  15.9  18.6          15.1  16.7  8.95  14.7  13.2  15.6   11.6 
## 10  15.4  16.2  20.1          15.4  17.6 10.3   15.9  14.7  16.5   11.4 
## # … with 190 more rows, and 69 more variables: CADM1 &lt;dbl&gt;, CD163 &lt;dbl&gt;,
## #   CD44 &lt;dbl&gt;, CDH5 &lt;dbl&gt;, CFH &lt;dbl&gt;, CFI &lt;dbl&gt;, CLU &lt;dbl&gt;, CP &lt;dbl&gt;,
## #   CTSD &lt;dbl&gt;, DKFZp686N02209 &lt;dbl&gt;, DSG2 &lt;dbl&gt;, ECM1 &lt;dbl&gt;, F11 &lt;dbl&gt;,
## #   F5 &lt;dbl&gt;, FCGBP &lt;dbl&gt;, `FETUA-Bovine` &lt;dbl&gt;, FETUB &lt;dbl&gt;, FGA &lt;dbl&gt;,
## #   FGG &lt;dbl&gt;, FHR3 &lt;dbl&gt;, FN1 &lt;dbl&gt;, GOLM1 &lt;dbl&gt;, HP &lt;dbl&gt;, HRG &lt;dbl&gt;,
## #   HYOU1 &lt;dbl&gt;, ICAM1 &lt;dbl&gt;, IGFBP3 &lt;dbl&gt;, IGHA2 &lt;dbl&gt;, IGHG2 &lt;dbl&gt;,
## #   ITIH4 &lt;dbl&gt;, KLKB1 &lt;dbl&gt;, KNG1 &lt;dbl&gt;, LAMP2 &lt;dbl&gt;, LCN2 &lt;dbl&gt;,
## #   LGALS3BP &lt;dbl&gt;, LRG1 &lt;dbl&gt;, LUM &lt;dbl&gt;, LYVE1 &lt;dbl&gt;, MMRN1 &lt;dbl&gt;,
## #   MPO &lt;dbl&gt;, MRC2 &lt;dbl&gt;, MST1 &lt;dbl&gt;, NCAM1 &lt;dbl&gt;, ORM1 &lt;dbl&gt;,
## #   PGCP &lt;dbl&gt;, PIGR &lt;dbl&gt;, PLTP &lt;dbl&gt;, PLXDC2 &lt;dbl&gt;, PON1 &lt;dbl&gt;,
## #   PRG4 &lt;dbl&gt;, PROC &lt;dbl&gt;, PTPRJ &lt;dbl&gt;, Q5JNX2 &lt;dbl&gt;, SERPINA1 &lt;dbl&gt;,
## #   SERPINA3 &lt;dbl&gt;, SERPINA6 &lt;dbl&gt;, SERPINA7 &lt;dbl&gt;, THBS1 &lt;dbl&gt;,
## #   TIMP1 &lt;dbl&gt;, TNC &lt;dbl&gt;, VTN &lt;dbl&gt;, VWF &lt;dbl&gt;, Sample &lt;chr&gt;,
## #   Group &lt;chr&gt;, Age &lt;dbl&gt;, Gender &lt;chr&gt;, Cancer_stage &lt;dbl&gt;,
## #   Tumour_location &lt;chr&gt;, Sub_group &lt;chr&gt;</code></pre>
<p>This dataset is in the wide format. It contains the intensity of the proteins in columns 1 to 72 for each of the 200 samples along the rows. Generally, omics datasets contain the features (proteins, transcripts, …) along the rows and the samples along the columns.</p>
<p>In columns 73 to 79, we sample metadata.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">crcdf[, <span class="dv">73</span><span class="op">:</span><span class="dv">79</span>]</code></pre></div>
<pre><code>## # A tibble: 200 x 7
##    Sample Group   Age Gender Cancer_stage Tumour_location Sub_group
##    &lt;chr&gt;  &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;         &lt;dbl&gt; &lt;chr&gt;           &lt;chr&gt;    
##  1 P1A10  CRC      60 female            1 colon           CRC      
##  2 P1A2   CRC      70 male              1 rectum          CRC      
##  3 P1A4   CRC      65 male              1 rectum          CRC      
##  4 P1A6   CRC      65 female            4 colon           CRC      
##  5 P1B12  CRC      62 female            3 colon           CRC      
##  6 P1B2   CRC      55 male              2 colon           CRC      
##  7 P1B3   CRC      61 male              2 rectum          CRC      
##  8 P1B6   CRC      52 male              4 rectum          CRC      
##  9 P1B9   CRC      89 female            4 colon           CRC      
## 10 P1C11  CRC      81 male              1 rectum          CRC      
## # … with 190 more rows</code></pre>
<p>A widely used data structure for omics data follows the convention described in the figure below:</p>
<div class="figure">
<img src="img/msnset.png" alt="An eSet-type of expression data container" />
<p class="caption">An eSet-type of expression data container</p>
</div>
<p>This typical omics data structure, as defined by the <code>eSet</code> class in the Bioconductor <code>Biobase</code> package, is represented below. It’s main features are</p>
<ul>
<li><p>An assay data slot containing the quantitative omics data (expression data), stored as a <code>matrix</code> and accessible with <code>exprs</code>. Features defined along the rows and samples along the columns.</p></li>
<li><p>A sample metadata slot containing sample co-variates, stored as an annotated <code>data.frame</code> and accessible with <code>pData</code>. This data frame is stored with rows representing samples and sample covariate along the columns, and its rows match the expression data columns exactly.</p></li>
<li><p>A feature metadata slot containing feature co-variates, stored as an annotated <code>data.frame</code> and accessible with <code>fData</code>. This dataframe’s rows match the expression data rows exactly.</p></li>
</ul>
<p>The coordinated nature of the high throughput data guarantees that the dimensions of the different slots will always match (i.e the columns in the expression data and then rows in the sample metadata, as well as the rows in the expression data and feature metadata) during data manipulation. The metadata slots can grow additional co-variates (columns) without affecting the other structures.</p>
<p>Below, we show how to transform the <code>crc</code> dataset into an <code>MSnSet</code> (implementing the data structure above for quantitative proteomics data) using the <code>readMSnSet2</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;MSnbase&quot;</span>)
i &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="dv">72</span> ## expression columns
e &lt;-<span class="st"> </span><span class="kw">t</span>(crcdf[, i]) ## expression data
<span class="kw">colnames</span>(e) &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="dv">200</span>
crc &lt;-<span class="st"> </span><span class="kw">readMSnSet2</span>(<span class="kw">data.frame</span>(e), <span class="dt">e =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">200</span>)

pd &lt;-<span class="st"> </span><span class="kw">data.frame</span>(crcdf[, <span class="op">-</span>i])
<span class="kw">rownames</span>(pd) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;X&quot;</span>, <span class="kw">rownames</span>(pd))
<span class="kw">pData</span>(crc) &lt;-<span class="st"> </span>pd
crc</code></pre></div>
<pre><code>## MSnSet (storageMode: lockedEnvironment)
## assayData: 72 features, 200 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: X1 X2 ... X200 (200 total)
##   varLabels: Sample Group ... Sub_group (7 total)
##   varMetadata: labelDescription
## featureData: none
## experimentData: use &#39;experimentData(object)&#39;
## Annotation:  
## - - - Processing information - - -
##  MSnbase version: 2.10.0</code></pre>
<p>Let’s also set the sample names.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sampleNames</span>(crc) &lt;-<span class="st"> </span>crc<span class="op">$</span>Sample</code></pre></div>
<p>To download and load the <code>MSnSet</code> direcly:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">download.file</span>(<span class="st">&quot;http://bit.ly/VisBiomedDataCrcMSnSet&quot;</span>, <span class="st">&quot;./data/crc.rda&quot;</span>)
<span class="kw">load</span>(<span class="st">&quot;./data/crc.rda&quot;</span>)
crc</code></pre></div>
<pre><code>## MSnSet (storageMode: lockedEnvironment)
## assayData: 72 features, 200 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: P1A10 P1A2 ... P3H6 (200 total)
##   varLabels: Sample Group ... Sub_group (7 total)
##   varMetadata: labelDescription
## featureData: none
## experimentData: use &#39;experimentData(object)&#39;
## Annotation:  
## - - - Processing information - - -
##  MSnbase version: 2.6.0</code></pre>
<p>Reference:</p>
<blockquote>
<p>See Surinova, S. et al. (2015) <em>Prediction of colorectal cancer diagnosis based on circulating plasma proteins</em>. EMBO Mol. Med., 7, 1166–1178 for details.</p>
</blockquote>
</div>
<div id="time-course-from-mulvey-et-al.-2015" class="section level2">
<h2><span class="header-section-number">2.4</span> Time course from Mulvey <em>et al.</em> 2015</h2>
<p>This data comes from</p>
<blockquote>
<p>Mulvey CM, Schröter C, Gatto L, Dikicioglu D, Fidaner IB, Christoforou A, Deery MJ, Cho LT, Niakan KK, Martinez-Arias A, Lilley KS. Dynamic Proteomic Profiling of Extra-Embryonic Endoderm Differentiation in Mouse Embryonic Stem Cells. Stem Cells. 2015 Sep;33(9):2712-25. doi: <a href="https://doi.org/10.1002/stem.2067">10.1002/stem.2067</a>. Epub 2015 Jun 23. PMID: 26059426.</p>
</blockquote>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;pRolocdata&quot;</span>)
<span class="kw">data</span>(mulvey2015norm)</code></pre></div>
<p>This <code>MSnSet</code>, available from the <em><a href="https://bioconductor.org/packages/3.9/pRolocdata">pRolocdata</a></em> package, measured the expression profiles of 2337 proteins along 6 time points in triplicate.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MSnbase<span class="op">::</span><span class="kw">exprs</span>(mulvey2015norm)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</code></pre></div>
<pre><code>##          rep1_0hr rep1_16hr rep1_24hr
## P48432   2.479592  1.698630 1.0350877
## Q62315-2 1.979592  1.342466 0.9181287
## P55821   1.780612  1.767123 1.1578947
## P17809   1.637755  1.157534 0.9941520
## Q8K3F7   1.852041  1.623288 1.3040936</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">pData</span>(mulvey2015norm)</code></pre></div>
<pre><code>##           rep times cond
## rep1_0hr    1     1    1
## rep1_16hr   1     2    1
## rep1_24hr   1     3    1
## rep1_48hr   1     4    1
## rep1_72hr   1     5    1
## rep1_XEN    1     6    1
## rep2_0hr    2     1    1
## rep2_16hr   2     2    1
## rep2_24hr   2     3    1
## rep2_48hr   2     4    1
## rep2_72hr   2     5    1
## rep2_XEN    2     6    1
## rep3_0hr    3     1    1
## rep3_16hr   3     2    1
## rep3_24hr   3     3    1
## rep3_48hr   3     4    1
## rep3_72hr   3     5    1
## rep3_XEN    3     6    1</code></pre>
</div>
<div id="all-data" class="section level2">
<h2><span class="header-section-number">2.5</span> ALL data</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;ALL&quot;</span>)
<span class="kw">data</span>(ALL)
ALL</code></pre></div>
<pre><code>## ExpressionSet (storageMode: lockedEnvironment)
## assayData: 12625 features, 128 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: 01005 01010 ... LAL4 (128 total)
##   varLabels: cod diagnosis ... date last seen (21 total)
##   varMetadata: labelDescription
## featureData: none
## experimentData: use &#39;experimentData(object)&#39;
##   pubMedIds: 14684422 16243790 
## Annotation: hgu95av2</code></pre>
<p>From the documentation page:</p>
<blockquote>
<p>The <em>Acute Lymphoblastic Leukemia Data from the Ritz Laboratory</em> consist of microarrays from 128 different individuals with acute lymphoblastic leukemia (ALL). A number of additional covariates are available. The data have been normalized (using rma) and it is the jointly normalized data that are available here.</p>
</blockquote>
<p>The <code>ALL</code> data is of class <code>ExpressionSet</code>, which implements the data structure above for microarray expression data, and contains normalised and summarised transcript intensities.</p>
<p>Below, we select will patients with B-cell lymphomas and BCR/ABL abnormality and negative controls.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(ALL<span class="op">$</span>BT)</code></pre></div>
<pre><code>## 
##  B B1 B2 B3 B4  T T1 T2 T3 T4 
##  5 19 36 23 12  5  1 15 10  2</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(ALL<span class="op">$</span>mol.biol)</code></pre></div>
<pre><code>## 
## ALL1/AF4  BCR/ABL E2A/PBX1      NEG   NUP-98  p15/p16 
##       10       37        5       74        1        1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ALL_bcrneg &lt;-<span class="st"> </span>ALL[, ALL<span class="op">$</span>mol.biol <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;NEG&quot;</span>, <span class="st">&quot;BCR/ABL&quot;</span>) <span class="op">&amp;</span><span class="st"> </span><span class="kw">grepl</span>(<span class="st">&quot;B&quot;</span>, ALL<span class="op">$</span>BT)]
ALL_bcrneg<span class="op">$</span>mol.biol &lt;-<span class="st"> </span><span class="kw">factor</span>(ALL_bcrneg<span class="op">$</span>mol.biol)</code></pre></div>
<p>We then use the <em><a href="https://bioconductor.org/packages/3.9/limma">limma</a></em> package to</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;limma&quot;</span>)
design &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="dv">0</span><span class="op">+</span>ALL_bcrneg<span class="op">$</span>mol.biol)
<span class="kw">colnames</span>(design) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;BCR.ABL&quot;</span>, <span class="st">&quot;NEG&quot;</span>)

## Step1: linear model. lmFit is a wrapper around lm in R
fit1 &lt;-<span class="st"> </span><span class="kw">lmFit</span>(ALL_bcrneg, design)

## Step 2: fit contrasts: find genes that respond to estrogen
contrast.matrix &lt;-<span class="st"> </span><span class="kw">makeContrasts</span>(BCR.ABL<span class="op">-</span>NEG, <span class="dt">levels =</span> design)
fit2 &lt;-<span class="st"> </span><span class="kw">contrasts.fit</span>(fit1, contrast.matrix)

## Step3: add empirical Bayes moderation
fit3 &lt;-<span class="st"> </span><span class="kw">eBayes</span>(fit2)

## Extract results and set them to the feature data
res &lt;-<span class="st"> </span><span class="kw">topTable</span>(fit3, <span class="dt">n =</span> <span class="ot">Inf</span>)
<span class="kw">fData</span>(ALL_bcrneg) &lt;-<span class="st"> </span>res[<span class="kw">featureNames</span>(ALL_bcrneg), ]</code></pre></div>
<p>This annotated <code>ExpressionSet</code> can be reproduced as shown above or downloaded and loaded using</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">download.file</span>(<span class="st">&quot;http://bit.ly/VisBiomedDataALL_bcrneg&quot;</span>,  <span class="st">&quot;./data/ALL_bcrneg.rda&quot;</span>)
<span class="kw">load</span>(<span class="st">&quot;./data/ALL_bcrneg.rda&quot;</span>)</code></pre></div>
<p>Reference:</p>
<blockquote>
<p>Sabina Chiaretti, Xiaochun Li, Robert Gentleman, Antonella Vitale, Marco Vignetti, Franco Mandelli, Jerome Ritz, and Robin Foa <em>Gene expression profile of adult T-cell acute lymphocytic leukemia identifies distinct subsets of patients with different response to therapy and survival</em>. Blood, 1 April 2004, Vol. 103, No. 7.</p>
</blockquote>
</div>
<div id="spatial-proteomics-data" class="section level2">
<h2><span class="header-section-number">2.6</span> Spatial proteomics data</h2>
<p>The goal of spatial proteomics is to study the sub-cellular localisation of proteins. The data below comes from</p>
<blockquote>
<p>Christoforou A, Mulvey CM, Breckels LM, Geladaki A, Hurrell T, Hayward PC, Naake T, Gatto L, Viner R, Martinez Arias A, Lilley KS. A draft map of the mouse pluripotent stem cell spatial proteome. Nat Commun. 2016 Jan 12;7:8992. doi: <a href="https://doi.org/10.1038/ncomms9992">10.1038/ncomms9992</a>. PMID: 26754106; PMCID: PMC4729960.</p>
</blockquote>
<p>and investigated the sub-cellular location of over 5000 proteins in mouse pluripotent stem cells using a mass spectrometry-based technique called hyperLOPIT.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;pRolocdata&quot;</span>)
<span class="kw">data</span>(hyperLOPIT2015)
<span class="kw">dim</span>(hyperLOPIT2015)</code></pre></div>
<pre><code>## [1] 5032   20</code></pre>
<p>In addition to the quantitative data, another important piece of data here are the spatial markers, proteins for which we can confidently assign them a sub-cellular location apriori. These markers are defined in the <code>markers</code> feature variable:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(<span class="kw">fData</span>(hyperLOPIT2015)<span class="op">$</span>markers)</code></pre></div>
<pre><code>## 
##                          40S Ribosome 
##                                    27 
##                          60S Ribosome 
##                                    43 
##                    Actin cytoskeleton 
##                                    13 
##                               Cytosol 
##                                    43 
## Endoplasmic reticulum/Golgi apparatus 
##                                   107 
##                              Endosome 
##                                    13 
##                  Extracellular matrix 
##                                    13 
##                              Lysosome 
##                                    33 
##                         Mitochondrion 
##                                   383 
##                   Nucleus - Chromatin 
##                                    64 
##               Nucleus - Non-chromatin 
##                                    85 
##                            Peroxisome 
##                                    17 
##                       Plasma membrane 
##                                    51 
##                            Proteasome 
##                                    34 
##                               unknown 
##                                  4106</code></pre>
<p>Here, the columns of the data represent fractions along a density gradient used to separate the sub-cellular proteome presented in this data.</p>
<p>We will be using a simplified version of this data in the next chapters. Below, we retain the first replicate (<code>hyperLOPIT2015$Replicate == 1]</code>) and the marker proteins (using the <code>markerMSnSet</code> function):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hlm &lt;-<span class="st"> </span>pRoloc<span class="op">::</span><span class="kw">markerMSnSet</span>(hyperLOPIT2015[, hyperLOPIT2015<span class="op">$</span>Replicate <span class="op">==</span><span class="st"> </span><span class="dv">1</span>])
<span class="kw">dim</span>(hlm)</code></pre></div>
<pre><code>## [1] 926  10</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">getMarkerClasses</span>(hlm) ## no unknowns anymore</code></pre></div>
<pre><code>##  [1] &quot;40S Ribosome&quot;                         
##  [2] &quot;60S Ribosome&quot;                         
##  [3] &quot;Actin cytoskeleton&quot;                   
##  [4] &quot;Cytosol&quot;                              
##  [5] &quot;Endoplasmic reticulum/Golgi apparatus&quot;
##  [6] &quot;Endosome&quot;                             
##  [7] &quot;Extracellular matrix&quot;                 
##  [8] &quot;Lysosome&quot;                             
##  [9] &quot;Mitochondrion&quot;                        
## [10] &quot;Nucleus - Chromatin&quot;                  
## [11] &quot;Nucleus - Non-chromatin&quot;              
## [12] &quot;Peroxisome&quot;                           
## [13] &quot;Plasma membrane&quot;                      
## [14] &quot;Proteasome&quot;</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="exploring-and-visualising-biomolecular-data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
