<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Visualisation of biomolecular data</title>
  <meta name="description" content="Visualisation of biomolecular data">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Visualisation of biomolecular data" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Visualisation of biomolecular data" />
  <meta name="github-repo" content="lgatto/VisualisingBiomolecularData" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Visualisation of biomolecular data" />
  
  <meta name="twitter:description" content="Visualisation of biomolecular data" />
  

<meta name="author" content="Laurent Gatto">


<meta name="date" content="2018-05-04">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  

<link rel="next" href="example-datasets-and-data-manipulation.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i><b>1.1</b> Installation</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#questions"><i class="fa fa-check"></i><b>1.2</b> Questions</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.3</b> License</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#useful-references"><i class="fa fa-check"></i><b>1.4</b> Useful references</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html"><i class="fa fa-check"></i><b>2</b> Example datasets and data manipulation</a><ul>
<li class="chapter" data-level="2.1" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html#raw-ms-data"><i class="fa fa-check"></i><b>2.1</b> Raw MS data</a></li>
<li class="chapter" data-level="2.2" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html#time-course-from-mulvey-et-al.-2015"><i class="fa fa-check"></i><b>2.2</b> Time course from Mulvey <em>et al.</em> 2015</a></li>
<li class="chapter" data-level="2.3" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html#the-iprg-data"><i class="fa fa-check"></i><b>2.3</b> The iPRG data</a></li>
<li class="chapter" data-level="2.4" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html#crc-training-data"><i class="fa fa-check"></i><b>2.4</b> CRC training data</a></li>
<li class="chapter" data-level="2.5" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html#all-data"><i class="fa fa-check"></i><b>2.5</b> ALL data</a></li>
<li class="chapter" data-level="2.6" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html#using-dplyr-for-data-manipulation"><i class="fa fa-check"></i><b>2.6</b> Using <code>dplyr</code> for data manipulation</a><ul>
<li class="chapter" data-level="2.6.1" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html#selecting-columns-and-filtering-rows"><i class="fa fa-check"></i><b>2.6.1</b> Selecting columns and filtering rows</a></li>
<li class="chapter" data-level="2.6.2" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html#pipes"><i class="fa fa-check"></i><b>2.6.2</b> Pipes</a></li>
<li class="chapter" data-level="2.6.3" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html#mutate"><i class="fa fa-check"></i><b>2.6.3</b> Mutate</a></li>
<li class="chapter" data-level="2.6.4" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html#split-apply-combine-data-analysis-and-the-summarize-function"><i class="fa fa-check"></i><b>2.6.4</b> Split-apply-combine data analysis and the summarize() function</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html"><i class="fa fa-check"></i><b>3</b> Exploring and visualising biomolecular data</a></li>
<li class="chapter" data-level="4" data-path="plotting-in-r.html"><a href="plotting-in-r.html"><i class="fa fa-check"></i><b>4</b> Plotting in R</a><ul>
<li class="chapter" data-level="4.1" data-path="plotting-in-r.html"><a href="plotting-in-r.html#base-graphics"><i class="fa fa-check"></i><b>4.1</b> Base graphics</a></li>
<li class="chapter" data-level="4.2" data-path="plotting-in-r.html"><a href="plotting-in-r.html#plotting-with-ggplot2"><i class="fa fa-check"></i><b>4.2</b> Plotting with <code>ggplot2</code></a></li>
<li class="chapter" data-level="4.3" data-path="plotting-in-r.html"><a href="plotting-in-r.html#building-complex-plots"><i class="fa fa-check"></i><b>4.3</b> Building complex plots</a></li>
<li class="chapter" data-level="4.4" data-path="plotting-in-r.html"><a href="plotting-in-r.html#customising-plots"><i class="fa fa-check"></i><b>4.4</b> Customising plots</a></li>
<li class="chapter" data-level="4.5" data-path="plotting-in-r.html"><a href="plotting-in-r.html#combine-ggplot-figures"><i class="fa fa-check"></i><b>4.5</b> Combine <code>ggplot</code> figures</a></li>
<li class="chapter" data-level="4.6" data-path="plotting-in-r.html"><a href="plotting-in-r.html#saving-your-figures"><i class="fa fa-check"></i><b>4.6</b> Saving your figures</a></li>
<li class="chapter" data-level="4.7" data-path="plotting-in-r.html"><a href="plotting-in-r.html#references"><i class="fa fa-check"></i><b>4.7</b> References</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="tools-and-plots.html"><a href="tools-and-plots.html"><i class="fa fa-check"></i><b>5</b> Tools and plots</a><ul>
<li class="chapter" data-level="5.1" data-path="tools-and-plots.html"><a href="tools-and-plots.html#transformations"><i class="fa fa-check"></i><b>5.1</b> Transformations</a><ul>
<li class="chapter" data-level="5.1.1" data-path="tools-and-plots.html"><a href="tools-and-plots.html#expression-data"><i class="fa fa-check"></i><b>5.1.1</b> Expression data</a></li>
<li class="chapter" data-level="5.1.2" data-path="tools-and-plots.html"><a href="tools-and-plots.html#fold-changes"><i class="fa fa-check"></i><b>5.1.2</b> Fold changes</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="tools-and-plots.html"><a href="tools-and-plots.html#comparing-samples-and-linear-models"><i class="fa fa-check"></i><b>5.2</b> Comparing samples and linear models</a></li>
<li class="chapter" data-level="5.3" data-path="tools-and-plots.html"><a href="tools-and-plots.html#plots-for-statistical-analyses"><i class="fa fa-check"></i><b>5.3</b> Plots for statistical analyses</a></li>
<li class="chapter" data-level="5.4" data-path="tools-and-plots.html"><a href="tools-and-plots.html#visualising-intersections"><i class="fa fa-check"></i><b>5.4</b> Visualising intersections</a></li>
<li class="chapter" data-level="5.5" data-path="tools-and-plots.html"><a href="tools-and-plots.html#unsupervised-learning"><i class="fa fa-check"></i><b>5.5</b> Unsupervised learning</a></li>
<li class="chapter" data-level="5.6" data-path="tools-and-plots.html"><a href="tools-and-plots.html#k-means-clustering"><i class="fa fa-check"></i><b>5.6</b> k-means clustering</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#how-does-k-means-work"><i class="fa fa-check"></i>How does k-means work</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#model-selection"><i class="fa fa-check"></i>Model selection</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#how-to-determine-the-number-of-clusters"><i class="fa fa-check"></i>How to determine the number of clusters</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#challenge"><i class="fa fa-check"></i>Challenge</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="tools-and-plots.html"><a href="tools-and-plots.html#hierarchical-clustering"><i class="fa fa-check"></i><b>5.7</b> Hierarchical clustering</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#how-does-hierarchical-clustering-work"><i class="fa fa-check"></i>How does hierarchical clustering work</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#defining-clusters"><i class="fa fa-check"></i>Defining clusters</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#challenge-1"><i class="fa fa-check"></i>Challenge</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="tools-and-plots.html"><a href="tools-and-plots.html#pre-processing"><i class="fa fa-check"></i><b>5.8</b> Pre-processing</a></li>
<li class="chapter" data-level="5.9" data-path="tools-and-plots.html"><a href="tools-and-plots.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>5.9</b> Principal component analysis (PCA)</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#how-does-it-work"><i class="fa fa-check"></i>How does it work</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#visualisation"><i class="fa fa-check"></i>Visualisation</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#data-pre-processing"><i class="fa fa-check"></i>Data pre-processing</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#final-comments-on-pca"><i class="fa fa-check"></i>Final comments on PCA</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#challenge-2"><i class="fa fa-check"></i>Challenge</a></li>
</ul></li>
<li class="chapter" data-level="5.10" data-path="tools-and-plots.html"><a href="tools-and-plots.html#t-distributed-stochastic-neighbour-embedding"><i class="fa fa-check"></i><b>5.10</b> t-Distributed Stochastic Neighbour Embedding</a><ul>
<li class="chapter" data-level="5.10.1" data-path="tools-and-plots.html"><a href="tools-and-plots.html#parameter-tuning"><i class="fa fa-check"></i><b>5.10.1</b> Parameter tuning</a></li>
</ul></li>
<li class="chapter" data-level="5.11" data-path="tools-and-plots.html"><a href="tools-and-plots.html#pca-loadings-vs-differential-expression"><i class="fa fa-check"></i><b>5.11</b> PCA loadings vs differential expression</a></li>
<li class="chapter" data-level="5.12" data-path="tools-and-plots.html"><a href="tools-and-plots.html#heatmaps"><i class="fa fa-check"></i><b>5.12</b> Heatmaps</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html"><i class="fa fa-check"></i><b>6</b> Using R and Bioconductor for MS-based proteomics</a><ul>
<li class="chapter" data-level="6.1" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#introduction-1"><i class="fa fa-check"></i><b>6.1</b> Introduction</a></li>
<li class="chapter" data-level="6.2" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#reading-ms-data-into-r"><i class="fa fa-check"></i><b>6.2</b> Reading MS data into R</a><ul>
<li class="chapter" data-level="6.2.1" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#raw-data"><i class="fa fa-check"></i><b>6.2.1</b> Raw data</a></li>
<li class="chapter" data-level="6.2.2" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#identification-data"><i class="fa fa-check"></i><b>6.2.2</b> Identification data</a></li>
<li class="chapter" data-level="6.2.3" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#quantitative-data"><i class="fa fa-check"></i><b>6.2.3</b> Quantitative data</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#manipulating-and-visualising-raw-data"><i class="fa fa-check"></i><b>6.3</b> Manipulating and visualising raw data</a></li>
<li class="chapter" data-level="6.4" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#peptide-spetrum-matches"><i class="fa fa-check"></i><b>6.4</b> Peptide spetrum matches</a></li>
<li class="chapter" data-level="6.5" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#missing-values"><i class="fa fa-check"></i><b>6.5</b> Missing values</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html"><i class="fa fa-check"></i><b>7</b> Interactive visualisation</a><ul>
<li class="chapter" data-level="7.1" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#interactivity-graphs"><i class="fa fa-check"></i><b>7.1</b> Interactivity graphs</a><ul>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#interactivity"><i class="fa fa-check"></i>Interactivity</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#layers"><i class="fa fa-check"></i>Layers</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#more-components"><i class="fa fa-check"></i>More components</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#interactive-apps"><i class="fa fa-check"></i><b>7.2</b> Interactive apps</a><ul>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#introduction-2"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#creation-of-our-fist-shiny-app"><i class="fa fa-check"></i>Creation of our fist shiny app</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#the-shiny-infrastructure"><i class="fa fa-check"></i>The shiny infrastructure</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#another-shiny-app"><i class="fa fa-check"></i>Another shiny app</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#single-file-app"><i class="fa fa-check"></i>Single-file app</a></li>
<li><a href="interactive-visualisation.html#theres-more-to-shiny">There’s more to <code>shiny</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="session-information.html"><a href="session-information.html"><i class="fa fa-check"></i><b>8</b> Session information</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Visualisation of biomolecular data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">Visualisation of biomolecular data</h1>
<h4 class="author"><em>Laurent Gatto</em></h4>
<h4 class="date"><em>2018-05-04</em></h4>
</div>
<div id="introduction" class="section level1">
<h1><span class="header-section-number">Chapter 1</span> Introduction</h1>
<p>This <em>Visualisation of biomolecular data</em> course was set up as part of the 2018 edition of the May Institute <em>Computation and statistics for mass spectrometry and proteomics</em> at Northeastern University, Boston MA. It is aimed at people who are already familiar with the R language and syntax, and who would like to get a hands-on introduction to visualisation, with a focus on biomolecular data in general, and proteomics in particular. This course is meant to be mostly hands-on, with an intuitive understanding of the underlying techniques.</p>
<p>Let’s use the famous <a href="https://en.wikipedia.org/wiki/Anscombe%27s_quartet"><em>Anscombe’s quartet</em></a> data as a motivating example. This data is composed of 4 pairs of values, <span class="math inline">\((x_1, y_1)\)</span> to <span class="math inline">\((x_4, y_4)\)</span>:</p>
<table>
<thead>
<tr class="header">
<th align="right">x1</th>
<th align="right">x2</th>
<th align="right">x3</th>
<th align="right">x4</th>
<th align="right">y1</th>
<th align="right">y2</th>
<th align="right">y3</th>
<th align="right">y4</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">10</td>
<td align="right">10</td>
<td align="right">10</td>
<td align="right">8</td>
<td align="right">8.04</td>
<td align="right">9.14</td>
<td align="right">7.46</td>
<td align="right">6.58</td>
</tr>
<tr class="even">
<td align="right">8</td>
<td align="right">8</td>
<td align="right">8</td>
<td align="right">8</td>
<td align="right">6.95</td>
<td align="right">8.14</td>
<td align="right">6.77</td>
<td align="right">5.76</td>
</tr>
<tr class="odd">
<td align="right">13</td>
<td align="right">13</td>
<td align="right">13</td>
<td align="right">8</td>
<td align="right">7.58</td>
<td align="right">8.74</td>
<td align="right">12.74</td>
<td align="right">7.71</td>
</tr>
<tr class="even">
<td align="right">9</td>
<td align="right">9</td>
<td align="right">9</td>
<td align="right">8</td>
<td align="right">8.81</td>
<td align="right">8.77</td>
<td align="right">7.11</td>
<td align="right">8.84</td>
</tr>
<tr class="odd">
<td align="right">11</td>
<td align="right">11</td>
<td align="right">11</td>
<td align="right">8</td>
<td align="right">8.33</td>
<td align="right">9.26</td>
<td align="right">7.81</td>
<td align="right">8.47</td>
</tr>
<tr class="even">
<td align="right">14</td>
<td align="right">14</td>
<td align="right">14</td>
<td align="right">8</td>
<td align="right">9.96</td>
<td align="right">8.10</td>
<td align="right">8.84</td>
<td align="right">7.04</td>
</tr>
<tr class="odd">
<td align="right">6</td>
<td align="right">6</td>
<td align="right">6</td>
<td align="right">8</td>
<td align="right">7.24</td>
<td align="right">6.13</td>
<td align="right">6.08</td>
<td align="right">5.25</td>
</tr>
<tr class="even">
<td align="right">4</td>
<td align="right">4</td>
<td align="right">4</td>
<td align="right">19</td>
<td align="right">4.26</td>
<td align="right">3.10</td>
<td align="right">5.39</td>
<td align="right">12.50</td>
</tr>
<tr class="odd">
<td align="right">12</td>
<td align="right">12</td>
<td align="right">12</td>
<td align="right">8</td>
<td align="right">10.84</td>
<td align="right">9.13</td>
<td align="right">8.15</td>
<td align="right">5.56</td>
</tr>
<tr class="even">
<td align="right">7</td>
<td align="right">7</td>
<td align="right">7</td>
<td align="right">8</td>
<td align="right">4.82</td>
<td align="right">7.26</td>
<td align="right">6.42</td>
<td align="right">7.91</td>
</tr>
<tr class="odd">
<td align="right">5</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">8</td>
<td align="right">5.68</td>
<td align="right">4.74</td>
<td align="right">5.73</td>
<td align="right">6.89</td>
</tr>
</tbody>
</table>
<p>Each of these <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> sets have the same variance, mean and correlation:</p>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">1</th>
<th align="right">2</th>
<th align="right">3</th>
<th align="right">4</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>var(x)</td>
<td align="right">11.0000000</td>
<td align="right">11.0000000</td>
<td align="right">11.0000000</td>
<td align="right">11.0000000</td>
</tr>
<tr class="even">
<td>mean(x)</td>
<td align="right">9.0000000</td>
<td align="right">9.0000000</td>
<td align="right">9.0000000</td>
<td align="right">9.0000000</td>
</tr>
<tr class="odd">
<td>var(y)</td>
<td align="right">4.1272691</td>
<td align="right">4.1276291</td>
<td align="right">4.1226200</td>
<td align="right">4.1232491</td>
</tr>
<tr class="even">
<td>mean(y)</td>
<td align="right">7.5009091</td>
<td align="right">7.5009091</td>
<td align="right">7.5000000</td>
<td align="right">7.5009091</td>
</tr>
<tr class="odd">
<td>cor(x,y)</td>
<td align="right">0.8164205</td>
<td align="right">0.8162365</td>
<td align="right">0.8162867</td>
<td align="right">0.8165214</td>
</tr>
</tbody>
</table>
<p>But…</p>
<p>While the <em>residuals</em> of the linear regression clearly indicate fundamental differences in these data, the most simple and straightforward approach is <em>visualisation</em> to highlight the fundamental differences in the datasets.</p>
<p><img src="VisualisingBiomolecularData_files/figure-html/anscombefig-1.png" width="672" /></p>
<p>See also another, more recent example: <a href="https://www.autodeskresearch.com/publications/samestats">The Datasaurus Dozen dataset</a>.</p>
<div class="figure">
<img src="https://d2f99xq7vri1nk.cloudfront.net/DinoSequentialSmaller.gif" alt="The Datasaurus Dozen dataset" />
<p class="caption">The Datasaurus Dozen dataset</p>
</div>
<p>We will focus on producing visualisation that will enable <strong>understanding important features</strong> of biomolecular data or the effect of their transformation and analyses. With this in mind, the beauty of the visualisations we will produce won’t be assessed by how visually attractive the figures are, but how they help us understand the data. In the process of data exploration and data analysis, we want to be able to quickly generate and interpret figures, to progress in our understanding. It is of course important to fine tune the graphics to make them visually appealing and improve communication with the audience, but we won’t focus on these aspects here.</p>
<p>An important aspect of data visualisation is data <strong>manipulation</strong>, transformation and the format of the data. These topics will be introduced and documented throughout the course.</p>
<p>A last feature of data that is intimately associated with its manipulation and visualisation is its <strong>structure</strong>. Some visualisations will be much easier if the data is formatted in a certain way. As such, we will describe different formats (long, wide) and data structures (dataframes and more specialised data structure).</p>
<p>The source code for this document is available on GitHub at <a href="https://github.com/lgatto/VisualisingBiomolecularData" class="uri">https://github.com/lgatto/VisualisingBiomolecularData</a></p>
<p>A short URL for this book is <a href="http://bit.ly/biomolvis" class="uri">http://bit.ly/biomolvis</a></p>
<div id="installation" class="section level2">
<h2><span class="header-section-number">1.1</span> Installation</h2>
<p>A set of packages that are used, either directly or indirectly are provided below. A complete session information with all packages used to compile this document is available at the end.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;affy&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;ALL&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;bioDist&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;ComplexHeatmap&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;ggplot2&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;ggvis&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;knitr&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;limma&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;magrittr&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;MSnbase&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;msdata&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;MSstats&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;MSstatsBioData&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;patchwork&quot;</span>) ## on GitHub at thomasp85/patchwork
<span class="kw">library</span>(<span class="st">&quot;plyr&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;pRoloc&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;pRolocdata&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;readr&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;RforProteomics&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;Rtsne&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;shiny&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;tidyr&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;tidyverse&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;UpSetR&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;Vennerable&quot;</span>) ## on GitHub at js229/Vennerable</code></pre></div>
<p>To install all dependencies, you can run the <a href="https://github.com/lgatto/VisualisingBiomolecularData/blob/master/src/install.R">installation script</a> by running the following line in R:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">&quot;http://bit.ly/biomolvis_install&quot;</span>)</code></pre></div>
</div>
<div id="questions" class="section level2">
<h2><span class="header-section-number">1.2</span> Questions</h2>
<p>For any questions or suggestions, please open an <a href="https://github.com/lgatto/VisualisingBiomolecularData/issues">issue</a>. Please do add the output of your session information and, of possible, a reproducible example describing your question or suggestion.</p>
</div>
<div id="license" class="section level2">
<h2><span class="header-section-number">1.3</span> License</h2>
<p>This material is licensed under the <a href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 License</a>. Some content is inspired by other sources, see the <em>Credit</em> section in the material.</p>
</div>
<div id="useful-references" class="section level2">
<h2><span class="header-section-number">1.4</span> Useful references</h2>
<p>Gatto L, Breckels LM, Naake T, Gibb S. <em>Visualization of proteomics data using R and bioconductor</em>. Proteomics. 2015 Apr;15(8):1375-89. doi: <a href="https://doi.org/10.1002/pmic.201400392">10.1002/pmic.201400392</a>. PMID: 25690415; PMCID: PMC4510819.</p>
<p>Key M. <em>A tutorial in displaying mass spectrometry-based proteomic data using heat maps</em>. BMC Bioinformatics. 2012;13 Suppl 16:S10. doi: <a href="http://dx.doi.org/10.1186/1471-2105-13-S16-S10">10.1186/1471-2105-13-S16-S10</a>. Epub 2012 Nov 5. PMID: 23176119; PMCID: PMC3489527.</p>
<p>Gatto L, Christoforou A. <em>Using R and Bioconductor for proteomics data analysis</em>. Biochim Biophys Acta. 2014 Jan;1844(1 Pt A):42-51. doi: <a href="http://dx.doi.org/10.1016/j.bbapap.2013.04.032">10.1016/j.bbapap.2013.04.032</a>. Epub 2013 May 18. PMID: 23692960.</p>
<p>Conway JR, Lex A, Gehlenborg N. UpSetR: an R package for the visualization of intersecting sets and their properties. Bioinformatics. 2017 Sep 15;33(18):2938-2940. doi: <a href="https://doi.org/10.1093/bioinformatics/btx364">10.1093/bioinformatics/btx364</a>. PMID: 28645171; PMCID: PMC5870712.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="example-datasets-and-data-manipulation.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
