<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Visualisation of biomolecular data</title>
  <meta name="description" content="Visualisation of biomolecular data">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Visualisation of biomolecular data" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Visualisation of biomolecular data" />
  <meta name="github-repo" content="lgatto/VisualisingBiomolecularData" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Visualisation of biomolecular data" />
  
  <meta name="twitter:description" content="Visualisation of biomolecular data" />
  

<meta name="author" content="Laurent Gatto">



  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="using-r-and-bioconductor-for-ms-based-proteomics.html">
<link rel="next" href="session-information.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script src="libs/htmlwidgets-1.2/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.7.1/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.0.0/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.0.0/js/crosstalk.min.js"></script>
<link href="libs/plotlyjs-1.29.2/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotlyjs-1.29.2/plotly-latest.min.js"></script>


<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i><b>1.1</b> Installation</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#questions"><i class="fa fa-check"></i><b>1.2</b> Questions</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.3</b> License</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#useful-references"><i class="fa fa-check"></i><b>1.4</b> Useful references</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="example-datasets.html"><a href="example-datasets.html"><i class="fa fa-check"></i><b>2</b> Example datasets</a><ul>
<li class="chapter" data-level="2.1" data-path="example-datasets.html"><a href="example-datasets.html#raw-ms-data"><i class="fa fa-check"></i><b>2.1</b> Raw MS data</a></li>
<li class="chapter" data-level="2.2" data-path="example-datasets.html"><a href="example-datasets.html#the-iprg-data"><i class="fa fa-check"></i><b>2.2</b> The iPRG data</a></li>
<li class="chapter" data-level="2.3" data-path="example-datasets.html"><a href="example-datasets.html#crc-training-data"><i class="fa fa-check"></i><b>2.3</b> CRC training data</a></li>
<li class="chapter" data-level="2.4" data-path="example-datasets.html"><a href="example-datasets.html#time-course-from-mulvey-et-al.-2015"><i class="fa fa-check"></i><b>2.4</b> Time course from Mulvey <em>et al.</em> 2015</a></li>
<li class="chapter" data-level="2.5" data-path="example-datasets.html"><a href="example-datasets.html#all-data"><i class="fa fa-check"></i><b>2.5</b> ALL data</a></li>
<li class="chapter" data-level="2.6" data-path="example-datasets.html"><a href="example-datasets.html#spatial-proteomics-data"><i class="fa fa-check"></i><b>2.6</b> Spatial proteomics data</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html"><i class="fa fa-check"></i><b>3</b> Exploring and visualising biomolecular data</a><ul>
<li class="chapter" data-level="" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html#raw-data"><i class="fa fa-check"></i>Raw data</a></li>
<li class="chapter" data-level="" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html#tracking-an-ion-from-the-chromatogram-to-the-ms2-spectra"><i class="fa fa-check"></i>Tracking an ion from the chromatogram to the MS2 spectra</a></li>
<li class="chapter" data-level="" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html#raw-data-qc"><i class="fa fa-check"></i>Raw data QC</a></li>
<li class="chapter" data-level="" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html#exploring-the-experimental-design"><i class="fa fa-check"></i>Exploring the experimental design</a></li>
<li class="chapter" data-level="" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html#quantitative-data"><i class="fa fa-check"></i>Quantitative data</a></li>
<li class="chapter" data-level="" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html#differential-expression"><i class="fa fa-check"></i>Differential expression</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="using-dplyr-for-data-manipulation.html"><a href="using-dplyr-for-data-manipulation.html"><i class="fa fa-check"></i><b>4</b> Using <code>dplyr</code> for data manipulation</a><ul>
<li class="chapter" data-level="4.1" data-path="using-dplyr-for-data-manipulation.html"><a href="using-dplyr-for-data-manipulation.html#selecting-columns-and-filtering-rows"><i class="fa fa-check"></i><b>4.1</b> Selecting columns and filtering rows</a></li>
<li class="chapter" data-level="4.2" data-path="using-dplyr-for-data-manipulation.html"><a href="using-dplyr-for-data-manipulation.html#pipes"><i class="fa fa-check"></i><b>4.2</b> Pipes</a></li>
<li class="chapter" data-level="4.3" data-path="using-dplyr-for-data-manipulation.html"><a href="using-dplyr-for-data-manipulation.html#mutate"><i class="fa fa-check"></i><b>4.3</b> Mutate</a></li>
<li class="chapter" data-level="4.4" data-path="using-dplyr-for-data-manipulation.html"><a href="using-dplyr-for-data-manipulation.html#split-apply-combine-data-analysis-and-the-summarize-function"><i class="fa fa-check"></i><b>4.4</b> Split-apply-combine data analysis and the summarize() function</a><ul>
<li><a href="using-dplyr-for-data-manipulation.html#the-summarize-function">The <code>summarize()</code> function</a></li>
<li class="chapter" data-level="" data-path="using-dplyr-for-data-manipulation.html"><a href="using-dplyr-for-data-manipulation.html#tallying"><i class="fa fa-check"></i>Tallying</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="plotting-in-r.html"><a href="plotting-in-r.html"><i class="fa fa-check"></i><b>5</b> Plotting in R</a><ul>
<li class="chapter" data-level="5.1" data-path="plotting-in-r.html"><a href="plotting-in-r.html#base-graphics"><i class="fa fa-check"></i><b>5.1</b> Base graphics</a></li>
<li class="chapter" data-level="5.2" data-path="plotting-in-r.html"><a href="plotting-in-r.html#plotting-with-ggplot2"><i class="fa fa-check"></i><b>5.2</b> Plotting with <code>ggplot2</code></a></li>
<li class="chapter" data-level="5.3" data-path="plotting-in-r.html"><a href="plotting-in-r.html#building-complex-plots"><i class="fa fa-check"></i><b>5.3</b> Building complex plots</a></li>
<li class="chapter" data-level="5.4" data-path="plotting-in-r.html"><a href="plotting-in-r.html#colour-scales"><i class="fa fa-check"></i><b>5.4</b> Colour scales</a></li>
<li class="chapter" data-level="5.5" data-path="plotting-in-r.html"><a href="plotting-in-r.html#customising-plots"><i class="fa fa-check"></i><b>5.5</b> Customising plots</a></li>
<li class="chapter" data-level="5.6" data-path="plotting-in-r.html"><a href="plotting-in-r.html#combine-ggplot-figures"><i class="fa fa-check"></i><b>5.6</b> Combine <code>ggplot</code> figures</a></li>
<li class="chapter" data-level="5.7" data-path="plotting-in-r.html"><a href="plotting-in-r.html#saving-your-figures"><i class="fa fa-check"></i><b>5.7</b> Saving your figures</a></li>
<li class="chapter" data-level="5.8" data-path="plotting-in-r.html"><a href="plotting-in-r.html#exercises"><i class="fa fa-check"></i><b>5.8</b> Exercises</a></li>
<li class="chapter" data-level="5.9" data-path="plotting-in-r.html"><a href="plotting-in-r.html#references"><i class="fa fa-check"></i><b>5.9</b> References</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="tools-and-plots.html"><a href="tools-and-plots.html"><i class="fa fa-check"></i><b>6</b> Tools and plots</a><ul>
<li class="chapter" data-level="6.1" data-path="tools-and-plots.html"><a href="tools-and-plots.html#transformations"><i class="fa fa-check"></i><b>6.1</b> Transformations</a><ul>
<li class="chapter" data-level="6.1.1" data-path="tools-and-plots.html"><a href="tools-and-plots.html#expression-data"><i class="fa fa-check"></i><b>6.1.1</b> Expression data</a></li>
<li class="chapter" data-level="6.1.2" data-path="tools-and-plots.html"><a href="tools-and-plots.html#fold-changes"><i class="fa fa-check"></i><b>6.1.2</b> Fold changes</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="tools-and-plots.html"><a href="tools-and-plots.html#comparing-samples-and-linear-models"><i class="fa fa-check"></i><b>6.2</b> Comparing samples and linear models</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#challenge"><i class="fa fa-check"></i>Challenge</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="tools-and-plots.html"><a href="tools-and-plots.html#plots-for-statistical-analyses"><i class="fa fa-check"></i><b>6.3</b> Plots for statistical analyses</a><ul>
<li class="chapter" data-level="6.3.1" data-path="tools-and-plots.html"><a href="tools-and-plots.html#p-value-distributions"><i class="fa fa-check"></i><b>6.3.1</b> P-value distributions</a></li>
<li class="chapter" data-level="6.3.2" data-path="tools-and-plots.html"><a href="tools-and-plots.html#volcano-plot"><i class="fa fa-check"></i><b>6.3.2</b> Volcano plot</a></li>
<li class="chapter" data-level="6.3.3" data-path="tools-and-plots.html"><a href="tools-and-plots.html#treating-the-plot-as-a-statistic"><i class="fa fa-check"></i><b>6.3.3</b> Treating the plot as a statistic</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="tools-and-plots.html"><a href="tools-and-plots.html#visualising-intersections"><i class="fa fa-check"></i><b>6.4</b> Visualising intersections</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#challenge-1"><i class="fa fa-check"></i>Challenge</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="tools-and-plots.html"><a href="tools-and-plots.html#unsupervised-learning"><i class="fa fa-check"></i><b>6.5</b> Unsupervised learning</a></li>
<li class="chapter" data-level="6.6" data-path="tools-and-plots.html"><a href="tools-and-plots.html#hierarchical-clustering"><i class="fa fa-check"></i><b>6.6</b> Hierarchical clustering</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#how-does-hierarchical-clustering-work"><i class="fa fa-check"></i>How does hierarchical clustering work</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#defining-clusters"><i class="fa fa-check"></i>Defining clusters</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#effect-of-distance-and-clustering-algorithm"><i class="fa fa-check"></i>Effect of distance and clustering algorithm</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#challenge-2"><i class="fa fa-check"></i>Challenge</a></li>
</ul></li>
<li class="chapter" data-level="6.7" data-path="tools-and-plots.html"><a href="tools-and-plots.html#k-means-clustering"><i class="fa fa-check"></i><b>6.7</b> k-means clustering</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#how-does-k-means-work"><i class="fa fa-check"></i>How does k-means work</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#model-selection"><i class="fa fa-check"></i>Model selection</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#how-to-determine-the-number-of-clusters"><i class="fa fa-check"></i>How to determine the number of clusters</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#challenge-3"><i class="fa fa-check"></i>Challenge</a></li>
</ul></li>
<li class="chapter" data-level="6.8" data-path="tools-and-plots.html"><a href="tools-and-plots.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>6.8</b> Principal component analysis (PCA)</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#how-does-it-work"><i class="fa fa-check"></i>How does it work</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#visualisation"><i class="fa fa-check"></i>Visualisation</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#data-pre-processing"><i class="fa fa-check"></i>Data pre-processing</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#final-comments-on-pca"><i class="fa fa-check"></i>Final comments on PCA</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#challenge-4"><i class="fa fa-check"></i>Challenge</a></li>
</ul></li>
<li class="chapter" data-level="6.9" data-path="tools-and-plots.html"><a href="tools-and-plots.html#t-distributed-stochastic-neighbour-embedding"><i class="fa fa-check"></i><b>6.9</b> t-Distributed Stochastic Neighbour Embedding</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#parameter-tuning"><i class="fa fa-check"></i>Parameter tuning</a></li>
</ul></li>
<li class="chapter" data-level="6.10" data-path="tools-and-plots.html"><a href="tools-and-plots.html#the-curse-of-dimensionality"><i class="fa fa-check"></i><b>6.10</b> The curse of dimensionality</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#as-the-number-of-noisy-proteins-increases-signal-may-be-lost"><i class="fa fa-check"></i>As the number of (noisy) proteins increases, signal may be lost</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#spurious-correlations-may-arise"><i class="fa fa-check"></i>Spurious correlations may arise</a></li>
</ul></li>
<li class="chapter" data-level="6.11" data-path="tools-and-plots.html"><a href="tools-and-plots.html#pca-loadings-vs-differential-expression"><i class="fa fa-check"></i><b>6.11</b> PCA loadings vs differential expression</a></li>
<li class="chapter" data-level="6.12" data-path="tools-and-plots.html"><a href="tools-and-plots.html#heatmaps"><i class="fa fa-check"></i><b>6.12</b> Heatmaps</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html"><i class="fa fa-check"></i><b>7</b> Using R and Bioconductor for MS-based proteomics</a><ul>
<li class="chapter" data-level="7.1" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#introduction-1"><i class="fa fa-check"></i><b>7.1</b> Introduction</a></li>
<li class="chapter" data-level="7.2" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#reading-ms-data-into-r"><i class="fa fa-check"></i><b>7.2</b> Reading MS data into R</a><ul>
<li class="chapter" data-level="" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#raw-data-1"><i class="fa fa-check"></i>Raw data</a></li>
<li class="chapter" data-level="" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#identification-data"><i class="fa fa-check"></i>Identification data</a></li>
<li class="chapter" data-level="" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#quantitative-data-1"><i class="fa fa-check"></i>Quantitative data</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#manipulating-and-visualising-raw-data"><i class="fa fa-check"></i><b>7.3</b> Manipulating and visualising raw data</a></li>
<li class="chapter" data-level="7.4" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#ms-maps"><i class="fa fa-check"></i><b>7.4</b> MS maps</a></li>
<li class="chapter" data-level="7.5" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#extracted-ion-chromatograms"><i class="fa fa-check"></i><b>7.5</b> Extracted ion chromatograms</a></li>
<li class="chapter" data-level="7.6" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#peptide-spetrum-matches"><i class="fa fa-check"></i><b>7.6</b> Peptide spetrum matches</a></li>
<li class="chapter" data-level="7.7" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#ms-imaging"><i class="fa fa-check"></i><b>7.7</b> MS imaging</a></li>
<li class="chapter" data-level="7.8" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#missing-values"><i class="fa fa-check"></i><b>7.8</b> Missing values</a></li>
<li class="chapter" data-level="7.9" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#pca-plots"><i class="fa fa-check"></i><b>7.9</b> PCA plots</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html"><i class="fa fa-check"></i><b>8</b> Interactive visualisation</a><ul>
<li class="chapter" data-level="8.1" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#interactivity-graphs-with-ggiz"><i class="fa fa-check"></i><b>8.1</b> Interactivity graphs with <code>ggiz</code></a><ul>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#interactivity"><i class="fa fa-check"></i>Interactivity</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#layers"><i class="fa fa-check"></i>Layers</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#more-components"><i class="fa fa-check"></i>More components</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#interactive-apps-with-plotly"><i class="fa fa-check"></i><b>8.2</b> Interactive apps with <code>plotly</code></a></li>
<li class="chapter" data-level="8.3" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#interactive-apps"><i class="fa fa-check"></i><b>8.3</b> Interactive apps</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#motivating-examples"><i class="fa fa-check"></i>Motivating examples</a><ul>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#longitudinal-proteomics-qc"><i class="fa fa-check"></i>Longitudinal proteomics QC</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#spatial-proteomics"><i class="fa fa-check"></i>Spatial proteomics</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#introduction-2"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#creation-of-our-fist-shiny-app"><i class="fa fa-check"></i>Creation of our fist shiny app</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#single-file-app"><i class="fa fa-check"></i>Single-file app</a></li>
<li><a href="interactive-visualisation.html#theres-more-to-shiny">There’s more to <code>shiny</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="session-information.html"><a href="session-information.html"><i class="fa fa-check"></i><b>9</b> Session information</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Visualisation of biomolecular data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="interactive-visualisation" class="section level1">
<h1><span class="header-section-number">Chapter 8</span> Interactive visualisation</h1>
<div id="interactivity-graphs-with-ggiz" class="section level2">
<h2><span class="header-section-number">8.1</span> Interactivity graphs with <code>ggiz</code></h2>
<blockquote>
<p>This section is based on the on-line <a href="http://ggvis.rstudio.com/"><code>ggvis</code> documentation</a></p>
</blockquote>
<p>The goal of ggvis is to make it easy to build interactive graphics for exploratory data analysis. <a href="http://ggvis.rstudio.com/"><code>ggvis</code></a> has a similar underlying theory to <code>ggplot2</code> (the grammar of graphics), but it’s expressed a little differently, and adds new features to make your plots interactive. <a href="http://ggvis.rstudio.com/"><code>ggvis</code></a> also incorporates shiny’s reactive programming model (see later) and <code>dplyr</code>’s grammar of data transformation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">crcdf &lt;-<span class="st"> </span>readr<span class="op">::</span><span class="kw">read_csv</span>(<span class="st">&quot;./data/CRC_train.csv&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;ggvis&quot;</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">ggvis</span>(crcdf, <span class="dt">x =</span> <span class="op">~</span>A1AG2, <span class="dt">y =</span> <span class="op">~</span>AFM)
p <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">layer_points</span>()</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">crcdf <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">ggvis</span>(<span class="dt">x =</span> <span class="op">~</span>A1AG2, <span class="dt">y =</span> <span class="op">~</span>AFM,
          <span class="dt">fill =</span> <span class="op">~</span>Group) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">layer_points</span>()</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">layer_points</span>(<span class="dt">fill =</span> <span class="op">~</span>Group)
p <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">layer_points</span>(<span class="dt">shape =</span> <span class="op">~</span>Group)</code></pre></div>
<p>To set fixed plotting parameters, use <code>:=</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">layer_points</span>(fill <span class="op">:</span><span class="er">=</span><span class="st"> &quot;red&quot;</span>, stroke <span class="op">:</span><span class="er">=</span><span class="st"> &quot;black&quot;</span>)
p <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">layer_points</span>(size <span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="dv">300</span>, opacity <span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="fl">0.4</span>)
p <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">layer_points</span>(shape <span class="op">:</span><span class="er">=</span><span class="st"> &quot;cross&quot;</span>)</code></pre></div>
<div id="interactivity" class="section level3 unnumbered">
<h3>Interactivity</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">layer_points</span>(
          size <span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="kw">input_slider</span>(<span class="dv">10</span>, <span class="dv">200</span>),
          opacity <span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="kw">input_slider</span>(<span class="dv">0</span>, <span class="dv">1</span>))</code></pre></div>
<ul>
<li><code>add_tooltip</code></li>
<li><code>input_slider()</code></li>
<li><code>input_checkbox()</code></li>
<li><code>input_checkboxgroup()</code></li>
<li><code>input_numeric()</code></li>
<li><code>input_radiobuttons()</code></li>
<li><code>input_select()</code></li>
<li><code>input_text()</code></li>
</ul>
<p>See the <a href="http://ggvis.rstudio.com/interactivity.html">interactivity vignette</a> for details.</p>
</div>
<div id="layers" class="section level3 unnumbered">
<h3>Layers</h3>
<p><strong>Simple layers</strong></p>
<ul>
<li><code>layer_points()</code>, with properties x, y, shape, stroke, fill, strokeOpacity, fillOpacity, and opacity.</li>
<li><code>layer_paths()</code>, for paths and polygons (using the fill argument).</li>
<li><code>layer_ribbons()</code> for filled areas.</li>
<li><code>layer_rects()</code>, <code>layer_text()</code>.</li>
</ul>
<p><strong>Compound layers</strong>, which combine data transformations with one or more simple layers.</p>
<ul>
<li><code>layer_lines()</code> which automatically orders by the x variable with <code>arrange()</code>.</li>
<li><code>layer_histograms()</code> and <code>layer_freqpolys()</code>, which first bin the data with <code>compute_bin()</code>.</li>
<li><code>layer_smooths()</code>, which fits and plots a smooth model to the data using <code>compute_smooth()</code>.</li>
</ul>
<p>See the <a href="http://ggvis.rstudio.com/layers.html">layers vignette</a> for details.</p>
<p>Like for <code>ggplot2</code>’s geoms, we can overly multiple layers:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">layer_points</span>() <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">layer_smooths</span>(stroke <span class="op">:</span><span class="er">=</span><span class="st"> &quot;red&quot;</span>)</code></pre></div>
</div>
<div id="more-components" class="section level3 unnumbered">
<h3>More components</h3>
<ul>
<li><p><code>scales</code>, to control the mapping between data and visual properties; see the <a href="http://ggvis.rstudio.com/properties-scales.html">properties and scales vignette</a>.</p></li>
<li><p><code>legends</code> and <code>axes</code> to control the appearance of the guides produced by the scales. See the <a href="http://ggvis.rstudio.com/axes-legends.html">axes and legends vignette</a>.</p></li>
</ul>
<blockquote>
<p>Challenge</p>
<p>Apply a PCA analysis on the <code>mulvey2015norm</code> data, and use <code>ggvis</code> to visualise that data along PC1 and PC2, controlling the point size using a slide bar.</p>
</blockquote>
<p><details></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">ggvis</span>(<span class="kw">data.frame</span>(pca<span class="op">$</span>x), <span class="dt">x =</span> <span class="op">~</span>PC1, <span class="dt">y =</span> <span class="op">~</span>PC2)
p <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">layer_points</span>(
          size <span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="kw">input_slider</span>(<span class="dv">10</span>, <span class="dv">200</span>),
          opacity <span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="kw">input_slider</span>(<span class="dv">0</span>, <span class="dv">1</span>))</code></pre></div>
<p></details></p>
</div>
</div>
<div id="interactive-apps-with-plotly" class="section level2">
<h2><span class="header-section-number">8.2</span> Interactive apps with <code>plotly</code></h2>
<p>The <a href="https://github.com/ropensci/plotly"><code>plotly</code></a> package can be used for creating interactive web graphics via the open source JavaScript graphing library <a href="https://github.com/plotly/plotly.js">plotly.js</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;plotly&quot;</span>)
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(<span class="dt">data =</span> crcdf,
            <span class="kw">aes</span>(<span class="dt">x =</span> A1AG2, <span class="dt">y =</span> AFM, <span class="dt">colour =</span> Group)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_point</span>()
<span class="kw">ggplotly</span>(p)</code></pre></div>
<div id="6ff679c23703" style="width:672px;height:480px;" class="plotly html-widget"></div>
<script type="application/json" data-for="6ff679c23703">{"x":{"data":[{"x":[14.238157,15.024111,15.631364,15.401365,16.003158,13.932424,14.341549,13.570857,15.833479,15.379964,16.054507,13.813165,14.417196,14.226599,17.504802,15.153401,14.962772,13.840415,14.842405,17.106094,17.695624,16.482747,14.907142,15.728593,16.949813,14.47464,15.195631,16.119599,15.101139,16.15688,14.802356,16.17295,15.065995,15.225577,15.573785,16.479537,15.106331,15.535962,14.310377,13.105397,14.853809,13.756143,14.646338,12.496269,15.477063,14.408347,15.230216,14.635675,14.463843,14.906309,15.112356,13.784786,12.911864,13.556481,14.440705,14.736968,12.899141,14.985355,14.069918,15.064921,15.666972,14.684303,15.040263,14.974207,15.18897,14.836378,13.904041,14.745929,15.006346,14.935695,14.458605,15.715643,15.079632,16.126295,15.400125,14.185257,14.967862,12.566896,13.863699,13.000453,15.428707,15.425013,15.319064,14.679902,14.032498,14.175604,14.81883,14.146699,14.537166,15.127702,14.612095,15.638066,16.250593,14.901858,15.640738,14.906852,15.645656,15.291199,13.097695,14.801609],"y":[16.103019,16.020714,16.1438,16.276415,16.958207,16.527715,16.430286,17.110525,15.881893,16.226206,16.212388,16.905254,16.290439,16.098064,15.570571,16.528333,15.530522,16.266124,15.672178,15.859402,16.085248,16.813347,15.952163,16.967749,16.482593,16.470773,16.878366,16.291421,16.015487,15.955984,16.163799,16.841158,16.430817,16.301941,16.907288,16.17906,16.260799,17.135373,15.756246,15.895164,15.980673,16.245436,15.976348,16.336003,16.105255,15.945204,16.18727,16.262259,15.615956,16.61246,15.689976,16.099301,16.663352,16.538229,16.263101,16.532692,16.197321,16.433091,15.472298,15.689884,15.293582,16.173061,15.822366,15.923416,15.334568,15.962639,15.488141,15.740257,15.705423,16.155999,15.166897,15.366985,15.689819,16.345171,16.214845,15.844659,16.100135,16.816858,16.274108,16.301133,15.68042,16.176484,15.696797,16.070374,16.330247,15.922015,16.407824,16.14053,16.241626,16.293145,15.387818,16.06951,15.962946,14.819808,16.225643,15.917716,16.279397,16.20803,16.307783,16.148471],"text":["~A1AG2: 14.23816<br />~AFM: 16.10302<br />~Group: CRC","~A1AG2: 15.02411<br />~AFM: 16.02071<br />~Group: CRC","~A1AG2: 15.63136<br />~AFM: 16.14380<br />~Group: CRC","~A1AG2: 15.40137<br />~AFM: 16.27642<br />~Group: CRC","~A1AG2: 16.00316<br />~AFM: 16.95821<br />~Group: CRC","~A1AG2: 13.93242<br />~AFM: 16.52772<br />~Group: CRC","~A1AG2: 14.34155<br />~AFM: 16.43029<br />~Group: CRC","~A1AG2: 13.57086<br />~AFM: 17.11052<br />~Group: CRC","~A1AG2: 15.83348<br />~AFM: 15.88189<br />~Group: CRC","~A1AG2: 15.37996<br />~AFM: 16.22621<br />~Group: CRC","~A1AG2: 16.05451<br />~AFM: 16.21239<br />~Group: CRC","~A1AG2: 13.81316<br />~AFM: 16.90525<br />~Group: CRC","~A1AG2: 14.41720<br />~AFM: 16.29044<br />~Group: CRC","~A1AG2: 14.22660<br />~AFM: 16.09806<br />~Group: CRC","~A1AG2: 17.50480<br />~AFM: 15.57057<br />~Group: CRC","~A1AG2: 15.15340<br />~AFM: 16.52833<br />~Group: CRC","~A1AG2: 14.96277<br />~AFM: 15.53052<br />~Group: CRC","~A1AG2: 13.84042<br />~AFM: 16.26612<br />~Group: CRC","~A1AG2: 14.84240<br />~AFM: 15.67218<br />~Group: CRC","~A1AG2: 17.10609<br />~AFM: 15.85940<br />~Group: CRC","~A1AG2: 17.69562<br />~AFM: 16.08525<br />~Group: CRC","~A1AG2: 16.48275<br />~AFM: 16.81335<br />~Group: CRC","~A1AG2: 14.90714<br />~AFM: 15.95216<br />~Group: CRC","~A1AG2: 15.72859<br />~AFM: 16.96775<br />~Group: CRC","~A1AG2: 16.94981<br />~AFM: 16.48259<br />~Group: CRC","~A1AG2: 14.47464<br />~AFM: 16.47077<br />~Group: CRC","~A1AG2: 15.19563<br />~AFM: 16.87837<br />~Group: CRC","~A1AG2: 16.11960<br />~AFM: 16.29142<br />~Group: CRC","~A1AG2: 15.10114<br />~AFM: 16.01549<br />~Group: CRC","~A1AG2: 16.15688<br />~AFM: 15.95598<br />~Group: CRC","~A1AG2: 14.80236<br />~AFM: 16.16380<br />~Group: CRC","~A1AG2: 16.17295<br />~AFM: 16.84116<br />~Group: CRC","~A1AG2: 15.06599<br />~AFM: 16.43082<br />~Group: CRC","~A1AG2: 15.22558<br />~AFM: 16.30194<br />~Group: CRC","~A1AG2: 15.57379<br />~AFM: 16.90729<br />~Group: CRC","~A1AG2: 16.47954<br />~AFM: 16.17906<br />~Group: CRC","~A1AG2: 15.10633<br />~AFM: 16.26080<br />~Group: CRC","~A1AG2: 15.53596<br />~AFM: 17.13537<br />~Group: CRC","~A1AG2: 14.31038<br />~AFM: 15.75625<br />~Group: CRC","~A1AG2: 13.10540<br />~AFM: 15.89516<br />~Group: CRC","~A1AG2: 14.85381<br />~AFM: 15.98067<br />~Group: CRC","~A1AG2: 13.75614<br />~AFM: 16.24544<br />~Group: CRC","~A1AG2: 14.64634<br />~AFM: 15.97635<br />~Group: CRC","~A1AG2: 12.49627<br />~AFM: 16.33600<br />~Group: CRC","~A1AG2: 15.47706<br />~AFM: 16.10525<br />~Group: CRC","~A1AG2: 14.40835<br />~AFM: 15.94520<br />~Group: CRC","~A1AG2: 15.23022<br />~AFM: 16.18727<br />~Group: CRC","~A1AG2: 14.63568<br />~AFM: 16.26226<br />~Group: CRC","~A1AG2: 14.46384<br />~AFM: 15.61596<br />~Group: CRC","~A1AG2: 14.90631<br />~AFM: 16.61246<br />~Group: CRC","~A1AG2: 15.11236<br />~AFM: 15.68998<br />~Group: CRC","~A1AG2: 13.78479<br />~AFM: 16.09930<br />~Group: CRC","~A1AG2: 12.91186<br />~AFM: 16.66335<br />~Group: CRC","~A1AG2: 13.55648<br />~AFM: 16.53823<br />~Group: CRC","~A1AG2: 14.44070<br />~AFM: 16.26310<br />~Group: CRC","~A1AG2: 14.73697<br />~AFM: 16.53269<br />~Group: CRC","~A1AG2: 12.89914<br />~AFM: 16.19732<br />~Group: CRC","~A1AG2: 14.98536<br />~AFM: 16.43309<br />~Group: CRC","~A1AG2: 14.06992<br />~AFM: 15.47230<br />~Group: CRC","~A1AG2: 15.06492<br />~AFM: 15.68988<br />~Group: CRC","~A1AG2: 15.66697<br />~AFM: 15.29358<br />~Group: CRC","~A1AG2: 14.68430<br />~AFM: 16.17306<br />~Group: CRC","~A1AG2: 15.04026<br />~AFM: 15.82237<br />~Group: CRC","~A1AG2: 14.97421<br />~AFM: 15.92342<br />~Group: CRC","~A1AG2: 15.18897<br />~AFM: 15.33457<br />~Group: CRC","~A1AG2: 14.83638<br />~AFM: 15.96264<br />~Group: CRC","~A1AG2: 13.90404<br />~AFM: 15.48814<br />~Group: CRC","~A1AG2: 14.74593<br />~AFM: 15.74026<br />~Group: CRC","~A1AG2: 15.00635<br />~AFM: 15.70542<br />~Group: CRC","~A1AG2: 14.93570<br />~AFM: 16.15600<br />~Group: CRC","~A1AG2: 14.45861<br />~AFM: 15.16690<br />~Group: CRC","~A1AG2: 15.71564<br />~AFM: 15.36698<br />~Group: CRC","~A1AG2: 15.07963<br />~AFM: 15.68982<br />~Group: CRC","~A1AG2: 16.12629<br />~AFM: 16.34517<br />~Group: CRC","~A1AG2: 15.40012<br />~AFM: 16.21485<br />~Group: CRC","~A1AG2: 14.18526<br />~AFM: 15.84466<br />~Group: CRC","~A1AG2: 14.96786<br />~AFM: 16.10014<br />~Group: CRC","~A1AG2: 12.56690<br />~AFM: 16.81686<br />~Group: CRC","~A1AG2: 13.86370<br />~AFM: 16.27411<br />~Group: CRC","~A1AG2: 13.00045<br />~AFM: 16.30113<br />~Group: CRC","~A1AG2: 15.42871<br />~AFM: 15.68042<br />~Group: CRC","~A1AG2: 15.42501<br />~AFM: 16.17648<br />~Group: CRC","~A1AG2: 15.31906<br />~AFM: 15.69680<br />~Group: CRC","~A1AG2: 14.67990<br />~AFM: 16.07037<br />~Group: CRC","~A1AG2: 14.03250<br />~AFM: 16.33025<br />~Group: CRC","~A1AG2: 14.17560<br />~AFM: 15.92202<br />~Group: CRC","~A1AG2: 14.81883<br />~AFM: 16.40782<br />~Group: CRC","~A1AG2: 14.14670<br />~AFM: 16.14053<br />~Group: CRC","~A1AG2: 14.53717<br />~AFM: 16.24163<br />~Group: CRC","~A1AG2: 15.12770<br />~AFM: 16.29314<br />~Group: CRC","~A1AG2: 14.61210<br />~AFM: 15.38782<br />~Group: CRC","~A1AG2: 15.63807<br />~AFM: 16.06951<br />~Group: CRC","~A1AG2: 16.25059<br />~AFM: 15.96295<br />~Group: CRC","~A1AG2: 14.90186<br />~AFM: 14.81981<br />~Group: CRC","~A1AG2: 15.64074<br />~AFM: 16.22564<br />~Group: CRC","~A1AG2: 14.90685<br />~AFM: 15.91772<br />~Group: CRC","~A1AG2: 15.64566<br />~AFM: 16.27940<br />~Group: CRC","~A1AG2: 15.29120<br />~AFM: 16.20803<br />~Group: CRC","~A1AG2: 13.09769<br />~AFM: 16.30778<br />~Group: CRC","~A1AG2: 14.80161<br />~AFM: 16.14847<br />~Group: CRC"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(248,118,109,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(248,118,109,1)"}},"hoveron":"points","name":"CRC","legendgroup":"CRC","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[14.522701,15.294532,14.317982,14.743974,14.473524,14.893401,14.865135,14.674806,14.277397,16.60895,14.963164,14.935335,15.247035,13.989382,14.436756,14.420802,14.35954,12.987059,17.126513,15.263615,14.505181,15.188011,14.739943,15.401143,14.865775,14.344033,14.172972,15.089892,15.262423,15.819156,15.5316,15.167554,14.95086,15.173166,15.187836,15.493931,15.839731,15.77099,14.156179,13.0177,14.127393,14.464952,14.615513,13.735682,14.233264,13.75953,12.338793,13.94161,15.061908,14.236917,14.689308,14.100373,14.70698,14.567964,13.812276,14.342497,14.465143,13.890418,13.371271,14.655093,15.26905,16.463754,15.273983,15.328769,14.791528,15.364615,15.429827,14.524318,14.563405,14.205807,15.060279,14.630145,15.022399,14.942586,15.585966,14.744899,14.172242,14.916165,13.277184,12.834799,14.444528,14.937208,15.753986,14.528422,14.074783,15.313827,14.537259,14.896764,13.378797,15.502738,13.87927,14.006765,14.837407,15.10888,15.008987,14.442917,15.116044,15.041488,13.922867,14.280702],"y":[15.940212,15.93552,16.0293,17.271903,15.997162,16.82491,16.655651,15.370228,16.217793,16.331384,16.340713,16.935429,16.195821,16.172152,15.686121,16.672321,16.125007,17.131209,15.820848,16.277179,16.733438,16.772306,15.705068,15.968328,15.682889,16.272519,15.838089,16.488223,16.302309,16.066615,16.803682,17.090921,16.279946,16.606197,16.171934,16.736472,16.58605,16.080103,16.040187,16.629872,16.097788,16.943532,16.031266,16.62834,16.228692,16.488454,15.961567,16.266452,16.583249,15.897937,15.195405,15.600361,16.306882,16.333232,15.170725,16.292834,15.997148,16.685588,16.578029,16.560325,16.740174,15.987205,16.387509,16.086814,15.901194,16.582726,16.089807,16.069582,16.02804,15.958482,16.685005,15.503696,16.070656,15.590423,15.908833,16.392224,16.313595,15.236765,17.009019,16.84931,16.333884,16.269961,15.66371,16.022942,16.560685,14.94242,16.041537,16.249299,16.698717,15.907995,16.200327,15.393334,15.694819,16.261503,16.156808,15.990122,16.625768,15.224106,16.388432,15.676063],"text":["~A1AG2: 14.52270<br />~AFM: 15.94021<br />~Group: Healthy","~A1AG2: 15.29453<br />~AFM: 15.93552<br />~Group: Healthy","~A1AG2: 14.31798<br />~AFM: 16.02930<br />~Group: Healthy","~A1AG2: 14.74397<br />~AFM: 17.27190<br />~Group: Healthy","~A1AG2: 14.47352<br />~AFM: 15.99716<br />~Group: Healthy","~A1AG2: 14.89340<br />~AFM: 16.82491<br />~Group: Healthy","~A1AG2: 14.86514<br />~AFM: 16.65565<br />~Group: Healthy","~A1AG2: 14.67481<br />~AFM: 15.37023<br />~Group: Healthy","~A1AG2: 14.27740<br />~AFM: 16.21779<br />~Group: Healthy","~A1AG2: 16.60895<br />~AFM: 16.33138<br />~Group: Healthy","~A1AG2: 14.96316<br />~AFM: 16.34071<br />~Group: Healthy","~A1AG2: 14.93534<br />~AFM: 16.93543<br />~Group: Healthy","~A1AG2: 15.24704<br />~AFM: 16.19582<br />~Group: Healthy","~A1AG2: 13.98938<br />~AFM: 16.17215<br />~Group: Healthy","~A1AG2: 14.43676<br />~AFM: 15.68612<br />~Group: Healthy","~A1AG2: 14.42080<br />~AFM: 16.67232<br />~Group: Healthy","~A1AG2: 14.35954<br />~AFM: 16.12501<br />~Group: Healthy","~A1AG2: 12.98706<br />~AFM: 17.13121<br />~Group: Healthy","~A1AG2: 17.12651<br />~AFM: 15.82085<br />~Group: Healthy","~A1AG2: 15.26361<br />~AFM: 16.27718<br />~Group: Healthy","~A1AG2: 14.50518<br />~AFM: 16.73344<br />~Group: Healthy","~A1AG2: 15.18801<br />~AFM: 16.77231<br />~Group: Healthy","~A1AG2: 14.73994<br />~AFM: 15.70507<br />~Group: Healthy","~A1AG2: 15.40114<br />~AFM: 15.96833<br />~Group: Healthy","~A1AG2: 14.86577<br />~AFM: 15.68289<br />~Group: Healthy","~A1AG2: 14.34403<br />~AFM: 16.27252<br />~Group: Healthy","~A1AG2: 14.17297<br />~AFM: 15.83809<br />~Group: Healthy","~A1AG2: 15.08989<br />~AFM: 16.48822<br />~Group: Healthy","~A1AG2: 15.26242<br />~AFM: 16.30231<br />~Group: Healthy","~A1AG2: 15.81916<br />~AFM: 16.06661<br />~Group: Healthy","~A1AG2: 15.53160<br />~AFM: 16.80368<br />~Group: Healthy","~A1AG2: 15.16755<br />~AFM: 17.09092<br />~Group: Healthy","~A1AG2: 14.95086<br />~AFM: 16.27995<br />~Group: Healthy","~A1AG2: 15.17317<br />~AFM: 16.60620<br />~Group: Healthy","~A1AG2: 15.18784<br />~AFM: 16.17193<br />~Group: Healthy","~A1AG2: 15.49393<br />~AFM: 16.73647<br />~Group: Healthy","~A1AG2: 15.83973<br />~AFM: 16.58605<br />~Group: Healthy","~A1AG2: 15.77099<br />~AFM: 16.08010<br />~Group: Healthy","~A1AG2: 14.15618<br />~AFM: 16.04019<br />~Group: Healthy","~A1AG2: 13.01770<br />~AFM: 16.62987<br />~Group: Healthy","~A1AG2: 14.12739<br />~AFM: 16.09779<br />~Group: Healthy","~A1AG2: 14.46495<br />~AFM: 16.94353<br />~Group: Healthy","~A1AG2: 14.61551<br />~AFM: 16.03127<br />~Group: Healthy","~A1AG2: 13.73568<br />~AFM: 16.62834<br />~Group: Healthy","~A1AG2: 14.23326<br />~AFM: 16.22869<br />~Group: Healthy","~A1AG2: 13.75953<br />~AFM: 16.48845<br />~Group: Healthy","~A1AG2: 12.33879<br />~AFM: 15.96157<br />~Group: Healthy","~A1AG2: 13.94161<br />~AFM: 16.26645<br />~Group: Healthy","~A1AG2: 15.06191<br />~AFM: 16.58325<br />~Group: Healthy","~A1AG2: 14.23692<br />~AFM: 15.89794<br />~Group: Healthy","~A1AG2: 14.68931<br />~AFM: 15.19540<br />~Group: Healthy","~A1AG2: 14.10037<br />~AFM: 15.60036<br />~Group: Healthy","~A1AG2: 14.70698<br />~AFM: 16.30688<br />~Group: Healthy","~A1AG2: 14.56796<br />~AFM: 16.33323<br />~Group: Healthy","~A1AG2: 13.81228<br />~AFM: 15.17072<br />~Group: Healthy","~A1AG2: 14.34250<br />~AFM: 16.29283<br />~Group: Healthy","~A1AG2: 14.46514<br />~AFM: 15.99715<br />~Group: Healthy","~A1AG2: 13.89042<br />~AFM: 16.68559<br />~Group: Healthy","~A1AG2: 13.37127<br />~AFM: 16.57803<br />~Group: Healthy","~A1AG2: 14.65509<br />~AFM: 16.56032<br />~Group: Healthy","~A1AG2: 15.26905<br />~AFM: 16.74017<br />~Group: Healthy","~A1AG2: 16.46375<br />~AFM: 15.98720<br />~Group: Healthy","~A1AG2: 15.27398<br />~AFM: 16.38751<br />~Group: Healthy","~A1AG2: 15.32877<br />~AFM: 16.08681<br />~Group: Healthy","~A1AG2: 14.79153<br />~AFM: 15.90119<br />~Group: Healthy","~A1AG2: 15.36462<br />~AFM: 16.58273<br />~Group: Healthy","~A1AG2: 15.42983<br />~AFM: 16.08981<br />~Group: Healthy","~A1AG2: 14.52432<br />~AFM: 16.06958<br />~Group: Healthy","~A1AG2: 14.56340<br />~AFM: 16.02804<br />~Group: Healthy","~A1AG2: 14.20581<br />~AFM: 15.95848<br />~Group: Healthy","~A1AG2: 15.06028<br />~AFM: 16.68501<br />~Group: Healthy","~A1AG2: 14.63015<br />~AFM: 15.50370<br />~Group: Healthy","~A1AG2: 15.02240<br />~AFM: 16.07066<br />~Group: Healthy","~A1AG2: 14.94259<br />~AFM: 15.59042<br />~Group: Healthy","~A1AG2: 15.58597<br />~AFM: 15.90883<br />~Group: Healthy","~A1AG2: 14.74490<br />~AFM: 16.39222<br />~Group: Healthy","~A1AG2: 14.17224<br />~AFM: 16.31359<br />~Group: Healthy","~A1AG2: 14.91616<br />~AFM: 15.23677<br />~Group: Healthy","~A1AG2: 13.27718<br />~AFM: 17.00902<br />~Group: Healthy","~A1AG2: 12.83480<br />~AFM: 16.84931<br />~Group: Healthy","~A1AG2: 14.44453<br />~AFM: 16.33388<br />~Group: Healthy","~A1AG2: 14.93721<br />~AFM: 16.26996<br />~Group: Healthy","~A1AG2: 15.75399<br />~AFM: 15.66371<br />~Group: Healthy","~A1AG2: 14.52842<br />~AFM: 16.02294<br />~Group: Healthy","~A1AG2: 14.07478<br />~AFM: 16.56068<br />~Group: Healthy","~A1AG2: 15.31383<br />~AFM: 14.94242<br />~Group: Healthy","~A1AG2: 14.53726<br />~AFM: 16.04154<br />~Group: Healthy","~A1AG2: 14.89676<br />~AFM: 16.24930<br />~Group: Healthy","~A1AG2: 13.37880<br />~AFM: 16.69872<br />~Group: Healthy","~A1AG2: 15.50274<br />~AFM: 15.90799<br />~Group: Healthy","~A1AG2: 13.87927<br />~AFM: 16.20033<br />~Group: Healthy","~A1AG2: 14.00676<br />~AFM: 15.39333<br />~Group: Healthy","~A1AG2: 14.83741<br />~AFM: 15.69482<br />~Group: Healthy","~A1AG2: 15.10888<br />~AFM: 16.26150<br />~Group: Healthy","~A1AG2: 15.00899<br />~AFM: 16.15681<br />~Group: Healthy","~A1AG2: 14.44292<br />~AFM: 15.99012<br />~Group: Healthy","~A1AG2: 15.11604<br />~AFM: 16.62577<br />~Group: Healthy","~A1AG2: 15.04149<br />~AFM: 15.22411<br />~Group: Healthy","~A1AG2: 13.92287<br />~AFM: 16.38843<br />~Group: Healthy","~A1AG2: 14.28070<br />~AFM: 15.67606<br />~Group: Healthy"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,191,196,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,191,196,1)"}},"hoveron":"points","name":"Healthy","legendgroup":"Healthy","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":26.2283105022831,"r":7.30593607305936,"b":40.1826484018265,"l":48.9497716894977},"plot_bgcolor":"rgba(235,235,235,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xaxis":{"domain":[0,1],"type":"linear","autorange":false,"range":[12.07095145,17.96346555],"tickmode":"array","ticktext":["13","14","15","16","17"],"tickvals":[13,14,15,16,17],"categoryorder":"array","categoryarray":["13","14","15","16","17"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"y","title":"A1AG2","titlefont":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"type":"linear","autorange":false,"range":[14.69720325,17.39450775],"tickmode":"array","ticktext":["15.0","15.5","16.0","16.5","17.0"],"tickvals":[15,15.5,16,16.5,17],"categoryorder":"array","categoryarray":["15.0","15.5","16.0","16.5","17.0"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"x","title":"AFM","titlefont":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":true,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"y":0.913385826771654},"annotations":[{"text":"Group","x":1.02,"y":1,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"left","yanchor":"bottom","legendTitle":true}],"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","modeBarButtonsToAdd":[{"name":"Collaborate","icon":{"width":1000,"ascent":500,"descent":-50,"path":"M487 375c7-10 9-23 5-36l-79-259c-3-12-11-23-22-31-11-8-22-12-35-12l-263 0c-15 0-29 5-43 15-13 10-23 23-28 37-5 13-5 25-1 37 0 0 0 3 1 7 1 5 1 8 1 11 0 2 0 4-1 6 0 3-1 5-1 6 1 2 2 4 3 6 1 2 2 4 4 6 2 3 4 5 5 7 5 7 9 16 13 26 4 10 7 19 9 26 0 2 0 5 0 9-1 4-1 6 0 8 0 2 2 5 4 8 3 3 5 5 5 7 4 6 8 15 12 26 4 11 7 19 7 26 1 1 0 4 0 9-1 4-1 7 0 8 1 2 3 5 6 8 4 4 6 6 6 7 4 5 8 13 13 24 4 11 7 20 7 28 1 1 0 4 0 7-1 3-1 6-1 7 0 2 1 4 3 6 1 1 3 4 5 6 2 3 3 5 5 6 1 2 3 5 4 9 2 3 3 7 5 10 1 3 2 6 4 10 2 4 4 7 6 9 2 3 4 5 7 7 3 2 7 3 11 3 3 0 8 0 13-1l0-1c7 2 12 2 14 2l218 0c14 0 25-5 32-16 8-10 10-23 6-37l-79-259c-7-22-13-37-20-43-7-7-19-10-37-10l-248 0c-5 0-9-2-11-5-2-3-2-7 0-12 4-13 18-20 41-20l264 0c5 0 10 2 16 5 5 3 8 6 10 11l85 282c2 5 2 10 2 17 7-3 13-7 17-13z m-304 0c-1-3-1-5 0-7 1-1 3-2 6-2l174 0c2 0 4 1 7 2 2 2 4 4 5 7l6 18c0 3 0 5-1 7-1 1-3 2-6 2l-173 0c-3 0-5-1-8-2-2-2-4-4-4-7z m-24-73c-1-3-1-5 0-7 2-2 3-2 6-2l174 0c2 0 5 0 7 2 3 2 4 4 5 7l6 18c1 2 0 5-1 6-1 2-3 3-5 3l-174 0c-3 0-5-1-7-3-3-1-4-4-5-6z"},"click":"function(gd) { \n        // is this being viewed in RStudio?\n        if (location.search == '?viewer_pane=1') {\n          alert('To learn about plotly for collaboration, visit:\\n https://cpsievert.github.io/plotly_book/plot-ly-for-collaboration.html');\n        } else {\n          window.open('https://cpsievert.github.io/plotly_book/plot-ly-for-collaboration.html', '_blank');\n        }\n      }"}],"cloud":false},"source":"A","attrs":{"6ff6343ac1b6":{"x":{},"y":{},"colour":{},"type":"scatter"}},"cur_data":"6ff6343ac1b6","visdat":{"6ff6343ac1b6":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1}},"base_url":"https://plot.ly"},"evals":["config.modeBarButtonsToAdd.0.click"],"jsHooks":{"render":[{"code":"function(el, x) { var ctConfig = crosstalk.var('plotlyCrosstalkOpts').set({\"on\":\"plotly_click\",\"persistent\":false,\"dynamic\":false,\"selectize\":false,\"opacityDim\":0.2,\"selected\":{\"opacity\":1}}); }","data":null}]}}</script>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">crcdf &lt;-<span class="st"> </span>crcdf[<span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">191</span><span class="op">:</span><span class="dv">200</span>), <span class="kw">c</span>(<span class="dv">70</span><span class="op">:</span><span class="dv">74</span>)]
x &lt;-<span class="st"> </span><span class="kw">gather</span>(crcdf,
            <span class="dt">key =</span> Protein, <span class="dt">value =</span> x,
            <span class="op">-</span>Sample,
            <span class="op">-</span>Group)
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(<span class="dt">data =</span> x,
            <span class="kw">aes</span>(<span class="dt">x =</span> Sample, <span class="dt">y =</span> x,
                <span class="dt">group =</span> Protein,
                <span class="dt">colour =</span> Protein)) <span class="op">+</span>
<span class="st">        </span><span class="kw">geom_line</span>()


<span class="kw">ggplotly</span>(p)</code></pre></div>
<div id="6ff6ca8e4c8" style="width:672px;height:480px;" class="plotly html-widget"></div>
<script type="application/json" data-for="6ff6ca8e4c8">{"x":{"data":[{"x":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20],"y":[10.275261,10.234434,10.241626,9.269195,10.476927,10.026299,9.633794,10.632307,11.214727,10.573414,null,8.851866,9.159705,10.97353,10.022583,9.847223,null,null,null,9.0326],"text":["~Sample: P1A10<br />~x: 10.275261<br />~Protein: ~<br />~Protein: TNC","~Sample: P1A2<br />~x: 10.234434<br />~Protein: ~<br />~Protein: TNC","~Sample: P1A4<br />~x: 10.241626<br />~Protein: ~<br />~Protein: TNC","~Sample: P1A6<br />~x:  9.269195<br />~Protein: ~<br />~Protein: TNC","~Sample: P1B12<br />~x: 10.476927<br />~Protein: ~<br />~Protein: TNC","~Sample: P1B2<br />~x: 10.026299<br />~Protein: ~<br />~Protein: TNC","~Sample: P1B3<br />~x:  9.633794<br />~Protein: ~<br />~Protein: TNC","~Sample: P1B6<br />~x: 10.632307<br />~Protein: ~<br />~Protein: TNC","~Sample: P1B9<br />~x: 11.214727<br />~Protein: ~<br />~Protein: TNC","~Sample: P1C11<br />~x: 10.573414<br />~Protein: ~<br />~Protein: TNC","~Sample: P3F3<br />~x:        NA<br />~Protein: ~<br />~Protein: TNC","~Sample: P3F5<br />~x:  8.851866<br />~Protein: ~<br />~Protein: TNC","~Sample: P3F7<br />~x:  9.159705<br />~Protein: ~<br />~Protein: TNC","~Sample: P3G10<br />~x: 10.973530<br />~Protein: ~<br />~Protein: TNC","~Sample: P3G2<br />~x: 10.022583<br />~Protein: ~<br />~Protein: TNC","~Sample: P3G4<br />~x:  9.847223<br />~Protein: ~<br />~Protein: TNC","~Sample: P3G7<br />~x:        NA<br />~Protein: ~<br />~Protein: TNC","~Sample: P3H2<br />~x:        NA<br />~Protein: ~<br />~Protein: TNC","~Sample: P3H4<br />~x:        NA<br />~Protein: ~<br />~Protein: TNC","~Sample: P3H6<br />~x:  9.032600<br />~Protein: ~<br />~Protein: TNC"],"type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(248,118,109,1)","dash":"solid"},"hoveron":"points","name":"TNC","legendgroup":"TNC","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20],"y":[12.066835,12.836885,12.311633,10.677644,14.504585,11.529156,11.57881,13.011477,11.298589,12.620521,11.008346,11.515896,11.950908,12.780903,10.518185,11.878245,11.997913,12.437102,10.90237,11.890238],"text":["~Sample: P1A10<br />~x: 12.066835<br />~Protein: NA<br />~Protein: VTN","~Sample: P1A2<br />~x: 12.836885<br />~Protein: NA<br />~Protein: VTN","~Sample: P1A4<br />~x: 12.311633<br />~Protein: NA<br />~Protein: VTN","~Sample: P1A6<br />~x: 10.677644<br />~Protein: NA<br />~Protein: VTN","~Sample: P1B12<br />~x: 14.504585<br />~Protein: NA<br />~Protein: VTN","~Sample: P1B2<br />~x: 11.529156<br />~Protein: NA<br />~Protein: VTN","~Sample: P1B3<br />~x: 11.578810<br />~Protein: NA<br />~Protein: VTN","~Sample: P1B6<br />~x: 13.011477<br />~Protein: NA<br />~Protein: VTN","~Sample: P1B9<br />~x: 11.298589<br />~Protein: NA<br />~Protein: VTN","~Sample: P1C11<br />~x: 12.620521<br />~Protein: NA<br />~Protein: VTN","~Sample: P3F3<br />~x: 11.008346<br />~Protein: NA<br />~Protein: VTN","~Sample: P3F5<br />~x: 11.515896<br />~Protein: NA<br />~Protein: VTN","~Sample: P3F7<br />~x: 11.950908<br />~Protein: NA<br />~Protein: VTN","~Sample: P3G10<br />~x: 12.780903<br />~Protein: NA<br />~Protein: VTN","~Sample: P3G2<br />~x: 10.518185<br />~Protein: NA<br />~Protein: VTN","~Sample: P3G4<br />~x: 11.878245<br />~Protein: NA<br />~Protein: VTN","~Sample: P3G7<br />~x: 11.997913<br />~Protein: NA<br />~Protein: VTN","~Sample: P3H2<br />~x: 12.437102<br />~Protein: NA<br />~Protein: VTN","~Sample: P3H4<br />~x: 10.902370<br />~Protein: NA<br />~Protein: VTN","~Sample: P3H6<br />~x: 11.890238<br />~Protein: NA<br />~Protein: VTN"],"type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(0,186,56,1)","dash":"solid"},"hoveron":"points","name":"VTN","legendgroup":"VTN","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20],"y":[10.661031,10.778432,11.185271,10.887855,11.254175,11.035014,11.710207,11.759593,10.824279,11.121305,10.217992,9.773403,9.489447,10.965852,10.222464,11.062184,10.250246,9.410451,10.700179,9.933152],"text":["~Sample: P1A10<br />~x: 10.661031<br />~Protein: NA<br />~Protein: VWF","~Sample: P1A2<br />~x: 10.778432<br />~Protein: NA<br />~Protein: VWF","~Sample: P1A4<br />~x: 11.185271<br />~Protein: NA<br />~Protein: VWF","~Sample: P1A6<br />~x: 10.887855<br />~Protein: NA<br />~Protein: VWF","~Sample: P1B12<br />~x: 11.254175<br />~Protein: NA<br />~Protein: VWF","~Sample: P1B2<br />~x: 11.035014<br />~Protein: NA<br />~Protein: VWF","~Sample: P1B3<br />~x: 11.710207<br />~Protein: NA<br />~Protein: VWF","~Sample: P1B6<br />~x: 11.759593<br />~Protein: NA<br />~Protein: VWF","~Sample: P1B9<br />~x: 10.824279<br />~Protein: NA<br />~Protein: VWF","~Sample: P1C11<br />~x: 11.121305<br />~Protein: NA<br />~Protein: VWF","~Sample: P3F3<br />~x: 10.217992<br />~Protein: NA<br />~Protein: VWF","~Sample: P3F5<br />~x:  9.773403<br />~Protein: NA<br />~Protein: VWF","~Sample: P3F7<br />~x:  9.489447<br />~Protein: NA<br />~Protein: VWF","~Sample: P3G10<br />~x: 10.965852<br />~Protein: NA<br />~Protein: VWF","~Sample: P3G2<br />~x: 10.222464<br />~Protein: NA<br />~Protein: VWF","~Sample: P3G4<br />~x: 11.062184<br />~Protein: NA<br />~Protein: VWF","~Sample: P3G7<br />~x: 10.250246<br />~Protein: NA<br />~Protein: VWF","~Sample: P3H2<br />~x:  9.410451<br />~Protein: NA<br />~Protein: VWF","~Sample: P3H4<br />~x: 10.700179<br />~Protein: NA<br />~Protein: VWF","~Sample: P3H6<br />~x:  9.933152<br />~Protein: NA<br />~Protein: VWF"],"type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(97,156,255,1)","dash":"solid"},"hoveron":"points","name":"VWF","legendgroup":"VWF","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":26.2283105022831,"r":7.30593607305936,"b":40.1826484018265,"l":37.2602739726027},"plot_bgcolor":"rgba(235,235,235,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xaxis":{"domain":[0,1],"type":"linear","autorange":false,"range":[0.4,20.6],"tickmode":"array","ticktext":["P1A10","P1A2","P1A4","P1A6","P1B12","P1B2","P1B3","P1B6","P1B9","P1C11","P3F3","P3F5","P3F7","P3G10","P3G2","P3G4","P3G7","P3H2","P3H4","P3H6"],"tickvals":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20],"categoryorder":"array","categoryarray":["P1A10","P1A2","P1A4","P1A6","P1B12","P1B2","P1B3","P1B6","P1B9","P1C11","P3F3","P3F5","P3F7","P3G10","P3G2","P3G4","P3G7","P3H2","P3H4","P3H6"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"y","title":"Sample","titlefont":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"type":"linear","autorange":false,"range":[8.56923005,14.78722095],"tickmode":"array","ticktext":["9","10","11","12","13","14"],"tickvals":[9,10,11,12,13,14],"categoryorder":"array","categoryarray":["9","10","11","12","13","14"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"x","title":"x","titlefont":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":true,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"y":0.913385826771654},"annotations":[{"text":"Protein","x":1.02,"y":1,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"left","yanchor":"bottom","legendTitle":true}],"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","modeBarButtonsToAdd":[{"name":"Collaborate","icon":{"width":1000,"ascent":500,"descent":-50,"path":"M487 375c7-10 9-23 5-36l-79-259c-3-12-11-23-22-31-11-8-22-12-35-12l-263 0c-15 0-29 5-43 15-13 10-23 23-28 37-5 13-5 25-1 37 0 0 0 3 1 7 1 5 1 8 1 11 0 2 0 4-1 6 0 3-1 5-1 6 1 2 2 4 3 6 1 2 2 4 4 6 2 3 4 5 5 7 5 7 9 16 13 26 4 10 7 19 9 26 0 2 0 5 0 9-1 4-1 6 0 8 0 2 2 5 4 8 3 3 5 5 5 7 4 6 8 15 12 26 4 11 7 19 7 26 1 1 0 4 0 9-1 4-1 7 0 8 1 2 3 5 6 8 4 4 6 6 6 7 4 5 8 13 13 24 4 11 7 20 7 28 1 1 0 4 0 7-1 3-1 6-1 7 0 2 1 4 3 6 1 1 3 4 5 6 2 3 3 5 5 6 1 2 3 5 4 9 2 3 3 7 5 10 1 3 2 6 4 10 2 4 4 7 6 9 2 3 4 5 7 7 3 2 7 3 11 3 3 0 8 0 13-1l0-1c7 2 12 2 14 2l218 0c14 0 25-5 32-16 8-10 10-23 6-37l-79-259c-7-22-13-37-20-43-7-7-19-10-37-10l-248 0c-5 0-9-2-11-5-2-3-2-7 0-12 4-13 18-20 41-20l264 0c5 0 10 2 16 5 5 3 8 6 10 11l85 282c2 5 2 10 2 17 7-3 13-7 17-13z m-304 0c-1-3-1-5 0-7 1-1 3-2 6-2l174 0c2 0 4 1 7 2 2 2 4 4 5 7l6 18c0 3 0 5-1 7-1 1-3 2-6 2l-173 0c-3 0-5-1-8-2-2-2-4-4-4-7z m-24-73c-1-3-1-5 0-7 2-2 3-2 6-2l174 0c2 0 5 0 7 2 3 2 4 4 5 7l6 18c1 2 0 5-1 6-1 2-3 3-5 3l-174 0c-3 0-5-1-7-3-3-1-4-4-5-6z"},"click":"function(gd) { \n        // is this being viewed in RStudio?\n        if (location.search == '?viewer_pane=1') {\n          alert('To learn about plotly for collaboration, visit:\\n https://cpsievert.github.io/plotly_book/plot-ly-for-collaboration.html');\n        } else {\n          window.open('https://cpsievert.github.io/plotly_book/plot-ly-for-collaboration.html', '_blank');\n        }\n      }"}],"cloud":false},"source":"A","attrs":{"6ff63d8cae5c":{"x":{},"y":{},"colour":{},"type":"scatter"}},"cur_data":"6ff63d8cae5c","visdat":{"6ff63d8cae5c":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1}},"base_url":"https://plot.ly"},"evals":["config.modeBarButtonsToAdd.0.click"],"jsHooks":{"render":[{"code":"function(el, x) { var ctConfig = crosstalk.var('plotlyCrosstalkOpts').set({\"on\":\"plotly_click\",\"persistent\":false,\"dynamic\":false,\"selectize\":false,\"opacityDim\":0.2,\"selected\":{\"opacity\":1}}); }","data":null}]}}</script>
<p>See the <a href="https://plotly-book.cpsievert.me/">plotly for R</a> book for more details.</p>
</div>
<div id="interactive-apps" class="section level2">
<h2><span class="header-section-number">8.3</span> Interactive apps</h2>
</div>
<div id="motivating-examples" class="section level2 unnumbered">
<h2>Motivating examples</h2>
<p>In this section, we show how to build interactive <code>shiny</code> applications. <code>shiny</code> is widely used to explore and visualise biomolecular data. As motivating examples, we present here two such example for proteomics data.</p>
<div id="longitudinal-proteomics-qc" class="section level3 unnumbered">
<h3>Longitudinal proteomics QC</h3>
<p><em><a href="http://bioconductor.org/packages/MSstatsQC">MSstatsQC</a></em> uses control charts to monitor the instrument performance by tracking system suitability metrics including total peak area, retention time and full width at half maximum (FWHM) and peak assymetry. Additional metrics can also be analyzed by including them to the input file.</p>
<p>An online application is available at <a href="https://eralpdogu.shinyapps.io/msstatsqc" class="uri">https://eralpdogu.shinyapps.io/msstatsqc</a></p>
</div>
<div id="spatial-proteomics" class="section level3 unnumbered">
<h3>Spatial proteomics</h3>
<p>The <em><a href="http://bioconductor.org/packages/pRoloc">pRoloc</a></em> package offers dedicated functionality to manipulate, annotate, analyse and visualise spatial proteomics data. Its companion package <em><a href="http://bioconductor.org/packages/pRolocGUI">pRolocGUI</a></em> offers support for interactive visualisation.</p>
<p>You can start an app to visualise one of the published data as shown below</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;pRoloc&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;pRolocdata&quot;</span>)
<span class="kw">data</span>(hyperLOPIT2015)
<span class="kw">library</span>(<span class="st">&quot;pRolocGUI&quot;</span>)
<span class="kw">pRolocVis</span>(hyperLOPIT2015)</code></pre></div>
<p>This same data is also available as a standalone on-line application at <a href="https://lgatto.shinyapps.io/christoforou2015" class="uri">https://lgatto.shinyapps.io/christoforou2015</a></p>
<p>Reference:</p>
<ul>
<li><p>Gatto L, Breckels LM, Wieczorek S, Burger T, Lilley KS. Mass-spectrometry-based spatial proteomics data analysis using pRoloc and pRolocdata. Bioinformatics. 2014 May 1;30(9):1322-4. doi: <a href="https://doi.org/10.1093/bioinformatics/btu013">10.1093/bioinformatics/btu013</a>. Epub 2014 Jan 11. PubMed PMID: 24413670; PubMed Central PMCID: PMC3998135.</p></li>
<li><p>Breckels LM, Mulvey CM, Lilley KS and Gatto L. A Bioconductor workflow for processing and analysing spatial proteomics data [version 1; referees: 1 approved, 1 approved with reservations]. F1000Research 2016, 5:2926 (doi: <a href="https://f1000research.com/articles/5-2926/">10.12688/f1000research.10411.1</a>)</p></li>
</ul>
</div>
<div id="introduction-2" class="section level3 unnumbered">
<h3>Introduction</h3>
<p>A useful <code>shiny</code> cheet sheet is available <a href="https://github.com/rstudio/cheatsheets/raw/master/shiny.pdf">here</a>.</p>
<blockquote>
<p>This section is based on RStudio <code>shiny</code> tutorials.</p>
</blockquote>
<p>From the <a href="https://shiny.rstudio.com/"><code>shiny</code></a> package <a href="https://shiny.rstudio.com/">website</a>:</p>
<blockquote>
<p>Shiny is an R package that makes it easy to build interactive web apps straight from R.</p>
</blockquote>
<p>When using <a href="https://shiny.rstudio.com/"><code>shiny</code></a>, one tends to aim for more complete, long-lasting applications, rather then <em>simply</em> and transient visualisations.</p>
<p>A <a href="https://shiny.rstudio.com/"><code>shiny</code></a> application is composed of a <em>ui</em> (user interface) and a <em>server</em> that exchange information using a programming paradigm called <em>reactive programming</em>: changes performed by the user to the <em>ui</em> trigger a reaction by the <em>server</em> and the output is updated accordingly.</p>
<ul>
<li><p>In the ui: define the components of the user interface (such as page layout, page title, input options and outputs), i.e what the user will see and interact with.</p></li>
<li><p>In the server: defines the computations in the R backend.</p></li>
<li><p>The <em>reactive</em> programming is implemented through <em>reactive</em> functions, which are functions that are only called when their respective inputs are changed.</p></li>
<li><p>An application is run with the <code>shiny::runApp()</code> function, that takes the directory containing the two files as input.</p></li>
</ul>
<p>Before looking at the details of such an architecture, let’s build a simple example from scratch, step by step. This app, shown below, uses the <code>faithful</code> data, describing the wainting time between eruptions and the duration of the reuption for the Old Faithful geyser in Yellowstone National Park, Wyoming, USA.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(faithful)</code></pre></div>
<pre><code>##   eruptions waiting
## 1     3.600      79
## 2     1.800      54
## 3     3.333      74
## 4     2.283      62
## 5     4.533      85
## 6     2.883      55</code></pre>
<p>It shows the distribution of waiting times along a histogram (produced by the <code>hist</code> function) and provides a slider to adjust the number of bins (the <code>breaks</code> argument to <code>hist</code>).</p>
<iframe src="https://lgatto.shinyapps.io/shiny-app1/?showcase=0" width="672" height="650px">
</iframe>
<p>The app can also be opened at <a href="https://lgatto.shinyapps.io/shiny-app1/" class="uri">https://lgatto.shinyapps.io/shiny-app1/</a></p>
</div>
<div id="creation-of-our-fist-shiny-app" class="section level3 unnumbered">
<h3>Creation of our fist shiny app</h3>
<ol style="list-style-type: decimal">
<li>Create a directory that will contain the app, such as for example <code>&quot;shinyapp&quot;</code>.</li>
<li>In this directory, create the <em>ui</em> and <em>server</em> files, named <code>ui.R</code> and <code>server.R</code>.</li>
<li>In the <code>ui.R</code> file, let’s defines a (fluid) page containing</li>
</ol>
<ul>
<li>a title panel with a page title;</li>
<li>a layout containing a sidebar and a main panel</li>
</ul>
<pre><code>library(shiny)

shinyUI(fluidPage(
    titlePanel(&quot;My Shiny App&quot;),
    sidebarLayout(
        sidebarPanel(
        ),
        mainPanel(
        )
    )
))</code></pre>
<ol start="4" style="list-style-type: decimal">
<li>In the <code>server.R</code> file, we define the <code>shinyServer</code> function that handles input and ouputs (none at this stage) and the R logic.</li>
</ol>
<pre><code>library(shiny)

shinyServer(function(input, output) {
})</code></pre>
<ol start="5" style="list-style-type: decimal">
<li>Let’s now add some items to the <em>ui</em>: a text input widget in the sidebar and a field to hold the text ouput.</li>
</ol>
<pre><code>library(shiny)

shinyUI(fluidPage(
    titlePanel(&quot;My Shiny App&quot;),
    sidebarLayout(
        sidebarPanel(
            textInput(&quot;textInput&quot;, &quot;Enter text here:&quot;)
        ),
        mainPanel(
            textOutput(&quot;textOutput&quot;)
        )
    )
))</code></pre>
<ol start="6" style="list-style-type: decimal">
<li>In the <code>server.R</code> file, we add in the <code>shinyServer</code> function some R code defining how to manipulate the user-provided text and render it using a shiny <code>textOuput</code>.</li>
</ol>
<pre><code>library(shiny)

shinyServer(function(input, output) {
    output$textOutput &lt;- renderText(paste(&quot;User-entered text: &quot;,
                                          input$textInput))
})</code></pre>
<ol start="7" style="list-style-type: decimal">
<li>Let’s now add a plot in the main panel in <code>ui.R</code> and some code to draw a histogram in <code>server.R</code>:</li>
</ol>
<pre><code>library(shiny)

shinyUI(fluidPage(
    titlePanel(&quot;My Shiny App&quot;),
    sidebarLayout(
        sidebarPanel(
            textInput(&quot;textInput&quot;, &quot;Enter text here:&quot;)
        ),
        mainPanel(
            textOutput(&quot;textOutput&quot;),
            plotOutput(&quot;distPlot&quot;)            
        )
    )
))</code></pre>
<pre><code>library(shiny)

shinyServer(function(input, output) {
    output$textOutput &lt;- renderText(paste(&quot;User-entered text: &quot;,
                                          input$textInput))
    output$distPlot &lt;- renderPlot({
        x    &lt;- faithful[, 2]
        hist(x)
    })
    
})</code></pre>
<ol start="8" style="list-style-type: decimal">
<li>We want to be able to control the number of breaks used to plot the histograms. We first add a <code>sliderInput</code> to the <em>ui</em> for the user to specify the number of bins, and then make use of that new input to parametrise the histogram.</li>
</ol>
<pre><code>library(shiny)

shinyUI(fluidPage(
    titlePanel(&quot;My Shiny App&quot;),
    sidebarLayout(
        sidebarPanel(
            textInput(&quot;textInput&quot;, &quot;Enter text here:&quot;),
            sliderInput(&quot;bins&quot;,
                        &quot;Number of bins:&quot;,
                        min = 1,
                        max = 50,
                        value = 30)            
        ),
        mainPanel(
            textOutput(&quot;textOutput&quot;),
            plotOutput(&quot;distPlot&quot;)            
        )
    )
))</code></pre>
<pre><code>library(shiny)

shinyServer(function(input, output) {
    output$textOutput &lt;- renderText(paste(&quot;User-entered text: &quot;,
                                          input$textInput))
    output$distPlot &lt;- renderPlot({
        x    &lt;- faithful[, 2]
        bins &lt;- seq(min(x), max(x), length.out = input$bins + 1)    
        hist(x, breaks = bins)
    })
    
})</code></pre>
<ol start="9" style="list-style-type: decimal">
<li>The next addition is to add a menu for the user to choose a set of predefined colours (that would be a <code>selectInput</code>) in the <code>ui.R</code> file and use that new input to parametrise the colour of the histogramme in the <code>server.R</code> file.</li>
</ol>
<pre><code>library(shiny)

shinyUI(fluidPage(
    titlePanel(&quot;My Shiny App&quot;),
    sidebarLayout(
        sidebarPanel(
            textInput(&quot;textInput&quot;, &quot;Enter text here:&quot;),
            sliderInput(&quot;bins&quot;,
                        &quot;Number of bins:&quot;,
                        min = 1,
                        max = 50,
                        value = 30),
            selectInput(&quot;col&quot;, &quot;Select a colour:&quot;,
                        choices = c(&quot;steelblue&quot;, &quot;darkgray&quot;, &quot;orange&quot;))            
        ),
        mainPanel(
            textOutput(&quot;textOutput&quot;),
            plotOutput(&quot;distPlot&quot;)
        )
    )
))</code></pre>
<pre><code>library(shiny)

shinyServer(function(input, output) {
    output$textOutput &lt;- renderText(paste(&quot;User-entered text: &quot;,
                                          input$textInput))
    output$distPlot &lt;- renderPlot({
        x    &lt;- faithful[, 2]
        bins &lt;- seq(min(x), max(x), length.out = input$bins + 1)
        hist(x, breaks = bins, col = input$col)
    })
    
})</code></pre>
<ol start="10" style="list-style-type: decimal">
<li>The last addition that we want is to visualise the actual data in the main panel. We add a <code>dataTableOutput</code> in <code>ui.R</code> and generate that table in <code>server.R</code> using a <code>renderDataTable</code> rendering function.</li>
</ol>
<pre><code>library(shiny)
## Define UI for application that draws a histogram
shinyUI(fluidPage(
    ## Application title
    titlePanel(&quot;My Shiny App&quot;),
    ## Sidebar with text, slide bar and menu selection inputs
    sidebarLayout(
        sidebarPanel(
            textInput(&quot;textInput&quot;, &quot;Enter text here:&quot;),
            sliderInput(&quot;bins&quot;,
                        &quot;Number of bins:&quot;,
                        min = 1,
                        max = 50,
                        value = 30),
            selectInput(&quot;col&quot;, &quot;Select a colour:&quot;,
                        choices = c(&quot;steelblue&quot;, &quot;darkgray&quot;, &quot;orange&quot;))
        ),
        ## Main panel showing user-entered text, a reactive plot and a
        ## dynamic table
        mainPanel(
            textOutput(&quot;textOutput&quot;),
            plotOutput(&quot;distPlot&quot;),
            dataTableOutput(&quot;dataTable&quot;)
        )
    )
))</code></pre>
<pre><code>library(shiny)
## Define server logic 
shinyServer(function(input, output) {
    output$textOutput &lt;- renderText(paste(&quot;User-entered text: &quot;,
                                          input$textInput))
    ## Expression that generates a histogram. The expression is
    ## wrapped in a call to renderPlot to indicate that:
    ##
    ##  1) It is &quot;reactive&quot; and therefore should be automatically
    ##     re-executed when inputs change
    ##  2) Its output type is a plot
    output$distPlot &lt;- renderPlot({
        x    &lt;- faithful[, 2]  ## Old Faithful Geyser data
        bins &lt;- seq(min(x), max(x), length.out = input$bins + 1)
        ## draw the histogram with the specified number of bins
        hist(x, breaks = bins, col = input$col, border = &#39;white&#39;)
    })
    output$dataTable &lt;- renderDataTable(faithful)
})</code></pre>
<blockquote>
<p>Challenge</p>
<p>Write and your <code>shinyapp</code> applications, as described above.</p>
</blockquote>
</div>
<div id="single-file-app" class="section level3 unnumbered">
<h3>Single-file app</h3>
<p>Instead of defining the <em>ui</em> and <em>server</em> in their respective files, they can be combined into list to be passed directly to <code>runApp</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ui &lt;-<span class="st"> </span><span class="kw">fluidPage</span>(...)
server &lt;-<span class="st"> </span><span class="cf">function</span>(input, output) { ... }

app &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">ui =</span> ui, <span class="dt">server =</span> server)
<span class="kw">runApp</span>(app)</code></pre></div>
<blockquote>
<p>Challenges</p>
<ol style="list-style-type: decimal">
<li><p>Create an app to visualise the <code>mulvey2015norm</code> data where the user can select along which features to view the data.</p></li>
<li><p>As above, where the visualisation is a PCA plot and the user chooses the PCs.</p></li>
</ol>
</blockquote>
</div>
<div id="theres-more-to-shiny" class="section level3 unnumbered">
<h3>There’s more to <code>shiny</code></h3>
<div id="sharing-shiny-apps" class="section level4 unnumbered">
<h4>Sharing <code>shiny</code> apps</h4>
<ul>
<li>Share the code file(s) and <code>runApp</code></li>
<li><code>runUrl</code></li>
<li><code>runGitHub</code></li>
<li><code>runGist</code></li>
<li><a href="http://wwwshinyapps.io">shinyapps</a></li>
<li>Shiny server (in-house)</li>
</ul>
</div>
<div id="more-interactivity" class="section level4 unnumbered">
<h4>More interactivity</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"> <span class="kw">plotOutput</span>(<span class="st">&quot;pca&quot;</span>,
            <span class="dt">hover =</span> <span class="st">&quot;hover&quot;</span>,
            <span class="dt">click =</span> <span class="st">&quot;click&quot;</span>,
            <span class="dt">dblclick =</span> <span class="st">&quot;dblClick&quot;</span>,
            <span class="dt">brush =</span> <span class="kw">brushOpts</span>(
                <span class="dt">id =</span> <span class="st">&quot;brush&quot;</span>,
                <span class="dt">resetOnNew =</span> <span class="ot">TRUE</span>))</code></pre></div>
<p>Example <a href="http://shiny.rstudio.com/gallery/plot-interaction-advanced.html">here</a>.</p>
</div>
<div id="references-1" class="section level4 unnumbered">
<h4>References</h4>
<ul>
<li><a href="http://shiny.rstudio.com/"><code>shiny</code> page</a></li>
<li><a href="https://github.com/rstudio/cheatsheets/raw/master/shiny.pdf"><code>shiny</code> cheat sheet</a></li>
</ul>

</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="using-r-and-bioconductor-for-ms-based-proteomics.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="session-information.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
