<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Visualisation of biomolecular data</title>
  <meta name="description" content="Visualisation of biomolecular data">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Visualisation of biomolecular data" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Visualisation of biomolecular data" />
  <meta name="github-repo" content="lgatto/VisualisingBiomolecularData" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Visualisation of biomolecular data" />
  
  <meta name="twitter:description" content="Visualisation of biomolecular data" />
  

<meta name="author" content="Laurent Gatto">


<meta name="date" content="2018-05-06">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="plotting-in-r.html">
<link rel="next" href="using-r-and-bioconductor-for-ms-based-proteomics.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i><b>1.1</b> Installation</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#questions"><i class="fa fa-check"></i><b>1.2</b> Questions</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.3</b> License</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#useful-references"><i class="fa fa-check"></i><b>1.4</b> Useful references</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html"><i class="fa fa-check"></i><b>2</b> Example datasets and data manipulation</a><ul>
<li class="chapter" data-level="2.1" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html#raw-ms-data"><i class="fa fa-check"></i><b>2.1</b> Raw MS data</a></li>
<li class="chapter" data-level="2.2" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html#time-course-from-mulvey-et-al.-2015"><i class="fa fa-check"></i><b>2.2</b> Time course from Mulvey <em>et al.</em> 2015</a></li>
<li class="chapter" data-level="2.3" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html#the-iprg-data"><i class="fa fa-check"></i><b>2.3</b> The iPRG data</a></li>
<li class="chapter" data-level="2.4" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html#crc-training-data"><i class="fa fa-check"></i><b>2.4</b> CRC training data</a></li>
<li class="chapter" data-level="2.5" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html#all-data"><i class="fa fa-check"></i><b>2.5</b> ALL data</a></li>
<li class="chapter" data-level="2.6" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html#spatial-proteomics-data"><i class="fa fa-check"></i><b>2.6</b> Spatial proteomics data</a></li>
<li class="chapter" data-level="2.7" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html#using-dplyr-for-data-manipulation"><i class="fa fa-check"></i><b>2.7</b> Using <code>dplyr</code> for data manipulation</a><ul>
<li class="chapter" data-level="2.7.1" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html#selecting-columns-and-filtering-rows"><i class="fa fa-check"></i><b>2.7.1</b> Selecting columns and filtering rows</a></li>
<li class="chapter" data-level="2.7.2" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html#pipes"><i class="fa fa-check"></i><b>2.7.2</b> Pipes</a></li>
<li class="chapter" data-level="2.7.3" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html#mutate"><i class="fa fa-check"></i><b>2.7.3</b> Mutate</a></li>
<li class="chapter" data-level="2.7.4" data-path="example-datasets-and-data-manipulation.html"><a href="example-datasets-and-data-manipulation.html#split-apply-combine-data-analysis-and-the-summarize-function"><i class="fa fa-check"></i><b>2.7.4</b> Split-apply-combine data analysis and the summarize() function</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="exploring-and-visualising-biomolecular-data.html"><a href="exploring-and-visualising-biomolecular-data.html"><i class="fa fa-check"></i><b>3</b> Exploring and visualising biomolecular data</a></li>
<li class="chapter" data-level="4" data-path="plotting-in-r.html"><a href="plotting-in-r.html"><i class="fa fa-check"></i><b>4</b> Plotting in R</a><ul>
<li class="chapter" data-level="4.1" data-path="plotting-in-r.html"><a href="plotting-in-r.html#base-graphics"><i class="fa fa-check"></i><b>4.1</b> Base graphics</a></li>
<li class="chapter" data-level="4.2" data-path="plotting-in-r.html"><a href="plotting-in-r.html#plotting-with-ggplot2"><i class="fa fa-check"></i><b>4.2</b> Plotting with <code>ggplot2</code></a></li>
<li class="chapter" data-level="4.3" data-path="plotting-in-r.html"><a href="plotting-in-r.html#building-complex-plots"><i class="fa fa-check"></i><b>4.3</b> Building complex plots</a></li>
<li class="chapter" data-level="4.4" data-path="plotting-in-r.html"><a href="plotting-in-r.html#customising-plots"><i class="fa fa-check"></i><b>4.4</b> Customising plots</a></li>
<li class="chapter" data-level="4.5" data-path="plotting-in-r.html"><a href="plotting-in-r.html#combine-ggplot-figures"><i class="fa fa-check"></i><b>4.5</b> Combine <code>ggplot</code> figures</a></li>
<li class="chapter" data-level="4.6" data-path="plotting-in-r.html"><a href="plotting-in-r.html#saving-your-figures"><i class="fa fa-check"></i><b>4.6</b> Saving your figures</a></li>
<li class="chapter" data-level="4.7" data-path="plotting-in-r.html"><a href="plotting-in-r.html#references"><i class="fa fa-check"></i><b>4.7</b> References</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="tools-and-plots.html"><a href="tools-and-plots.html"><i class="fa fa-check"></i><b>5</b> Tools and plots</a><ul>
<li class="chapter" data-level="5.1" data-path="tools-and-plots.html"><a href="tools-and-plots.html#transformations"><i class="fa fa-check"></i><b>5.1</b> Transformations</a><ul>
<li class="chapter" data-level="5.1.1" data-path="tools-and-plots.html"><a href="tools-and-plots.html#expression-data"><i class="fa fa-check"></i><b>5.1.1</b> Expression data</a></li>
<li class="chapter" data-level="5.1.2" data-path="tools-and-plots.html"><a href="tools-and-plots.html#fold-changes"><i class="fa fa-check"></i><b>5.1.2</b> Fold changes</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="tools-and-plots.html"><a href="tools-and-plots.html#comparing-samples-and-linear-models"><i class="fa fa-check"></i><b>5.2</b> Comparing samples and linear models</a></li>
<li class="chapter" data-level="5.3" data-path="tools-and-plots.html"><a href="tools-and-plots.html#plots-for-statistical-analyses"><i class="fa fa-check"></i><b>5.3</b> Plots for statistical analyses</a></li>
<li class="chapter" data-level="5.4" data-path="tools-and-plots.html"><a href="tools-and-plots.html#visualising-intersections"><i class="fa fa-check"></i><b>5.4</b> Visualising intersections</a></li>
<li class="chapter" data-level="5.5" data-path="tools-and-plots.html"><a href="tools-and-plots.html#unsupervised-learning"><i class="fa fa-check"></i><b>5.5</b> Unsupervised learning</a></li>
<li class="chapter" data-level="5.6" data-path="tools-and-plots.html"><a href="tools-and-plots.html#hierarchical-clustering"><i class="fa fa-check"></i><b>5.6</b> Hierarchical clustering</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#how-does-hierarchical-clustering-work"><i class="fa fa-check"></i>How does hierarchical clustering work</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#defining-clusters"><i class="fa fa-check"></i>Defining clusters</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#effect-of-distance-and-clustering-algorithm"><i class="fa fa-check"></i>Effect of distance and clustering algorithm</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="tools-and-plots.html"><a href="tools-and-plots.html#k-means-clustering"><i class="fa fa-check"></i><b>5.7</b> k-means clustering</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#how-does-k-means-work"><i class="fa fa-check"></i>How does k-means work</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#model-selection"><i class="fa fa-check"></i>Model selection</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#how-to-determine-the-number-of-clusters"><i class="fa fa-check"></i>How to determine the number of clusters</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#challenge"><i class="fa fa-check"></i>Challenge</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="tools-and-plots.html"><a href="tools-and-plots.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>5.8</b> Principal component analysis (PCA)</a><ul>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#how-does-it-work"><i class="fa fa-check"></i>How does it work</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#visualisation"><i class="fa fa-check"></i>Visualisation</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#data-pre-processing"><i class="fa fa-check"></i>Data pre-processing</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#final-comments-on-pca"><i class="fa fa-check"></i>Final comments on PCA</a></li>
<li class="chapter" data-level="" data-path="tools-and-plots.html"><a href="tools-and-plots.html#challenge-1"><i class="fa fa-check"></i>Challenge</a></li>
</ul></li>
<li class="chapter" data-level="5.9" data-path="tools-and-plots.html"><a href="tools-and-plots.html#t-distributed-stochastic-neighbour-embedding"><i class="fa fa-check"></i><b>5.9</b> t-Distributed Stochastic Neighbour Embedding</a><ul>
<li class="chapter" data-level="5.9.1" data-path="tools-and-plots.html"><a href="tools-and-plots.html#parameter-tuning"><i class="fa fa-check"></i><b>5.9.1</b> Parameter tuning</a></li>
</ul></li>
<li class="chapter" data-level="5.10" data-path="tools-and-plots.html"><a href="tools-and-plots.html#pca-loadings-vs-differential-expression"><i class="fa fa-check"></i><b>5.10</b> PCA loadings vs differential expression</a></li>
<li class="chapter" data-level="5.11" data-path="tools-and-plots.html"><a href="tools-and-plots.html#heatmaps"><i class="fa fa-check"></i><b>5.11</b> Heatmaps</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html"><i class="fa fa-check"></i><b>6</b> Using R and Bioconductor for MS-based proteomics</a><ul>
<li class="chapter" data-level="6.1" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#introduction-1"><i class="fa fa-check"></i><b>6.1</b> Introduction</a></li>
<li class="chapter" data-level="6.2" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#reading-ms-data-into-r"><i class="fa fa-check"></i><b>6.2</b> Reading MS data into R</a><ul>
<li class="chapter" data-level="" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#raw-data"><i class="fa fa-check"></i>Raw data</a></li>
<li class="chapter" data-level="" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#identification-data"><i class="fa fa-check"></i>Identification data</a></li>
<li class="chapter" data-level="" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#quantitative-data"><i class="fa fa-check"></i>Quantitative data</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#manipulating-and-visualising-raw-data"><i class="fa fa-check"></i><b>6.3</b> Manipulating and visualising raw data</a></li>
<li class="chapter" data-level="6.4" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#ms-maps"><i class="fa fa-check"></i><b>6.4</b> MS maps</a></li>
<li class="chapter" data-level="6.5" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#peptide-spetrum-matches"><i class="fa fa-check"></i><b>6.5</b> Peptide spetrum matches</a></li>
<li class="chapter" data-level="6.6" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#ms-imaging"><i class="fa fa-check"></i><b>6.6</b> MS imaging</a></li>
<li class="chapter" data-level="6.7" data-path="using-r-and-bioconductor-for-ms-based-proteomics.html"><a href="using-r-and-bioconductor-for-ms-based-proteomics.html#missing-values"><i class="fa fa-check"></i><b>6.7</b> Missing values</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html"><i class="fa fa-check"></i><b>7</b> Interactive visualisation</a><ul>
<li class="chapter" data-level="7.1" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#interactivity-graphs"><i class="fa fa-check"></i><b>7.1</b> Interactivity graphs</a><ul>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#interactivity"><i class="fa fa-check"></i>Interactivity</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#layers"><i class="fa fa-check"></i>Layers</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#more-components"><i class="fa fa-check"></i>More components</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#interactive-apps"><i class="fa fa-check"></i><b>7.2</b> Interactive apps</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#motivating-examples"><i class="fa fa-check"></i>Motivating examples</a><ul>
<li class="chapter" data-level="7.2.1" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#longitudinal-proteomics-qc"><i class="fa fa-check"></i><b>7.2.1</b> Longitudinal proteomics QC</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#spatial-proteomics"><i class="fa fa-check"></i>Spatial proteomics</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#introduction-2"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#creation-of-our-fist-shiny-app"><i class="fa fa-check"></i>Creation of our fist shiny app</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#the-shiny-infrastructure"><i class="fa fa-check"></i>The shiny infrastructure</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#another-shiny-app"><i class="fa fa-check"></i>Another shiny app</a></li>
<li class="chapter" data-level="" data-path="interactive-visualisation.html"><a href="interactive-visualisation.html#single-file-app"><i class="fa fa-check"></i>Single-file app</a></li>
<li><a href="interactive-visualisation.html#theres-more-to-shiny">There’s more to <code>shiny</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="session-information.html"><a href="session-information.html"><i class="fa fa-check"></i><b>8</b> Session information</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Visualisation of biomolecular data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="tools-and-plots" class="section level1">
<h1><span class="header-section-number">Chapter 5</span> Tools and plots</h1>
<div id="transformations" class="section level2">
<h2><span class="header-section-number">5.1</span> Transformations</h2>
<div id="expression-data" class="section level3">
<h3><span class="header-section-number">5.1.1</span> Expression data</h3>
<p>Let’s start with the comparison of two vectors of matching expression intensities such as those from two samples in the <code>iprg3</code> dataset. Let’s extract the intensities of samples JD_06232014_sample1-A.raw (second column) and JD_06232014_sample1_B.raw (third column) and produce a scatter plot of one against the other.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span>iprg3[[<span class="dv">2</span>]]
y &lt;-<span class="st"> </span>iprg3[[<span class="dv">3</span>]]
<span class="kw">plot</span>(x, y)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/trans1-1.png" width="672" /></p>
<p>Due to the distribution of the raw intensities, where most of the intensities are low with very few high intensities (see density plots below), the majority of points are squeezed close to the origin of the scatter plot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(<span class="kw">density</span>(<span class="kw">na.omit</span>(x)), <span class="dt">col =</span> <span class="st">&quot;blue&quot;</span>)
<span class="kw">lines</span>(<span class="kw">density</span>(<span class="kw">na.omit</span>(y)), <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/denstrans1-1.png" width="672" /></p>
<p>This has negative effects as it (1) leads to overplotting in the low intensity range and (2) gives too much confidence in the correlation of the two vectors. A simple way to avoid this effect is to directly log-tranform the data or set the graph axes to log scales:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))
<span class="kw">plot</span>(<span class="kw">log10</span>(x), <span class="kw">log10</span>(y))
<span class="kw">plot</span>(x, y, <span class="dt">log =</span> <span class="st">&quot;xy&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/plotrans1log-1.png" width="672" /></p>
<p>We will see better visualisations to detect correlation between sample replicates below.</p>
<p>It is possible to generalise to production of scatter plots to more samples using the <code>pairs</code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">pairs</span>(iprg3[<span class="dv">2</span><span class="op">:</span><span class="dv">6</span>], <span class="dt">log =</span> <span class="st">&quot;xy&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/pairs-1.png" width="672" /></p>
<p>A lot of space is wasted by repeating the same sets of plots in the upper right and lower left triangles of the matrix. See the <code>pairs</code> documentation page.</p>
<p>A general technique to overcome overplotting is to set the alpha scale (transparency), of to use <code>graphics::smoothScatter</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))
<span class="kw">plot</span>(x, y, <span class="dt">pch =</span> <span class="dv">19</span>, <span class="dt">col =</span> <span class="st">&quot;#00000010&quot;</span>, <span class="dt">log =</span> <span class="st">&quot;xy&quot;</span>)
<span class="kw">smoothScatter</span>(<span class="kw">log10</span>(x), <span class="kw">log10</span>(y))</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/smoothscatter-1.png" width="672" /></p>
</div>
<div id="fold-changes" class="section level3">
<h3><span class="header-section-number">5.1.2</span> Fold changes</h3>
<p>Log-transformation also comes handy when computing fold-changes. Below we calculate the fold-changes and log2 fold-changes (omitting missing values)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fc &lt;-<span class="st"> </span><span class="kw">na.omit</span>(iprg3[[<span class="dv">2</span>]] <span class="op">/</span><span class="st"> </span>iprg3[[<span class="dv">3</span>]])
lfc &lt;-<span class="st"> </span><span class="kw">log2</span>(fc)</code></pre></div>
<p>Below, we see how the log2 fold-changes become symmetrical around zero (the absence of change), with positive values corresponding to up-regulation and negative values to down-regulation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(<span class="kw">density</span>(lfc), <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">5</span>))
<span class="kw">abline</span>(<span class="dt">v =</span> <span class="kw">median</span>(lfc))
<span class="kw">lines</span>(<span class="kw">density</span>(fc), <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>)
<span class="kw">abline</span>(<span class="dt">v =</span> <span class="kw">median</span>(fc), <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/lfcplot-1.png" width="672" /></p>
<p><strong>Note</strong>: when the data is already log-transformed, log fold-changes are computed by subtracting values.</p>
</div>
</div>
<div id="comparing-samples-and-linear-models" class="section level2">
<h2><span class="header-section-number">5.2</span> Comparing samples and linear models</h2>
<p>Let’s return to the scatter plot example above and focus on three replicates from consitions 1 and 4, remove missing values and log-tranform the intensites.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">log2</span>(<span class="kw">na.omit</span>(iprg3[, <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">11</span>)]))</code></pre></div>
<p>Below, we use the pairs function and print the pairwise correlations in the upper right traingle.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## put (absolute) correlations on the upper panels,
## with size proportional to the correlations.
## From ?pairs
panel.cor &lt;-<span class="st"> </span><span class="cf">function</span>(x, y, <span class="dt">digits =</span> <span class="dv">2</span>, <span class="dt">prefix =</span> <span class="st">&quot;&quot;</span>, cex.cor, ...) {
    usr &lt;-<span class="st"> </span><span class="kw">par</span>(<span class="st">&quot;usr&quot;</span>); <span class="kw">on.exit</span>(<span class="kw">par</span>(usr))
    <span class="kw">par</span>(<span class="dt">usr =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>))
    r &lt;-<span class="st"> </span><span class="kw">abs</span>(<span class="kw">cor</span>(x, y))
    txt &lt;-<span class="st"> </span><span class="kw">format</span>(<span class="kw">c</span>(r, <span class="fl">0.123456789</span>), <span class="dt">digits =</span> digits)[<span class="dv">1</span>]
    txt &lt;-<span class="st"> </span><span class="kw">paste0</span>(prefix, txt)
    <span class="cf">if</span> (<span class="kw">missing</span>(cex.cor)) cex.cor &lt;-<span class="st"> </span><span class="fl">0.8</span><span class="op">/</span><span class="kw">strwidth</span>(txt)
    <span class="kw">text</span>(<span class="fl">0.5</span>, <span class="fl">0.5</span>, txt, <span class="dt">cex =</span> cex.cor <span class="op">*</span><span class="st"> </span>r)
}
<span class="kw">pairs</span>(x, <span class="dt">lower.panel =</span> panel.smooth, <span class="dt">upper.panel =</span> panel.cor)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/linmodpairs-1.png" width="672" /></p>
<p>It is often assumed that high correlation is a halmark of good replication. Rather than focus on the correlation of the data, a better measurement would be to look a the log2 fold-changes, i.e. the distance between repeated measurements. The ideal way to visualise this is on an MA-plot:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))
r1 &lt;-<span class="st"> </span>x[[<span class="dv">1</span>]]
r2 &lt;-<span class="st"> </span>x[[<span class="dv">2</span>]]
M &lt;-<span class="st"> </span>r1 <span class="op">-</span><span class="st"> </span>r2
A &lt;-<span class="st"> </span>(r1 <span class="op">+</span><span class="st"> </span>r2)<span class="op">/</span><span class="dv">2</span>
<span class="kw">plot</span>(A, M); <span class="kw">grid</span>()
<span class="kw">library</span>(<span class="st">&quot;affy&quot;</span>)
affy<span class="op">::</span><span class="kw">ma.plot</span>(A, M)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/ma-1.png" width="1152" /></p>
<p>See also this <a href="http://simplystatistics.org/2015/08/12/correlation-is-not-a-measure-of-reproducibility/">post</a> on the <em>Simply Statistics</em> blog.</p>
<p><code>abline(0, 1)</code> can be used to add a line with intercept 0 and slop 1. It we want to add the line that models the data linearly, we can calculate the parameters using the <code>lm</code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">lmod &lt;-<span class="st"> </span><span class="kw">lm</span>(r2 <span class="op">~</span><span class="st"> </span>r1)
<span class="kw">summary</span>(lmod)</code></pre></div>
<pre><code>## 
## Call:
## lm(formula = r2 ~ r1)
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -3.4939 -0.0721  0.0126  0.0881  3.4595 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept) 0.348190   0.091842   3.791 0.000153 ***
## r1          0.985878   0.003688 267.357  &lt; 2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 0.3263 on 3024 degrees of freedom
## Multiple R-squared:  0.9594, Adjusted R-squared:  0.9594 
## F-statistic: 7.148e+04 on 1 and 3024 DF,  p-value: &lt; 2.2e-16</code></pre>
<p>which can be used to add the adequate line that reflects the (linear) relationship between the two data</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(r1, r2)
<span class="kw">abline</span>(lmod, <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/lmodplot-1.png" width="672" /></p>
<p>As we have seen in the beginning of this section, it is essential not to rely solely on the correlation value, but look at the data. This also holds true for linear (or any) modelling, which can be done by plotting the model:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">2</span>))
<span class="kw">plot</span>(lmod)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/lmodplot2-1.png" width="672" /></p>
<ul>
<li><p><em>Cook’s distance</em> is a commonly used estimate of the influence of a data point when performing a least-squares regression analysis and can be used to highlight points that particularly influence the regression.</p></li>
<li><p><em>Leverage</em> quantifies the influence of a given observation on the regression due to its location in the space of the inputs.</p></li>
</ul>
<p>See also <code>?influence.measures</code>.</p>
<blockquote>
<p><strong>Challenge</strong></p>
<ol style="list-style-type: decimal">
<li>Take any of the <code>iprg3</code> replicates, model and plot their linear relationship.</li>
<li>The Anscombe quartet is available as <code>anscombe</code>. Load it, create a linear model for one <span class="math inline">\((x_i, y_i)\)</span> pair of your choice and visualise/check the model.</li>
</ol>
</blockquote>
<p><details></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x3 &lt;-<span class="st"> </span>anscombe[, <span class="dv">3</span>]
y3 &lt;-<span class="st"> </span>anscombe[, <span class="dv">7</span>]
lmod &lt;-<span class="st"> </span><span class="kw">lm</span>(y3 <span class="op">~</span><span class="st"> </span>x3)
<span class="kw">summary</span>(lmod)</code></pre></div>
<pre><code>## 
## Call:
## lm(formula = y3 ~ x3)
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -1.1586 -0.6146 -0.2303  0.1540  3.2411 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(&gt;|t|)   
## (Intercept)   3.0025     1.1245   2.670  0.02562 * 
## x3            0.4997     0.1179   4.239  0.00218 **
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 1.236 on 9 degrees of freedom
## Multiple R-squared:  0.6663, Adjusted R-squared:  0.6292 
## F-statistic: 17.97 on 1 and 9 DF,  p-value: 0.002176</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">2</span>))
<span class="kw">plot</span>(lmod)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/anscombechallenge-1.png" width="672" /> </details></p>
</div>
<div id="plots-for-statistical-analyses" class="section level2">
<h2><span class="header-section-number">5.3</span> Plots for statistical analyses</h2>
<p>Let’s use the <code>ALL_bclneg</code> dataset, that we already have analysed with <em><a href="http://bioconductor.org/packages/limma">limma</a></em> in the <em>Data</em> chapter.</p>
<p>Whenever performing a statistical test, it is important to quality check the distribution of non-adjusted p-values. Below, we see an enrichment of small p-values, as opposed to a uniform distribution to be expected under the null hypothesis of absence of changes between groups.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">fvarLabels</span>(ALL_bcrneg)</code></pre></div>
<pre><code>## [1] &quot;logFC&quot;     &quot;AveExpr&quot;   &quot;t&quot;         &quot;P.Value&quot;   &quot;adj.P.Val&quot; &quot;B&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">hist</span>(<span class="kw">fData</span>(ALL_bcrneg)<span class="op">$</span>P.Value)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/histpv-1.png" width="672" /></p>
<p>The histograms below illustrate <a href="http://varianceexplained.org/statistics/interpreting-pvalue-histogram/">other distributions</a> to beware of.</p>
<div class="figure">
<img src="img/plot_melted-1.png" />

</div>
<p>Another important visualisation for statistical results are the <em>Volcano plots</em>, that show the relationship between the significance of the individual tests (adjusted p-values) and their magnitude of the effect (log2 fold-changes).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">lfc &lt;-<span class="st"> </span><span class="kw">fData</span>(ALL_bcrneg)<span class="op">$</span>logFC
bh &lt;-<span class="st"> </span><span class="kw">fData</span>(ALL_bcrneg)<span class="op">$</span>adj.P.Val
<span class="kw">plot</span>(lfc, <span class="op">-</span><span class="kw">log10</span>(bh),
     <span class="dt">main =</span> <span class="st">&quot;Volcano plot&quot;</span>,
     <span class="dt">xlab =</span> <span class="kw">expression</span>(log[<span class="dv">2</span>]<span class="op">~</span>fold<span class="op">-</span>change),
     <span class="dt">ylab =</span> <span class="kw">expression</span>(<span class="op">-</span>log[<span class="dv">10</span>]<span class="op">~</span>adjusted<span class="op">~</span>p<span class="op">-</span>value))
<span class="kw">grid</span>()</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/volc-1.png" width="672" /></p>
<p>The volcano plot can further be annotated using vertical and horizontal lines depicting thresholds of interest or points can be colour-coded based on their interest.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">lfc &lt;-<span class="st"> </span><span class="kw">fData</span>(ALL_bcrneg)<span class="op">$</span>logFC
bh &lt;-<span class="st"> </span><span class="kw">fData</span>(ALL_bcrneg)<span class="op">$</span>adj.P.Val
sign &lt;-<span class="st"> </span><span class="kw">abs</span>(lfc) <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span> <span class="op">&amp;</span><span class="st"> </span>bh <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.01</span>
<span class="kw">plot</span>(lfc, <span class="op">-</span><span class="kw">log10</span>(bh),
     <span class="dt">main =</span> <span class="st">&quot;Volcano plot&quot;</span>,
     <span class="dt">col =</span> <span class="kw">ifelse</span>(sign, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;black&quot;</span>),
     <span class="dt">pch =</span> <span class="kw">ifelse</span>(sign, <span class="dv">19</span>, <span class="dv">1</span>),
     <span class="dt">xlab =</span> <span class="kw">expression</span>(log[<span class="dv">2</span>]<span class="op">~</span>fold<span class="op">-</span>change),
     <span class="dt">ylab =</span> <span class="kw">expression</span>(<span class="op">-</span>log[<span class="dv">10</span>]<span class="op">~</span>adjusted<span class="op">~</span>p<span class="op">-</span>value))
<span class="kw">grid</span>()
<span class="kw">abline</span>(<span class="dt">v =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>), <span class="dt">lty =</span> <span class="st">&quot;dotted&quot;</span>)
<span class="kw">abline</span>(<span class="dt">h =</span> <span class="op">-</span><span class="kw">log10</span>(<span class="fl">0.05</span>), <span class="dt">lty =</span> <span class="st">&quot;dotted&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/volc2-1.png" width="672" /></p>
<p>It is also possible to identify and label individual points on the plot using the <code>identify</code> function</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">i &lt;-<span class="st"> </span><span class="kw">identify</span>(lfc, <span class="op">-</span><span class="kw">log10</span>(bh), <span class="kw">featureNames</span>(ALL_bcrneg))</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/volc3-1.png" width="672" /></p>
</div>
<div id="visualising-intersections" class="section level2">
<h2><span class="header-section-number">5.4</span> Visualising intersections</h2>
<p>Venn and Euler diagrams are popular representation when comparing sets and their intersection. Two useful R packages to generate such plots are <em><a href="https://CRAN.R-project.org/package=venneuler">venneuler</a></em> and <em><a href="https://github.com/js229/Vennerable">Vennerable</a></em>.</p>
<p>We will use the <code>crc</code> feature names to generate a test data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">123</span>)
x &lt;-<span class="st"> </span><span class="kw">replicate</span>(<span class="dv">3</span>, <span class="kw">sample</span>(<span class="kw">featureNames</span>(crc), <span class="dv">35</span>), <span class="dt">simplify =</span> <span class="ot">FALSE</span>)
<span class="kw">names</span>(x) &lt;-<span class="st"> </span>LETTERS[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]
(v &lt;-<span class="st"> </span><span class="kw">Venn</span>(x))</code></pre></div>
<pre><code>## A Venn object on 3 sets named
## A,B,C 
## 000 100 010 110 001 101 011 111 
##   0  11   9   6   5  10  12   8</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(v)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/venn-1.png" width="672" /></p>
<p>The <em><a href="https://CRAN.R-project.org/package=UpSetR">UpSetR</a></em> visualises intersections of sets as a matrix in which the rows represent the sets and the columns represent their intersection sizes. For each set that is part of a given intersection, a black filled circle is placed in the corresponding matrix cell. If a set is not part of the intersection, a light gray circle is shown. A vertical black line connects the topmost black circle with the bottom most black circle in each column to emphasise the column-based relationships. The size of the intersections is shown as a bar chart placed on top of the matrix so that each column lines up with exactly one bar. A second bar chart showing the size of the each set is shown to the left of the matrix.</p>
<p>We will first make use of the <code>fromList</code> function to convert our list to a <code>UpSetR</code> compatible input and then generate the figure:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;UpSetR&quot;</span>)
x2 &lt;-<span class="st"> </span><span class="kw">fromList</span>(x)
<span class="kw">upset</span>(x2)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/upset1-1.png" width="672" /></p>
<p>The following <a href="https://twitter.com/ngehlenborg/status/986354574989709312">tweet</a> by the author of the package illustrates how Venn and upset diagrams relate to each other.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">upset</span>(x2, <span class="dt">order.by =</span> <span class="st">&quot;freq&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/upse2-1.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">upset</span>(x2, <span class="dt">order.by =</span> <span class="st">&quot;degree&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/upse2-2.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">upset</span>(x2, <span class="dt">order.by =</span> <span class="kw">c</span>(<span class="st">&quot;freq&quot;</span>, <span class="st">&quot;degree&quot;</span>))</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/upse2-3.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">upset</span>(x2, <span class="dt">order.by =</span> <span class="kw">c</span>(<span class="st">&quot;degree&quot;</span>, <span class="st">&quot;freq&quot;</span>))</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/upse2-4.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">upset</span>(x2, <span class="dt">sets =</span> <span class="kw">c</span>(<span class="st">&quot;A&quot;</span>, <span class="st">&quot;B&quot;</span>))</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/upse2-5.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">upset</span>(x2, <span class="dt">sets =</span> <span class="kw">c</span>(<span class="st">&quot;B&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;A&quot;</span>), <span class="dt">keep.order =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/upse2-6.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">upset</span>(x2, <span class="dt">group.by =</span> <span class="st">&quot;sets&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/upse2-7.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Add set D with a single intersection
x3 &lt;-<span class="st"> </span>x2
x3<span class="op">$</span>D &lt;-<span class="st"> </span><span class="dv">0</span>
x3[<span class="dv">1</span>, <span class="st">&quot;D&quot;</span>] &lt;-<span class="st"> </span><span class="dv">1</span>
<span class="kw">head</span>(x3)</code></pre></div>
<pre><code>##   A B C D
## 1 1 0 1 1
## 2 1 0 1 0
## 3 1 0 0 0
## 4 1 0 0 0
## 5 1 0 0 0
## 6 1 0 0 0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">upset</span>(x3)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/upse2-8.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">upset</span>(x3, <span class="dt">empty.intersections =</span> <span class="st">&quot;on&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/upse2-9.png" width="672" /></p>
<p>Visualising intersections with <code>UpSetR</code> shines with more that 4 sets, as Venn diagrams become practically useless.</p>
<blockquote>
<p><strong>Challenge</strong></p>
<p>Generate a bigger dataset containing 10 sets. Try to generate Venn and upset diagrams as shown above.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">123</span>)
x &lt;-<span class="st"> </span><span class="kw">replicate</span>(<span class="dv">10</span>, <span class="kw">sample</span>(<span class="kw">featureNames</span>(crc), <span class="dv">35</span>), <span class="dt">simplify =</span> <span class="ot">FALSE</span>)
<span class="kw">names</span>(x) &lt;-<span class="st"> </span>LETTERS[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</code></pre></div>
</blockquote>
<p>When the number of sets become larger, the options above, as well as <code>nsets</code>, the number of sets (default is 5) and <code>nintersects</code>, the number of intersectios (default is 40) becomes useful.</p>
</div>
<div id="unsupervised-learning" class="section level2">
<h2><span class="header-section-number">5.5</span> Unsupervised learning</h2>
<p>In <strong>unsupervised learning</strong> (UML), no labels are provided, and the learning algorithm focuses solely on detecting structure in unlabelled input data. One generally differentiates between</p>
<ul>
<li><p><strong>Clustering</strong>, where the goal is to find homogeneous subgroups within the data; the grouping is based on distance between observations.</p></li>
<li><p><strong>Dimensionality reduction</strong>, where the goal is to identify patterns in the features of the data. Dimensionality reduction is often used to facilitate visualisation of the data, as well as a pre-processing method before supervised learning.</p></li>
</ul>
<p>UML presents specific challenges and benefits:</p>
<ul>
<li>there is no single goal in UML</li>
<li>there is generally much more unlabelled data available than labelled data.</li>
</ul>
<p>Unsupervised learning techniques are paramount for exploratory data analysis and visualisation.</p>
</div>
<div id="hierarchical-clustering" class="section level2">
<h2><span class="header-section-number">5.6</span> Hierarchical clustering</h2>
<div id="how-does-hierarchical-clustering-work" class="section level3 unnumbered">
<h3>How does hierarchical clustering work</h3>
<p><strong>Initialisation</strong>: Starts by assigning each of the n points its own cluster</p>
<p><strong>Iteration</strong></p>
<ol style="list-style-type: decimal">
<li>Find the two nearest clusters, and join them together, leading to n-1 clusters</li>
<li>Continue the cluster merging process until all are grouped into a single cluster</li>
</ol>
<p><strong>Termination:</strong> All observations are grouped within a single cluster.</p>
<div class="figure"><span id="fig:hcldata"></span>
<img src="VisualisingBiomolecularData_files/figure-html/hcldata-1.png" alt="Hierarchical clustering: initialisation (left) and colour-coded results after iteration (right)." width="1152" />
<p class="caption">
Figure 5.1: Hierarchical clustering: initialisation (left) and colour-coded results after iteration (right).
</p>
</div>
<p>The results of hierarchical clustering are typically visualised along a <strong>dendrogram</strong>, where the distance between the clusters is proportional to the branch lengths.</p>
<div class="figure"><span id="fig:hcldendro"></span>
<img src="VisualisingBiomolecularData_files/figure-html/hcldendro-1.png" alt="Visualisation of the hierarchical clustering results on a dendrogram" width="672" />
<p class="caption">
Figure 5.2: Visualisation of the hierarchical clustering results on a dendrogram
</p>
</div>
<p>In R:</p>
<ul>
<li>Calculate the distance using <code>dist</code>, typically the Euclidean distance.</li>
<li>Hierarchical clustering on this distance matrix using <code>hclust</code></li>
</ul>
<blockquote>
<p>Challenge</p>
<p>Apply hierarchical clustering on the <code>mulvey2015norm</code> data and generate a sample dendrogram using the dedicated <code>plot</code> method as shown below.</p>
</blockquote>
<p><details></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">d &lt;-<span class="st"> </span><span class="kw">dist</span>(<span class="kw">t</span>(MSnbase<span class="op">::</span><span class="kw">exprs</span>(mulvey2015norm)))
hcl &lt;-<span class="st"> </span><span class="kw">hclust</span>(d)
hcl</code></pre></div>
<pre><code>## 
## Call:
## hclust(d = d)
## 
## Cluster method   : complete 
## Distance         : euclidean 
## Number of objects: 18</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(hcl, <span class="dt">main =</span> <span class="st">&quot;Mulvey et al. 2016&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/hclsol-1.png" width="672" /> </details></p>
</div>
<div id="defining-clusters" class="section level3 unnumbered">
<h3>Defining clusters</h3>
<p>After producing the hierarchical clustering result, we need to <em>cut the tree (dendrogram)</em> at a specific height to defined the clusters. For example, on our test dataset above, we could decide to cut it at a distance around 15, that would produce 3 clusters, namely the XEN samples, the late time points 72 and 48 hours, and the early time points.</p>
<div class="figure"><span id="fig:cuthcl"></span>
<img src="VisualisingBiomolecularData_files/figure-html/cuthcl-1.png" alt="Cutting the dendrogram at height 15." width="672" />
<p class="caption">
Figure 5.3: Cutting the dendrogram at height 15.
</p>
</div>
<p>In R we can us the <code>cutree</code> function to</p>
<ul>
<li>cut the tree at a specific height: <code>cutree(hcl, h = 15)</code></li>
<li>cut the tree to get a certain number of clusters: <code>cutree(hcl, k = 3)</code></li>
</ul>
<blockquote>
<p>Challenge</p>
<ul>
<li><p>Cut the hierarchical clustering result at a height to obtain 4 clusters by setting <code>h</code>.</p></li>
<li><p>Cut the hierarchical clustering result at a height to obtain 4 clusters by setting directly <code>k</code>, and verify that both provide the same results.</p></li>
<li><p>What happens if we wanted to obtain 5 clusters.</p></li>
</ul>
</blockquote>
<p><details></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(hcl)
<span class="kw">abline</span>(<span class="dt">h =</span> <span class="dv">11</span>, <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>)
<span class="kw">cutree</span>(hcl, <span class="dt">k =</span> <span class="dv">4</span>)
<span class="kw">cutree</span>(hcl, <span class="dt">h =</span> <span class="dv">11</span>)
<span class="kw">identical</span>(<span class="kw">cutree</span>(hcl, <span class="dt">k =</span> <span class="dv">4</span>), <span class="kw">cutree</span>(hcl, <span class="dt">h =</span> <span class="dv">11</span>))
<span class="kw">plot</span>(hcl, <span class="dt">labels =</span> <span class="kw">cutree</span>(hcl, <span class="dt">k =</span> <span class="dv">4</span>))</code></pre></div>
<p></details></p>
<p>For more advanced dendrogram visualisations, see <em><a href="https://CRAN.R-project.org/package=dendextend">dendextend</a></em>.</p>
</div>
<div id="effect-of-distance-and-clustering-algorithm" class="section level3 unnumbered">
<h3>Effect of distance and clustering algorithm</h3>
<p>It is important to highlight that the type of distance and the type of clustering algorithms will affect the result. Let’s observe how using the euclidean or Pearson correlation distances between expression profiles can fundamentally change the results.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## toy example: 3 genes, 5 samples from Olga Vitek
gene1 &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">6</span>, <span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">7</span>)
gene2 &lt;-<span class="st"> </span>gene1 <span class="op">+</span><span class="st"> </span><span class="dv">4</span>
gene3 &lt;-<span class="st"> </span>gene2<span class="op">/</span><span class="dv">3</span> <span class="op">+</span><span class="st"> </span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">0</span>, <span class="dv">4</span>, <span class="dv">0</span>)
e &lt;-<span class="st"> </span><span class="kw">rbind</span>(gene1, gene2, gene3)
<span class="kw">dimnames</span>(e) &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="kw">paste0</span>(<span class="st">&quot;gene&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>),
                    <span class="kw">paste0</span>(<span class="st">&quot;sample&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>))
e</code></pre></div>
<pre><code>##        sample1   sample2 sample3  sample4   sample5
## gene1 1.000000  6.000000       2 4.000000  7.000000
## gene2 5.000000 10.000000       6 8.000000 11.000000
## gene3 1.666667  5.333333       2 6.666667  3.666667</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">matplot</span>(<span class="kw">t</span>(e), <span class="dt">type =</span> <span class="st">&quot;b&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;Samples&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Expression&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/disttoyexample-1.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;bioDist&quot;</span>)
<span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))
<span class="kw">plot</span>(<span class="kw">hclust</span>(<span class="kw">euc</span>(e)), <span class="dt">main =</span> <span class="st">&quot;Euclidean distane&quot;</span>)
<span class="kw">plot</span>(<span class="kw">hclust</span>(<span class="kw">cor.dist</span>(e)), <span class="dt">main =</span> <span class="st">&quot;Pearson correlational distance&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/hclbiodist-1.png" width="672" /></p>
<p>Many of the machine learning methods that are regularly used are sensitive to difference scales. This applies to unsupervised methods as well as supervised methods. Below, we show how centering and scaling the data (row-wise, hence the transposition) affects the euclidean distance but not the Pearson correlation distance.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">e2 &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">scale</span>(<span class="kw">t</span>(e)))
## or manually
## e2 &lt;- (e - rowMeans(e)) / genefilter::rowSds(e)
<span class="kw">matplot</span>(<span class="kw">t</span>(e2), <span class="dt">type =</span> <span class="st">&quot;b&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;Samples&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Expression&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/hclscale-1.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">2</span>))
<span class="kw">plot</span>(<span class="kw">hclust</span>(<span class="kw">euc</span>(e)), <span class="dt">main =</span> <span class="st">&quot;Euclidean distance&quot;</span>)
<span class="kw">plot</span>(<span class="kw">hclust</span>(<span class="kw">euc</span>(e2)), <span class="dt">main =</span> <span class="st">&quot;Euclidean distance (scaled/centred)&quot;</span>)
<span class="kw">plot</span>(<span class="kw">hclust</span>(<span class="kw">cor.dist</span>(e)), <span class="dt">main =</span> <span class="st">&quot;Pearson correlational distance&quot;</span>)
<span class="kw">plot</span>(<span class="kw">hclust</span>(<span class="kw">cor.dist</span>(e2)), <span class="dt">main =</span> <span class="st">&quot;Pearson correlational distance (scaled/centred)&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/hclscale-2.png" width="672" /></p>
<p>Finally, the clustering method itself (the <code>method</code> argument in <code>hclust</code>) will also influence the results:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ALL_sign &lt;-<span class="st"> </span>ALL_bcrneg[<span class="kw">fData</span>(ALL_bcrneg)<span class="op">$</span>adj.P.Val <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.005</span>, ]
<span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>))
<span class="kw">plot</span>(<span class="kw">hclust</span>(<span class="kw">dist</span>(ALL_sign), <span class="dt">method =</span> <span class="st">&quot;complete&quot;</span>))
<span class="kw">plot</span>(<span class="kw">hclust</span>(<span class="kw">dist</span>(ALL_sign), <span class="dt">method =</span> <span class="st">&quot;single&quot;</span>))
<span class="kw">plot</span>(<span class="kw">hclust</span>(<span class="kw">dist</span>(ALL_sign), <span class="dt">method =</span> <span class="st">&quot;average&quot;</span>))</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/allsignhcl-1.png" width="960" /></p>
</div>
</div>
<div id="k-means-clustering" class="section level2">
<h2><span class="header-section-number">5.7</span> k-means clustering</h2>
<p>The k-means clustering algorithms aims at partitioning <em>n</em> observations into a fixed number of <em>k</em> clusters. The algorithm will find homogeneous clusters.</p>
<p>In R, we use</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">stats<span class="op">::</span><span class="kw">kmeans</span>(x, <span class="dt">centers =</span> <span class="dv">3</span>, <span class="dt">nstart =</span> <span class="dv">10</span>)</code></pre></div>
<p>where</p>
<ul>
<li><code>x</code> is a numeric data matrix</li>
<li><code>centers</code> is the pre-defined number of clusters</li>
<li>the k-means algorithm has a random component and can be repeated <code>nstart</code> times to improve the returned model</li>
</ul>
<blockquote>
<p>Challenge:</p>
<ul>
<li><p>Run the k-means algorithm on the reduce <code>hlm</code> spatial proteomics dataset, save the results in a new variable <code>cl</code>, and explore its output when printed. Set the number of clusters to be the number of annotated sub-cellular niches.</p></li>
<li><p>The actual results of the algorithms, i.e. the cluster membership can be accessed in the <code>clusters</code> element of the clustering result output. Use it to colour the inferred clusters to generate a figure like that shown below.</p></li>
</ul>
</blockquote>
<p><details></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cl &lt;-<span class="st"> </span><span class="kw">kmeans</span>(MSnbase<span class="op">::</span><span class="kw">exprs</span>(hlm), <span class="kw">length</span>(<span class="kw">getMarkerClasses</span>(hlm)), <span class="dt">nstart =</span> <span class="dv">10</span>)
<span class="kw">pairs</span>(MSnbase<span class="op">::</span><span class="kw">exprs</span>(hlm)[, <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">5</span>, <span class="dv">9</span>)], <span class="dt">col =</span> cl<span class="op">$</span>cluster)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/solkmplot-1.png" width="672" /> </details></p>
<div id="how-does-k-means-work" class="section level3 unnumbered">
<h3>How does k-means work</h3>
<p>Below, we use a manually generated dataset with 3 clusters.</p>
<p><strong>Initialisation</strong>: randomly assign class membership</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">123</span>)
x &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">x =</span> <span class="kw">rnorm</span>(<span class="dv">100</span>, <span class="dv">5</span>, <span class="dv">4</span>),
        <span class="dt">y =</span> <span class="kw">rnorm</span>(<span class="dv">100</span>, <span class="dv">5</span>, <span class="dv">4</span>))
init &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="dv">3</span>, <span class="kw">nrow</span>(x), <span class="dt">replace =</span> <span class="ot">TRUE</span>)
<span class="kw">plot</span>(x, <span class="dt">col =</span> init)</code></pre></div>
<div class="figure"><span id="fig:kmworksinit"></span>
<img src="VisualisingBiomolecularData_files/figure-html/kmworksinit-1.png" alt="k-means random intialisation" width="672" />
<p class="caption">
Figure 5.4: k-means random intialisation
</p>
</div>
<p><strong>Iteration</strong>:</p>
<ol style="list-style-type: decimal">
<li>Calculate the centre of each subgroup as the average position of all observations is that subgroup.</li>
<li>Each observation is then assigned to the group of its nearest centre.</li>
</ol>
<p>It’s also possible to stop the algorithm after a certain number of iterations, or once the centres move less than a certain distance.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))
<span class="kw">plot</span>(x, <span class="dt">col =</span> init)
centres &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="cf">function</span>(i) <span class="kw">colMeans</span>(x[init <span class="op">==</span><span class="st"> </span>i, ], ))
centres &lt;-<span class="st"> </span><span class="kw">t</span>(centres)
<span class="kw">points</span>(centres[, <span class="dv">1</span>], centres[, <span class="dv">2</span>], <span class="dt">pch =</span> <span class="dv">19</span>, <span class="dt">col =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>)

tmp &lt;-<span class="st"> </span><span class="kw">dist</span>(<span class="kw">rbind</span>(centres, x))
tmp &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(tmp)[, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]

ki &lt;-<span class="st"> </span><span class="kw">apply</span>(tmp, <span class="dv">1</span>, which.min)
ki &lt;-<span class="st"> </span>ki[<span class="op">-</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>)]

<span class="kw">plot</span>(x, <span class="dt">col =</span> ki)
<span class="kw">points</span>(centres[, <span class="dv">1</span>], centres[, <span class="dv">2</span>], <span class="dt">pch =</span> <span class="dv">19</span>, <span class="dt">col =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>)</code></pre></div>
<div class="figure"><span id="fig:kmworksiter"></span>
<img src="VisualisingBiomolecularData_files/figure-html/kmworksiter-1.png" alt="k-means iteration: calculate centers (left) and assign new cluster membership (right)" width="1152" />
<p class="caption">
Figure 5.5: k-means iteration: calculate centers (left) and assign new cluster membership (right)
</p>
</div>
<p><strong>Termination</strong>: Repeat iteration until no point changes its cluster membership.</p>
<div class="figure">
<img src="https://upload.wikimedia.org/wikipedia/commons/e/ea/K-means_convergence.gif" alt="k-means convergence (credit Wikipedia)" />
<p class="caption">k-means convergence (credit Wikipedia)</p>
</div>
</div>
<div id="model-selection" class="section level3 unnumbered">
<h3>Model selection</h3>
<p>Due to the random initialisation, one can obtain different clustering results. When k-means is run multiple times, the best outcome, i.e. the one that generates the smallest <em>total within cluster sum of squares (SS)</em>, is selected. The total within SS is calculated as:</p>
<p>For each cluster results:</p>
<ul>
<li>for each observation, determine the squared euclidean distance from observation to centre of cluster</li>
<li>sum all distances</li>
</ul>
<p>Note that this is a <strong>local minimum</strong>; there is no guarantee to obtain a global minimum.</p>
<p>Below, we see that if we repeat k-means on our <code>x</code> data multiple times, setting the number of iterations to 1 or greater and check whether you repeatedly obtain the same results.</p>
<p><details></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cl1 &lt;-<span class="st"> </span><span class="kw">kmeans</span>(x, <span class="dt">centers =</span> <span class="dv">3</span>, <span class="dt">nstart =</span> <span class="dv">10</span>)
cl2 &lt;-<span class="st"> </span><span class="kw">kmeans</span>(x, <span class="dt">centers =</span> <span class="dv">3</span>, <span class="dt">nstart =</span> <span class="dv">10</span>)
<span class="kw">table</span>(cl1<span class="op">$</span>cluster, cl2<span class="op">$</span>cluster)</code></pre></div>
<pre><code>##    
##      1  2  3
##   1  0 29  0
##   2  0  0 33
##   3 38  0  0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cl3 &lt;-<span class="st"> </span><span class="kw">kmeans</span>(x, <span class="dt">centers =</span> <span class="dv">3</span>, <span class="dt">nstart =</span> <span class="dv">1</span>)
cl4 &lt;-<span class="st"> </span><span class="kw">kmeans</span>(x, <span class="dt">centers =</span> <span class="dv">3</span>, <span class="dt">nstart =</span> <span class="dv">1</span>)
<span class="kw">table</span>(cl3<span class="op">$</span>cluster, cl4<span class="op">$</span>cluster)</code></pre></div>
<pre><code>##    
##      1  2  3
##   1  0 37  1
##   2 27  0  2
##   3  0  0 33</code></pre>
<p></details></p>
</div>
<div id="how-to-determine-the-number-of-clusters" class="section level3 unnumbered">
<h3>How to determine the number of clusters</h3>
<ol style="list-style-type: decimal">
<li>Run k-means with <code>k=1</code>, <code>k=2</code>, …, <code>k=n</code></li>
<li>Record total within SS for each value of k.</li>
<li>Choose k at the <em>elbow</em> position, as illustrated below.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ks &lt;-<span class="st"> </span><span class="dv">5</span><span class="op">:</span><span class="dv">20</span>
x &lt;-<span class="st"> </span>MSnbase<span class="op">::</span><span class="kw">exprs</span>(hlm)
tot_within_ss &lt;-<span class="st"> </span><span class="kw">sapply</span>(ks, <span class="cf">function</span>(k) {
    cl &lt;-<span class="st"> </span><span class="kw">kmeans</span>(x, k, <span class="dt">nstart =</span> <span class="dv">10</span>)
    cl<span class="op">$</span>tot.withinss
})
<span class="kw">plot</span>(ks, tot_within_ss, <span class="dt">type =</span> <span class="st">&quot;b&quot;</span>,
     <span class="dt">ylab =</span> <span class="st">&quot;Total within squared distances&quot;</span>,
     <span class="dt">xlab =</span> <span class="st">&quot;Values of k tested&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/kmelbow-1.png" width="672" /></p>
<p>There exists other metrics, other than the total within cluster sum of squares that can be applied, such as the gap statistic (see <code>cluster::clusGap</code>), or the Akaike (AIC) and Bayesian (BIC) information criteria.</p>
</div>
<div id="challenge" class="section level3 unnumbered">
<h3>Challenge</h3>
<p>Let’s use what we have learned to cluster the 2337 proteins from the <code>mulvey2015norm</code> data in 20 clusters.</p>
<ol style="list-style-type: decimal">
<li><p>Use k-means to cluster the <code>mulvey2015norm</code> data, setting <code>centers    = 20</code>. Take care in repeating the clustering more than once.</p></li>
<li><p>To plot the expression profiles for the 20 clusters, I suggest to use <code>gplot2</code>. Do do so, create a 2337 proteins by 18 sample dataframe (or tibble), appending the protein accession numbers (from the feature data - you can use the <code>MSnbase::ms2df</code> helper function.) and cluster numbers as 2 additional columns.</p></li>
<li><p>Use <code>gather</code> to transform the data in a long format.</p></li>
<li><p>Use <code>ggplot2</code> to reproduce the figure below. Optional: use <code>stat_summary</code> to add a mean profile for each cluster of proteins.</p></li>
</ol>
<div class="figure">
<img src="img/kmeans-mulvey-2.png" alt="kmeans clustering on mulvey2015norm" />
<p class="caption">kmeans clustering on <code>mulvey2015norm</code></p>
</div>
<p><details></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;pRolocdata&quot;</span>)
<span class="kw">data</span>(mulvey2015norm)

cl &lt;-<span class="st"> </span><span class="kw">kmeans</span>(MSnbase<span class="op">::</span><span class="kw">exprs</span>(mulvey2015norm),
             <span class="dt">centers =</span> <span class="dv">16</span>, <span class="dt">nstart =</span> <span class="dv">10</span>, <span class="dt">iter.max =</span> <span class="dv">50</span>)

x &lt;-<span class="st"> </span><span class="kw">ms2df</span>(mulvey2015norm, <span class="dt">fcol =</span> <span class="st">&quot;Accession&quot;</span>)
x[[<span class="st">&quot;cluster&quot;</span>]] &lt;-<span class="st"> </span>cl<span class="op">$</span>cluster
tb &lt;-<span class="st"> </span><span class="kw">gather</span>(x, <span class="dt">key =</span> sample, <span class="dt">value =</span> expression, <span class="op">-</span>cluster, <span class="op">-</span>Accession) <span class="op">%&gt;%</span>
<span class="st">    </span>as_tibble

## Check dimensions
<span class="kw">stopifnot</span>(<span class="kw">nrow</span>(tb) <span class="op">==</span><span class="st"> </span><span class="kw">prod</span>(<span class="kw">dim</span>(mulvey2015norm)))

pd &lt;-<span class="st"> </span><span class="kw">pData</span>(mulvey2015norm)
tb<span class="op">$</span>time &lt;-<span class="st"> </span>pd[tb[[<span class="st">&quot;sample&quot;</span>]], <span class="st">&quot;times&quot;</span>]
tb<span class="op">$</span>rep &lt;-<span class="st"> </span>pd[tb[[<span class="st">&quot;sample&quot;</span>]], <span class="st">&quot;rep&quot;</span>]

## Plotting
kmp &lt;-<span class="st"> </span><span class="kw">ggplot</span>(<span class="dt">data =</span> tb,
              <span class="kw">aes</span>(<span class="dt">x =</span> <span class="kw">paste</span>(time, rep), <span class="dt">y =</span> expression,
                  <span class="dt">group =</span> Accession, <span class="dt">colour =</span> <span class="kw">as.factor</span>(cluster))) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>cluster) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">&quot;none&quot;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_x_discrete</span>(<span class="st">&quot;Time course&quot;</span>)

kmp2 &lt;-<span class="st"> </span>kmp <span class="op">+</span>
<span class="st">    </span><span class="kw">stat_summary</span>(<span class="kw">aes</span>(<span class="dt">group =</span> cluster),
                 <span class="dt">fun.y =</span> mean, <span class="dt">geom =</span> <span class="st">&quot;line&quot;</span>,
                 <span class="dt">colour =</span> <span class="st">&quot;black&quot;</span>)</code></pre></div>
<p></details></p>
</div>
</div>
<div id="principal-component-analysis-pca" class="section level2">
<h2><span class="header-section-number">5.8</span> Principal component analysis (PCA)</h2>
<p><strong>Dimensionality reduction</strong> techniques are widely used and versatile techniques that can be used to:</p>
<ul>
<li>find structure in features</li>
<li>pre-processing for other ML algorithms, and</li>
<li>aid in visualisation.</li>
</ul>
<p>The basic principle of dimensionality reduction techniques is to transform the data into a new space that summarise properties of the whole data set along a reduced number of dimensions. These are then ideal candidates used to visualise the data along these reduced number of informative dimensions.</p>
<div id="how-does-it-work" class="section level3 unnumbered">
<h3>How does it work</h3>
<p>Principal Component Analysis (PCA) is a technique that transforms the original n-dimensional data into a new n-dimensional space.</p>
<ul>
<li>These new dimensions are linear combinations of the original data, i.e. they are composed of proportions of the original variables.</li>
<li>Along these new dimensions, called principal components, the data expresses most of its variability along the first PC, then second, …</li>
<li>Principal components are orthogonal to each other, i.e. non-correlated.</li>
</ul>
<div class="figure"><span id="fig:pcaex"></span>
<img src="VisualisingBiomolecularData_files/figure-html/pcaex-1.png" alt="Original data (left). PC1 will maximise the variability while minimising the residuals (centre). PC2 is orthogonal to PC1 (right)." width="1152" />
<p class="caption">
Figure 5.6: Original data (left). PC1 will maximise the variability while minimising the residuals (centre). PC2 is orthogonal to PC1 (right).
</p>
</div>
<p>In R, we can use the <code>prcomp</code> function.</p>
<p>Let’s apply PCA on the <code>mulvey2015norm</code> samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pca &lt;-<span class="st"> </span><span class="kw">prcomp</span>(MSnbase<span class="op">::</span><span class="kw">exprs</span>(<span class="kw">t</span>(mulvey2015norm)))
<span class="kw">summary</span>(pca)</code></pre></div>
<pre><code>## Importance of components:
##                           PC1    PC2     PC3     PC4     PC5     PC6
## Standard deviation     12.803 5.8737 2.32468 1.92711 1.50167 1.31702
## Proportion of Variance  0.748 0.1574 0.02466 0.01695 0.01029 0.00791
## Cumulative Proportion   0.748 0.9054 0.93003 0.94698 0.95727 0.96518
##                            PC7     PC8     PC9    PC10    PC11    PC12
## Standard deviation     1.27894 1.09257 1.06266 0.96421 0.83033 0.80686
## Proportion of Variance 0.00746 0.00545 0.00515 0.00424 0.00315 0.00297
## Cumulative Proportion  0.97265 0.97809 0.98325 0.98749 0.99063 0.99360
##                           PC13    PC14    PC15   PC16    PC17      PC18
## Standard deviation     0.72439 0.64802 0.57176 0.2582 0.25217 3.468e-15
## Proportion of Variance 0.00239 0.00192 0.00149 0.0003 0.00029 0.000e+00
## Cumulative Proportion  0.99600 0.99791 0.99941 0.9997 1.00000 1.000e+00</code></pre>
<p>A summary of the <code>prcomp</code> output shows that along PC1 along, we are able to retain over 75% of the total variance in the data.</p>
</div>
<div id="visualisation" class="section level3 unnumbered">
<h3>Visualisation</h3>
<p>A <strong>biplot</strong> features all original points re-mapped (rotated) along the first two PCs as well as the original features as vectors along the same PCs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">biplot</span>(pca)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/mulveybiplot-1.png" width="672" /></p>
<p>Feature vectors that are in the same direction in PC space are also correlated in the original data space. This can be seen below when running PCA on the protein (rather than the samples).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">biplot</span>(<span class="kw">prcomp</span>(MSnbase<span class="op">::</span><span class="kw">exprs</span>(mulvey2015norm)))</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/pca2-1.png" width="672" /></p>
<p>One important piece of information when using PCA is the proportion of variance explained along the PCs, in particular when dealing with high dimensional data, as PC1 and PC2 (that are generally used for visualisation), might only account for an insufficient proportion of variance to be relevant on their own.</p>
<p>In the code chunk below, I extract the standard deviations from the PCA result to calculate the variances, then obtain the percentage of and cumulative variance along the PCs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">var &lt;-<span class="st"> </span>pca<span class="op">$</span>sdev<span class="op">^</span><span class="dv">2</span>
(pve &lt;-<span class="st"> </span>var<span class="op">/</span><span class="kw">sum</span>(var))</code></pre></div>
<pre><code>##  [1] 7.479554e-01 1.574190e-01 2.465846e-02 1.694543e-02 1.028939e-02
##  [6] 7.914534e-03 7.463453e-03 5.446791e-03 5.152624e-03 4.242092e-03
## [11] 3.145870e-03 2.970534e-03 2.394341e-03 1.916061e-03 1.491659e-03
## [16] 3.042167e-04 2.901557e-04 5.489151e-32</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">cumsum</span>(pve)</code></pre></div>
<pre><code>##  [1] 0.7479554 0.9053744 0.9300329 0.9469783 0.9572677 0.9651822 0.9726457
##  [8] 0.9780924 0.9832451 0.9874872 0.9906330 0.9936036 0.9959979 0.9979140
## [15] 0.9994056 0.9997098 1.0000000 1.0000000</code></pre>
<blockquote>
<p>Challenge</p>
<ul>
<li><p>Repeat the PCA analysis above, reproducing the biplot and preparing a barplot of the percentage of variance explained by each PC.</p></li>
<li><p>It is often useful to produce custom figures using the data coordinates in PCA space, which can be accessed as <code>x</code> in the <code>prcomp</code> object. Reproduce the PCA plots below, along PC1 and PC2 and PC3 and PC4 respectively.</p></li>
</ul>
</blockquote>
<p><img src="VisualisingBiomolecularData_files/figure-html/pcax-1.png" width="1152" /></p>
<p><details></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))
<span class="kw">plot</span>(pca<span class="op">$</span>x[, <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>], <span class="dt">col =</span> mulvey2015norm<span class="op">$</span>times)
<span class="kw">plot</span>(pca<span class="op">$</span>x[, <span class="dv">3</span><span class="op">:</span><span class="dv">4</span>], <span class="dt">col =</span> mulvey2015norm<span class="op">$</span>times)</code></pre></div>
<p></details></p>
</div>
<div id="data-pre-processing" class="section level3 unnumbered">
<h3>Data pre-processing</h3>
<p>We haven’t looked at other <code>prcomp</code> parameters, other that the first one, <code>x</code>. There are two other ones that are or importance, in particular in the light of the section on pre-processing above, which are <code>center</code> and <code>scale.</code>. The former is set to <code>TRUE</code> by default, while the second one is set the <code>FALSE</code>.</p>
<p><details></p>
<p><strong>Example</strong> Repeat the analysis comparing the need for scaling on the <code>mtcars</code> dataset, but using PCA instead of hierarchical clustering. When comparing the two.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))
<span class="kw">biplot</span>(<span class="kw">prcomp</span>(mtcars, <span class="dt">scale =</span> <span class="ot">FALSE</span>), <span class="dt">main =</span> <span class="st">&quot;No scaling&quot;</span>)  ## 1
<span class="kw">biplot</span>(<span class="kw">prcomp</span>(mtcars, <span class="dt">scale =</span> <span class="ot">TRUE</span>), <span class="dt">main =</span> <span class="st">&quot;With scaling&quot;</span>) ## 2</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/scalepcasol-1.png" width="672" /></p>
<p>Without scaling, <code>disp</code> and <code>hp</code> are the features with the highest loadings along PC1 and 2 (all others are negligible), which are also those with the highest units of measurement. Scaling removes this effect.</p>
<p></details></p>
</div>
<div id="final-comments-on-pca" class="section level3 unnumbered">
<h3>Final comments on PCA</h3>
<p>Real datasets often come with <strong>missing values</strong>. In R, these should be encoded using <code>NA</code>. Unfortunately, standard PCA cannot deal with missing values, and observations containing <code>NA</code> values will be dropped automatically. This is a viable solution only when the proportion of missing values is low. Alternatively, the <em>NIPALS</em> (non-linear iterative partial least squares) implementation does support missing values (see <code>nipals::nipals</code>).</p>
<p>Finally, we should be careful when using categorical data in any of the unsupervised methods described above. Categories are generally represented as factors, which are encoded as integer levels, and might give the impression that a distance between levels is a relevant measure (which it is not, unless the factors are ordered). In such situations, categorical data can be dropped, or it is possible to encode categories as binary <strong>dummy variables</strong>. For example, if we have 3 categories, say <code>A</code>, <code>B</code> and <code>C</code>, we would create two dummy variables to encode the categories as:</p>
<pre><code>##   x y
## A 1 0
## B 0 1
## C 0 0</code></pre>
<p>so that the distance between each category are approximately equal to 1.</p>
</div>
<div id="challenge-1" class="section level3 unnumbered">
<h3>Challenge</h3>
<p>Produce the PCA plot for the <code>ALL_bcrneg</code> samples, and annotating the <code>NEG</code> and <code>BCR/ABL</code> samples on the plot. Do you think that the two first components offer enough resolution?</p>
<p><details></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pca &lt;-<span class="st"> </span><span class="kw">prcomp</span>(<span class="kw">t</span>(MSnbase<span class="op">::</span><span class="kw">exprs</span>(ALL_bcrneg)), <span class="dt">scale =</span> <span class="ot">TRUE</span>, <span class="dt">center =</span> <span class="ot">TRUE</span>)
<span class="kw">plot</span>(pca<span class="op">$</span>x[, <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>], <span class="dt">col =</span> ALL_bcrneg<span class="op">$</span>mol.bio, <span class="dt">cex =</span> <span class="dv">2</span>)
<span class="kw">legend</span>(<span class="st">&quot;bottomright&quot;</span>, <span class="dt">legend =</span> <span class="kw">unique</span>(ALL_bcrneg<span class="op">$</span>mol.bio),
       <span class="dt">pch =</span> <span class="dv">1</span>, <span class="dt">col =</span> <span class="kw">c</span>(<span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>), <span class="dt">bty =</span> <span class="st">&quot;n&quot;</span>)
<span class="kw">text</span>(pca<span class="op">$</span>x[, <span class="dv">1</span>], pca<span class="op">$</span>x[, <span class="dv">2</span>], <span class="kw">sampleNames</span>(ALL_bcrneg), <span class="dt">cex =</span> <span class="fl">0.8</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/pcachallenge0-1.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))
<span class="kw">barplot</span>(pca<span class="op">$</span>sdev<span class="op">^</span><span class="dv">2</span><span class="op">/</span><span class="kw">sum</span>(pca<span class="op">$</span>sdev<span class="op">^</span><span class="dv">2</span>),
        <span class="dt">xlab=</span><span class="st">&quot;Principle component&quot;</span>,
        <span class="dt">ylab=</span><span class="st">&quot;% of variance&quot;</span>)

<span class="kw">barplot</span>(<span class="kw">cumsum</span>(pca<span class="op">$</span>sdev<span class="op">^</span><span class="dv">2</span><span class="op">/</span><span class="kw">sum</span>(pca<span class="op">$</span>sdev<span class="op">^</span><span class="dv">2</span>) ),
        <span class="dt">xlab=</span><span class="st">&quot;Principle component&quot;</span>,
        <span class="dt">ylab=</span><span class="st">&quot;Pumulative % of variance&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/pcachallenge0-2.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Conclusion: the two first principle components are insufficient</code></pre></div>
<p></details></p>
<p>Repeat the kmeans clustering analysis on the <code>hml</code> data above and visualise the results on a PCA plot.</p>
<p><details></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cl &lt;-<span class="st"> </span><span class="kw">kmeans</span>(MSnbase<span class="op">::</span><span class="kw">exprs</span>(hlm), <span class="kw">length</span>(<span class="kw">getMarkerClasses</span>(hlm)), <span class="dt">nstart =</span> <span class="dv">10</span>)
pca &lt;-<span class="st"> </span><span class="kw">prcomp</span>(MSnbase<span class="op">::</span><span class="kw">exprs</span>(hlm), <span class="dt">scale =</span> <span class="ot">TRUE</span>, <span class="dt">center =</span> <span class="ot">TRUE</span>)
<span class="kw">plot</span>(pca<span class="op">$</span>x[, <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>], <span class="dt">col =</span> cl<span class="op">$</span>cluster)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/pcachallenge2-1.png" width="672" /> </details></p>
<p>Produce a PCA with the proteins of the complete <code>hyperLopit2015</code> experiment. Label the results using <code>final.assignment</code> feature variable (corresponding to the proteins assignment to they sub-cellular localisation) and scale to size of the points based on the assignment scores available in the <code>svm.score</code> feature variable (this might need to transform the scores to make differences visisble).</p>
<p><details></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">prcomp</span>(MSnbase<span class="op">::</span><span class="kw">exprs</span>(hyperLOPIT2015))
<span class="kw">plot</span>(x<span class="op">$</span>x[, <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>],
     <span class="dt">pch =</span> <span class="kw">ifelse</span>(<span class="kw">fData</span>(hyperLOPIT2015)<span class="op">$</span>final.assignment <span class="op">==</span><span class="st"> &quot;unknown&quot;</span>, <span class="dv">1</span>, <span class="dv">19</span>),
     <span class="dt">col =</span> <span class="kw">as.factor</span>(<span class="kw">fData</span>(hyperLOPIT2015)<span class="op">$</span>final.assignment),
     <span class="dt">cex =</span> <span class="kw">exp</span>(<span class="kw">fData</span>(hyperLOPIT2015)<span class="op">$</span>svm.score) <span class="op">-</span><span class="st"> </span><span class="dv">1</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/pcachallenge3-1.png" width="672" /> </details></p>
</div>
</div>
<div id="t-distributed-stochastic-neighbour-embedding" class="section level2">
<h2><span class="header-section-number">5.9</span> t-Distributed Stochastic Neighbour Embedding</h2>
<p><a href="https://lvdmaaten.github.io/tsne/">t-Distributed Stochastic Neighbour Embedding</a> (t-SNE) is a <em>non-linear</em> dimensionality reduction technique, i.e. that different regions of the data space will be subjected to different transformations. t-SNE will compress small distances, thus bringing close neighbours together, and will ignore large distances. It is particularly well suited for <a href="https://distill.pub/2016/misread-tsne/">very high dimensional data</a>.</p>
<p>In R, we can use the <code>Rtsne</code> function from the <em><a href="https://CRAN.R-project.org/package=Rtsne">Rtsne</a></em>.</p>
<p>As with PCA, the data can be scaled and centred prior the running t-SNE (see the <code>pca_center</code> and <code>pca_scale</code> arguments). The algorithm is stochastic, and will produce different results at each repetition.</p>
<div id="parameter-tuning" class="section level3">
<h3><span class="header-section-number">5.9.1</span> Parameter tuning</h3>
<p>t-SNE has two important parameters that can substantially influence the clustering of the data</p>
<ul>
<li><strong>Perplexity</strong>: balances global and local aspects of the data.</li>
<li><strong>Iterations</strong>: number of iterations before the clustering is stopped.</li>
</ul>
<p>It is important to adapt these for different data. The figure below shows a 5032 by 20 dataset that represent protein sub-cellular localisation.</p>
<div class="figure">
<img src="https://raw.githubusercontent.com/lgatto/visualisation/master/figure/tsnesplots-1.png" alt="Effect of different perplexity and iterations when running t-SNE" />
<p class="caption">Effect of different perplexity and iterations when running t-SNE</p>
</div>
<p>As a comparison, below are the same data with PCA (left) and t-SNE (right).</p>
<div class="figure">
<img src="https://raw.githubusercontent.com/lgatto/visualisation/master/figure/tsneex-1.png" alt="PCA and t-SNE on hyperLOPIT" />
<p class="caption">PCA and t-SNE on hyperLOPIT</p>
</div>
</div>
</div>
<div id="pca-loadings-vs-differential-expression" class="section level2">
<h2><span class="header-section-number">5.10</span> PCA loadings vs differential expression</h2>
<p>Let’s now compare the PCA loadings as calculated above and the p-values. As we can see below, large loadings may or may not correspond to small p-values.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pca &lt;-<span class="st"> </span><span class="kw">prcomp</span>(<span class="kw">t</span>(MSnbase<span class="op">::</span><span class="kw">exprs</span>(ALL_bcrneg)), <span class="dt">scale =</span> <span class="ot">TRUE</span>, <span class="dt">center =</span> <span class="ot">TRUE</span>)
<span class="kw">plot</span>(pca<span class="op">$</span>rotation[, <span class="dv">1</span>], <span class="kw">fData</span>(ALL_bcrneg)<span class="op">$</span>adj.P.Val,
     <span class="dt">xlab =</span> <span class="st">&quot;PCA Loading&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Adjusted p-values&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/pcaloadings-1.png" width="672" /></p>
<p>Let’s now repeat the same camparison, focusing on differentially expressed genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(sign &lt;-<span class="st"> </span><span class="kw">fData</span>(ALL_bcrneg)<span class="op">$</span>adj.P.Val <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>)</code></pre></div>
<pre><code>## 
## FALSE  TRUE 
## 12442   183</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ALL_sign &lt;-<span class="st"> </span>ALL_bcrneg[sign, ]
pca_sign &lt;-<span class="st"> </span><span class="kw">prcomp</span>(<span class="kw">t</span>(MSnbase<span class="op">::</span><span class="kw">exprs</span>(ALL_sign)), <span class="dt">center =</span> <span class="ot">TRUE</span>, <span class="dt">scale =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>Below, we see that there is better separation when we focus on differentially expressed genes and better consistency between p-values and loadings. However we can’t do this in practice!</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(pca_sign<span class="op">$</span>x[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>], <span class="dt">col =</span> ALL_sign<span class="op">$</span>mol.biol, <span class="dt">pch =</span> <span class="dv">19</span>)
<span class="kw">legend</span>(<span class="st">&quot;bottomleft&quot;</span>, <span class="dt">legend =</span> <span class="kw">unique</span>(ALL_sign<span class="op">$</span>mol.biol),
       <span class="dt">col =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, <span class="dt">pch =</span> <span class="dv">19</span>, <span class="dt">bty =</span> <span class="st">&quot;n&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/pcasign-1.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(pca_sign<span class="op">$</span>rotation[, <span class="dv">1</span>], <span class="kw">fData</span>(ALL_sign)<span class="op">$</span>adj.P.Val,
     <span class="dt">xlab =</span> <span class="st">&quot;PCA Loading&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Adjusted p-values&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/pcasign-2.png" width="672" /></p>
</div>
<div id="heatmaps" class="section level2">
<h2><span class="header-section-number">5.11</span> Heatmaps</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">heatmap</span>(MSnbase<span class="op">::</span><span class="kw">exprs</span>(mulvey2015norm))</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/heatmap_mulvey-1.png" width="672" /></p>
<p>A heatmap is composed of two hierarchical clusters (one along the rows, one along the columns, leading to their re-ordering based on their similarity) and a intensity matrix. Each of these components is subject to parameters and options.</p>
<p>As we have seen above, the distance used for clustering can have a substantial effect on the results, which is conformed below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">heatmap</span>(MSnbase<span class="op">::</span><span class="kw">exprs</span>(mulvey2015norm), <span class="dt">distfun =</span> cor.dist)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/heatmapcor-1.png" width="672" /></p>
<p>Another important argument, <code>scale</code> controls whether rows, columns or none are scaled. Let’s re-use the toy data from the hierarchical clustering section below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">heatmap</span>(e, <span class="dt">scale =</span> <span class="st">&quot;none&quot;</span>, <span class="dt">main =</span> <span class="st">&quot;No scaling&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/heatmapscale-1.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">heatmap</span>(e, <span class="dt">scale =</span> <span class="st">&quot;row&quot;</span>, <span class="dt">main =</span> <span class="st">&quot;Scaling along rows&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/heatmapscale-2.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">heatmap</span>(e, <span class="dt">scale =</span> <span class="st">&quot;column&quot;</span>, <span class="dt">main =</span> <span class="st">&quot;Scaling along columns&quot;</span>)</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/heatmapscale-3.png" width="672" /></p>
<p>Based on the caveats above, it is essential to present and interpret heatmaps with great care. It is of course possible to use any type of data, not only expression data, to build a heatmap.</p>
<p>There exists several packages that allow to produce heatmaps with various levels of sophistication, such as <code>heatmap.2</code> from the <em><a href="https://CRAN.R-project.org/package=gplots">gplots</a></em> package, the <em><a href="http://bioconductor.org/packages/Heatplus">Heatplus</a></em> package, or the <em><a href="http://bioconductor.org/packages/ComplexHeatmap">ComplexHeatmap</a></em> packages, demonstrated below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;ComplexHeatmap&quot;</span>)
x &lt;-<span class="st"> </span>MSnbase<span class="op">::</span><span class="kw">exprs</span>(mulvey2015norm)
hcl &lt;-<span class="st"> </span><span class="kw">hclust</span>(<span class="kw">dist</span>(x))
cl &lt;-<span class="st"> </span><span class="kw">cutree</span>(hcl, <span class="dt">k =</span> <span class="dv">12</span>)
ha1 &lt;-<span class="st"> </span><span class="kw">HeatmapAnnotation</span>(<span class="kw">data.frame</span>(<span class="dt">time =</span> <span class="kw">factor</span>(mulvey2015norm<span class="op">$</span>time)))
ha2 &lt;-<span class="st"> </span><span class="kw">HeatmapAnnotation</span>(<span class="dt">boxplot =</span> <span class="kw">anno_boxplot</span>(x))
ha3 &lt;-<span class="st"> </span><span class="kw">rowAnnotation</span>(<span class="kw">data.frame</span>(<span class="dt">cluster =</span> <span class="kw">factor</span>(cl)))
<span class="kw">Heatmap</span>(x,
        <span class="dt">top_annotation =</span> ha1,
        <span class="dt">bottom_annotation =</span> ha2,
        <span class="dt">column_names_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>),
        <span class="dt">row_names_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">3</span>),
        <span class="dt">bottom_annotation_height =</span> <span class="kw">unit</span>(<span class="dv">3</span>, <span class="st">&quot;cm&quot;</span>)) <span class="op">+</span><span class="st"> </span>ha3</code></pre></div>
<p><img src="VisualisingBiomolecularData_files/figure-html/complexheatmap-1.png" width="672" /></p>
<p>Finally, the <em><a href="https://github.com/talgalili/heatmaply">heatmaply</a></em> can be used to generate interactive heatmaps.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;heatmaply&quot;</span>)
<span class="kw">heatmaply</span>(x[<span class="dv">1</span><span class="op">:</span><span class="dv">100</span>, ])</code></pre></div>
<p>Reference:</p>
<blockquote>
<p>Key M. <em>A tutorial in displaying mass spectrometry-based proteomic data using heat maps</em>. BMC Bioinformatics. 2012;13 Suppl 16:S10. doi: <a href="https://doi.org/10.1186/1471-2105-13-S16-S10">10.1186/1471-2105-13-S16-S10</a>. Epub 2012 Nov 5. Review. PMID: 23176119; PMCID: PMC3489527.</p>
</blockquote>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="plotting-in-r.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="using-r-and-bioconductor-for-ms-based-proteomics.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
